--- a/src/main/java/io/papermc/paper/util/MCUtil.java
+++ b/src/main/java/io/papermc/paper/util/MCUtil.java
@@ -26,7 +_,6 @@
 import net.minecraft.world.phys.Vec3;
 import org.bukkit.Location;
 import org.bukkit.NamespacedKey;
-import org.bukkit.craftbukkit.util.CraftLocation;
 import org.bukkit.craftbukkit.util.CraftNamespacedKey;
 import org.bukkit.craftbukkit.util.Waitable;
 
@@ -155,22 +_,16 @@
 
     /**
      * Converts a NMS World/Vector to Bukkit Location
-     *
-     * @deprecated use CraftLocation methods instead
      */
-    @Deprecated
     public static Location toLocation(Level world, Vec3 pos) {
-        return CraftLocation.toBukkit(pos, world);
+        return new Location(world.getWorld(), pos.x(), pos.y(), pos.z());
     }
 
     /**
      * Converts a NMS World/Vector to Bukkit Location
-     *
-     * @deprecated use CraftLocation methods instead
      */
-    @Deprecated
     public static Location toLocation(Level world, Vec3 pos, float yaw, float pitch) {
-        return CraftLocation.toBukkit(pos, world, yaw, pitch);
+        return new Location(world.getWorld(), pos.x(), pos.y(), pos.z(), yaw, pitch);
     }
 
     public static BlockPos toBlockPos(Position pos) {
