--- a/src/main/java/io/papermc/paper/util/ServerWorkerThread.java
+++ b/src/main/java/io/papermc/paper/util/ServerWorkerThread.java
@@ -5,9 +_,9 @@
 
 public class ServerWorkerThread extends Thread {
     private static final AtomicInteger threadId = new AtomicInteger(1);
-    public ServerWorkerThread(Runnable target, String poolName, int priorityModifier) {
+    public ServerWorkerThread(Runnable target, String poolName, int prioritityModifier) {
         super(target, "Worker-" + poolName + "-" + threadId.getAndIncrement());
-        setPriority(Thread.NORM_PRIORITY + priorityModifier); // Deprioritize over main
+        setPriority(Thread.NORM_PRIORITY+prioritityModifier); // Deprioritize over main
         this.setDaemon(true);
         this.setUncaughtExceptionHandler(Util::onThreadException);
     }
