--- a/src/main/java/org/bukkit/craftbukkit/Main.java
+++ b/src/main/java/org/bukkit/craftbukkit/Main.java
@@ -87,7 +_,7 @@
                         .ofType(SimpleDateFormat.class)
                         .describedAs("Log date format");
 
-                this.accepts("log-pattern", "Specifies the log filename pattern")
+                this.accepts("log-pattern", "Specfies the log filename pattern")
                         .withRequiredArg()
                         .ofType(String.class)
                         .defaultsTo("server.log")
