From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Ivo Quiring <falk.quiring@gmail.com>
Date: Mon, 28 Jul 2025 15:55:09 +0200
Subject: [PATCH] Add pulvinar to bukkit


diff --git a/src/main/java/net/kissenpvp/api/base/Pulvinar.java b/src/main/java/net/kissenpvp/api/base/Pulvinar.java
new file mode 100644
index 0000000000000000000000000000000000000000..1d8db257982e258d5a367debdbc471adf38670fd
--- /dev/null
+++ b/src/main/java/net/kissenpvp/api/base/Pulvinar.java
@@ -0,0 +1,6 @@
+package net.kissenpvp.api.base;
+
+/**
+ * Represents the Pulvinar core which provides access to the kissen functions.
+ */
+public interface Pulvinar extends Kissen {}
diff --git a/src/main/java/net/kissenpvp/api/event/SystemMessageEvent.java b/src/main/java/net/kissenpvp/api/event/SystemMessageEvent.java
new file mode 100644
index 0000000000000000000000000000000000000000..4b35e5def2c99431e27f34ed7210fed55a0cde0f
--- /dev/null
+++ b/src/main/java/net/kissenpvp/api/event/SystemMessageEvent.java
@@ -0,0 +1,52 @@
+package net.kissenpvp.api.event;
+
+import net.kyori.adventure.text.Component;
+import org.bukkit.event.Cancellable;
+import org.bukkit.event.Event;
+import org.bukkit.event.HandlerList;
+import org.jetbrains.annotations.NotNull;
+
+public class SystemMessageEvent extends Event implements Cancellable
+{
+    private static final HandlerList HANDLER_LIST = new HandlerList();
+
+    private Component message;
+    private boolean cancelled;
+
+    public SystemMessageEvent(@NotNull Component message)
+    {
+        this.message = message;
+        cancelled = false;
+    }
+
+    public @NotNull Component getMessage()
+    {
+        return message;
+    }
+
+    public void setMessage(@NotNull Component message)
+    {
+        this.message = message;
+    }
+
+    @Override public @NotNull HandlerList getHandlers()
+    {
+        return HANDLER_LIST;
+    }
+
+    @NotNull
+    public static HandlerList getHandlerList()
+    {
+        return HANDLER_LIST;
+    }
+
+    @Override public boolean isCancelled()
+    {
+        return cancelled;
+    }
+
+    @Override public void setCancelled(boolean cancel)
+    {
+        cancelled = cancel;
+    }
+}
diff --git a/src/main/java/org/bukkit/Bukkit.java b/src/main/java/org/bukkit/Bukkit.java
index bb21401d142e2a9385d52afe40c54d793103afe2..59b323e021e2ddd2224252a2e6bfb008c372157a 100644
--- a/src/main/java/org/bukkit/Bukkit.java
+++ b/src/main/java/org/bukkit/Bukkit.java
@@ -15,6 +15,7 @@ import java.util.UUID;
 import java.util.function.Consumer;
 import java.util.logging.Logger;
 import io.papermc.paper.configuration.ServerConfiguration;
+import net.kissenpvp.api.base.Pulvinar;
 import net.kyori.adventure.text.Component;
 import org.bukkit.Warning.WarningState;
 import org.bukkit.advancement.Advancement;
@@ -71,6 +72,23 @@ import org.jetbrains.annotations.Nullable;
 public final class Bukkit {
     private static Server server;
 
+    // Pulvinar start - add base instance
+    private static Pulvinar pulvinar;
+
+    @NotNull
+    public static Pulvinar getPulvinar() {
+        return pulvinar;
+    }
+
+    public static void setPulvinar(@NotNull Pulvinar pulvinar) {
+        if (Bukkit.pulvinar != null) {
+            throw new UnsupportedOperationException("Cannot redefine singleton Pulvinar");
+        }
+
+        Bukkit.pulvinar = pulvinar;
+    }
+    // Pulvinar end - add base instance
+
     /**
      * Static class cannot be initialized.
      */
diff --git a/src/main/java/org/bukkit/OfflinePlayer.java b/src/main/java/org/bukkit/OfflinePlayer.java
index c1f19d5c1b819873fce1185b4fbf4c72eda26309..22197254808819bef8542586faf6a23f8b380b67 100644
--- a/src/main/java/org/bukkit/OfflinePlayer.java
+++ b/src/main/java/org/bukkit/OfflinePlayer.java
@@ -20,7 +20,7 @@ import org.jspecify.annotations.Nullable;
  * player needing to be online.
  */
 @NullMarked
-public interface OfflinePlayer extends ServerOperator, AnimalTamer, ConfigurationSerializable, io.papermc.paper.persistence.PersistentDataViewHolder { // Paper - Add Offline PDC API
+public interface OfflinePlayer extends ServerOperator, AnimalTamer, ConfigurationSerializable, io.papermc.paper.persistence.PersistentDataViewHolder, net.kissenpvp.api.network.actor.PlayerClient { // Paper - Add Offline PDC API // Pulvinar - Add PlayerClient Actor
 
     /**
      * Checks if this player is currently online
diff --git a/src/main/java/org/bukkit/command/ConsoleCommandSender.java b/src/main/java/org/bukkit/command/ConsoleCommandSender.java
index f309c2ed39bcc700e5b491ff549bf0608a07aab8..d1cd49aebffe50859634118ec3ec324bead8845d 100644
--- a/src/main/java/org/bukkit/command/ConsoleCommandSender.java
+++ b/src/main/java/org/bukkit/command/ConsoleCommandSender.java
@@ -2,5 +2,5 @@ package org.bukkit.command;
 
 import org.bukkit.conversations.Conversable;
 
-public interface ConsoleCommandSender extends CommandSender, Conversable {
+public interface ConsoleCommandSender extends CommandSender, Conversable, net.kissenpvp.api.network.actor.ConsoleClient { // Pulvinar
 }
diff --git a/src/main/java/org/bukkit/entity/HumanEntity.java b/src/main/java/org/bukkit/entity/HumanEntity.java
index 90cef1ab811571e0352e2e704b11b0ab6f777883..8b73113807f044c99e7395a36e13aa22c979dd36 100644
--- a/src/main/java/org/bukkit/entity/HumanEntity.java
+++ b/src/main/java/org/bukkit/entity/HumanEntity.java
@@ -27,7 +27,7 @@ import org.jspecify.annotations.Nullable;
  * Represents a human entity, such as an NPC or a player
  */
 @NullMarked
-public interface HumanEntity extends LivingEntity, AnimalTamer, InventoryHolder {
+public interface HumanEntity extends LivingEntity, AnimalTamer, InventoryHolder, net.kissenpvp.api.network.actor.Actor { // Pulvinar - Make all human entities also extend net.kissenpvp.api.network.actor.Actor
 
     // Paper start
     @Override
diff --git a/src/main/java/org/bukkit/entity/Player.java b/src/main/java/org/bukkit/entity/Player.java
index 18ab7ed5819510f6ee6de380cdbb0cff2493f58f..9100d5af3882a261f24674a2a4d1bd51c01e97c6 100644
--- a/src/main/java/org/bukkit/entity/Player.java
+++ b/src/main/java/org/bukkit/entity/Player.java
@@ -63,7 +63,7 @@ import org.jspecify.annotations.Nullable;
  * Represents a player, connected or not
  */
 @NullMarked
-public interface Player extends HumanEntity, Conversable, OfflinePlayer, PluginMessageRecipient, net.kyori.adventure.identity.Identified, net.kyori.adventure.bossbar.BossBarViewer, com.destroystokyo.paper.network.NetworkClient { // Paper
+public interface Player extends HumanEntity, Conversable, OfflinePlayer, PluginMessageRecipient, net.kyori.adventure.identity.Identified, net.kyori.adventure.bossbar.BossBarViewer, com.destroystokyo.paper.network.NetworkClient, net.kissenpvp.api.network.actor.ReachablePlayerClient { // Paper // Pulvinar - Add ReachablePlayerClient actor
 
     // Paper start
     @Override
diff --git a/src/main/java/org/bukkit/plugin/Plugin.java b/src/main/java/org/bukkit/plugin/Plugin.java
index 95d2ada67d86bc9d764d4455f4ef74ce88621695..9f57d0f100c3dd9506a1c8203ac5f74fc55a5918 100644
--- a/src/main/java/org/bukkit/plugin/Plugin.java
+++ b/src/main/java/org/bukkit/plugin/Plugin.java
@@ -17,7 +17,7 @@ import org.jetbrains.annotations.Nullable;
  * <p>
  * The use of {@link PluginBase} is recommended for actual Implementation
  */
-public interface Plugin extends TabExecutor, io.papermc.paper.plugin.lifecycle.event.LifecycleEventOwner, Namespaced { // Paper
+public interface Plugin extends TabExecutor, io.papermc.paper.plugin.lifecycle.event.LifecycleEventOwner, Namespaced, net.kissenpvp.api.base.KissenPlugin { // Paper // Pulvinar
     /**
      * Returns the folder that the plugin data files are located in. The
      * folder may not yet exist.
