From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Ivo Quiring <falk.quiring@gmail.com>
Date: Fri, 24 Oct 2025 21:36:45 +0200
Subject: [PATCH] Add Kissen Api Interface


diff --git a/src/main/java/net/kissenpvp/api/base/Pulvinar.java b/src/main/java/net/kissenpvp/api/base/Pulvinar.java
new file mode 100644
index 0000000000000000000000000000000000000000..cf64189c9135e54c5ab15f60a087f6c63d925291
--- /dev/null
+++ b/src/main/java/net/kissenpvp/api/base/Pulvinar.java
@@ -0,0 +1,25 @@
+package net.kissenpvp.api.base;
+
+
+import net.kissenpvp.api.database.ConnectionProvider;
+import org.jetbrains.annotations.NotNull;
+import org.jspecify.annotations.NonNull;
+
+/**
+ * Represents Pulvinar which provides access to the kissen functions.
+ */
+public interface Pulvinar extends Kissen {
+
+    /**
+     * Provides access to the server's private database connection provider.
+     * <p>
+     * The returned provider exposes the JDBC connection used exclusively by this
+     * server instance. This is primarily intended for server-specific operations and
+     * is especially important in proxy-based environments where each server
+     * requires its own isolated database connection.
+     *
+     * @return the private {@link ConnectionProvider} for this server instance
+     */
+    @NonNull @NotNull ConnectionProvider privateConnectionProvider();
+
+}
\ No newline at end of file
diff --git a/src/main/java/org/bukkit/Bukkit.java b/src/main/java/org/bukkit/Bukkit.java
index cc57cbe8cd3d2fe1d7e9709d03579f45cf6d78ed..2036cfdbfa47514a548b072a774469c11ac5143a 100644
--- a/src/main/java/org/bukkit/Bukkit.java
+++ b/src/main/java/org/bukkit/Bukkit.java
@@ -70,6 +70,23 @@ import org.jetbrains.annotations.Nullable;
 public final class Bukkit {
     private static Server server;
 
+    // Pulvinar start - add kissen api interface
+    private static net.kissenpvp.api.base.Pulvinar pulvinar;
+
+    @NotNull
+    public static net.kissenpvp.api.base.Pulvinar getPulvinar() {
+        return pulvinar;
+    }
+
+    public static void setPulvinar(@NotNull net.kissenpvp.api.base.Pulvinar pulvinar) {
+        if (Bukkit.pulvinar != null) {
+            throw new UnsupportedOperationException("Cannot redefine singleton Pulvinar");
+        }
+
+        Bukkit.pulvinar = pulvinar;
+    }
+    // Pulvinar end - add kissen api interface
+
     /**
      * Static class cannot be initialized.
      */
