--- a/src/main/java/io/papermc/paper/PaperServerInternalAPIBridge.java
+++ b/src/main/java/io/papermc/paper/PaperServerInternalAPIBridge.java
@@ -1,20 +_,13 @@
 package io.papermc.paper;
 
-import com.destroystokyo.paper.PaperSkinParts;
-import com.destroystokyo.paper.SkinParts;
-import io.papermc.paper.adventure.PaperAdventure;
 import io.papermc.paper.command.brigadier.CommandSourceStack;
-import io.papermc.paper.datacomponent.item.PaperResolvableProfile;
-import io.papermc.paper.datacomponent.item.ResolvableProfile;
 import io.papermc.paper.world.damagesource.CombatEntry;
 import io.papermc.paper.world.damagesource.FallLocationType;
 import io.papermc.paper.world.damagesource.PaperCombatEntryWrapper;
 import io.papermc.paper.world.damagesource.PaperCombatTrackerWrapper;
-import net.kyori.adventure.text.Component;
 import net.minecraft.Optionull;
 import net.minecraft.commands.PermissionSource;
 import net.minecraft.world.damagesource.FallLocation;
-import net.minecraft.world.entity.decoration.Mannequin;
 import org.bukkit.block.Biome;
 import org.bukkit.craftbukkit.block.CraftBiome;
 import org.bukkit.craftbukkit.damage.CraftDamageEffect;
@@ -92,20 +_,5 @@
         }
 
         return new RestrictedPredicate(predicate);
-    }
-
-    @Override
-    public ResolvableProfile defaultMannequinProfile() {
-        return new PaperResolvableProfile(net.minecraft.world.entity.decoration.Mannequin.DEFAULT_PROFILE);
-    }
-
-    @Override
-    public SkinParts.Mutable allSkinParts() {
-        return new PaperSkinParts.Mutable(net.minecraft.world.entity.decoration.Mannequin.ALL_LAYERS);
-    }
-
-    @Override
-    public Component defaultMannequinDescription() {
-        return PaperAdventure.asAdventure(Mannequin.DEFAULT_DESCRIPTION);
     }
 }
