--- a/src/main/java/org/bukkit/craftbukkit/potion/CraftPotionUtil.java
+++ b/src/main/java/org/bukkit/craftbukkit/potion/CraftPotionUtil.java
@@ -15,6 +_,7 @@
 
     private static final BiMap<PotionType, PotionType> upgradeable = ImmutableBiMap.<PotionType, PotionType>builder()
             // Start generate - CraftPotionUtil#upgradeable
+            // @GeneratedFrom 1.21.8
             .put(PotionType.HARMING, PotionType.STRONG_HARMING)
             .put(PotionType.HEALING, PotionType.STRONG_HEALING)
             .put(PotionType.LEAPING, PotionType.STRONG_LEAPING)
@@ -28,6 +_,7 @@
             .build();
     private static final BiMap<PotionType, PotionType> extendable = ImmutableBiMap.<PotionType, PotionType>builder()
             // Start generate - CraftPotionUtil#extendable
+            // @GeneratedFrom 1.21.8
             .put(PotionType.FIRE_RESISTANCE, PotionType.LONG_FIRE_RESISTANCE)
             .put(PotionType.INVISIBILITY, PotionType.LONG_INVISIBILITY)
             .put(PotionType.LEAPING, PotionType.LONG_LEAPING)
