--- a/src/main/java/org/bukkit/craftbukkit/inventory/CraftItemCraftResult.java
+++ b/src/main/java/org/bukkit/craftbukkit/inventory/CraftItemCraftResult.java
@@ -14,12 +_,12 @@
     private final List<ItemStack> overflowItems;
 
     public CraftItemCraftResult(ItemStack result) {
-        this.result = result;
+        this.result = Objects.requireNonNullElseGet(result, () -> new ItemStack(Material.AIR));
         this.resultMatrix = new ItemStack[9];
         this.overflowItems = new ArrayList<>();
 
         for (int i = 0; i < this.resultMatrix.length; i++) {
-            this.resultMatrix[i] = ItemStack.empty();
+            this.resultMatrix[i] = new ItemStack(Material.AIR);
         }
     }
 
@@ -39,6 +_,6 @@
     }
 
     public void setResultMatrix(int i, ItemStack itemStack) {
-        this.resultMatrix[i] = itemStack;
+        this.resultMatrix[i] = Objects.requireNonNullElseGet(itemStack, () -> new ItemStack(Material.AIR));
     }
 }
