--- a/src/main/java/org/bukkit/craftbukkit/inventory/CraftItemType.java
+++ b/src/main/java/org/bukkit/craftbukkit/inventory/CraftItemType.java
@@ -160,11 +_,11 @@
     public int getBurnDuration() {
         FuelValues fuelValues = MinecraftServer.getServer().fuelValues();
         net.minecraft.world.item.ItemStack stack = new net.minecraft.world.item.ItemStack(this.getHandle());
-
+        
         if (!fuelValues.isFuel(stack)) {
             return 0;
         }
-
+        
         return fuelValues.burnDuration(stack);
     }
 
