--- a/src/main/java/org/bukkit/craftbukkit/inventory/CraftMetaArmorStand.java
+++ b/src/main/java/org/bukkit/craftbukkit/inventory/CraftMetaArmorStand.java
@@ -7,14 +_,14 @@
 import net.minecraft.core.component.DataComponents;
 import net.minecraft.nbt.CompoundTag;
 import net.minecraft.world.entity.EntityType;
-import net.minecraft.world.item.component.TypedEntityData;
+import net.minecraft.world.item.component.CustomData;
 import org.bukkit.Material;
 import org.bukkit.configuration.serialization.DelegateDeserialization;
 
 @DelegateDeserialization(SerializableMeta.class)
 public class CraftMetaArmorStand extends CraftMetaItem implements com.destroystokyo.paper.inventory.meta.ArmorStandMeta {
 
-    static final ItemMetaKeyType<TypedEntityData<EntityType<?>>> ENTITY_TAG = new ItemMetaKeyType<>(DataComponents.ENTITY_DATA, "entity-tag");
+    static final ItemMetaKeyType<CustomData> ENTITY_TAG = new ItemMetaKeyType<>(DataComponents.ENTITY_DATA, "entity-tag");
 
     static final ItemMetaKey ENTITY_ID = new ItemMetaKey("id", "entity-id");
     static final ItemMetaKey INVISIBLE = new ItemMetaKey("Invisible", "invisible");
@@ -40,7 +_,7 @@
         super(tag, extraHandledDcts);
 
         getOrEmpty(tag, CraftMetaArmorStand.ENTITY_TAG).ifPresent((nbt) -> {
-            this.entityTag = nbt.copyTagWithEntityId();
+            this.entityTag = nbt.copyTag();
         });
     }
 
@@ -98,7 +_,7 @@
         super.applyToItem(tag);
 
         if (this.entityTag != null) {
-            tag.put(CraftMetaArmorStand.ENTITY_TAG, TypedEntityData.decodeEntity(this.entityTag));
+            tag.put(CraftMetaArmorStand.ENTITY_TAG, CustomData.of(this.entityTag));
         }
     }
 
@@ -113,7 +_,7 @@
     }
 
     boolean isArmorStandEmpty() {
-        return this.entityTag == null || this.entityTag.size() == 1 && this.entityTag.contains("id"); // consider armor stand empty if tag is empty.
+        return this.entityTag == null || this.entityTag.isEmpty(); // consider armor stand empty if tag is empty.
     }
 
     @Override
