--- a/src/main/java/org/bukkit/craftbukkit/inventory/CraftMetaSkull.java
+++ b/src/main/java/org/bukkit/craftbukkit/inventory/CraftMetaSkull.java
@@ -4,7 +_,6 @@
 import com.mojang.authlib.GameProfile;
 import java.util.Map;
 import java.util.Objects;
-import com.mojang.datafixers.util.Either;
 import net.minecraft.Util;
 import net.minecraft.core.UUIDUtil;
 import net.minecraft.core.component.DataComponentPatch;
@@ -12,7 +_,6 @@
 import net.minecraft.nbt.CompoundTag;
 import net.minecraft.nbt.NbtOps;
 import net.minecraft.resources.ResourceLocation;
-import net.minecraft.world.entity.player.PlayerSkin;
 import net.minecraft.world.item.component.ResolvableProfile;
 import org.bukkit.Bukkit;
 import org.bukkit.NamespacedKey;
@@ -161,13 +_,12 @@
     @Override
     public OfflinePlayer getOwningPlayer() {
         if (this.hasOwner()) {
-            final GameProfile gameProfile = this.profile.partialProfile(); // The partial profile is always guaranteed to have a non-null uuid and name.
-            if (Objects.equals(gameProfile.id(), Util.NIL_UUID)) {
-                return Bukkit.getOfflinePlayer(gameProfile.id());
+            if (this.profile.id().filter(u -> !u.equals(Util.NIL_UUID)).isPresent()) {
+                return Bukkit.getOfflinePlayer(this.profile.id().get());
             }
 
-            if (!gameProfile.name().isEmpty()) {
-                return Bukkit.getOfflinePlayer(gameProfile.name());
+            if (this.profile.name().filter(s -> !s.isEmpty()).isPresent()) {
+                return Bukkit.getOfflinePlayer(this.profile.name().get());
             }
         }
 
@@ -183,13 +_,13 @@
         if (name == null) {
             this.setProfile(null);
         } else {
-            // Attempt to fetch an already resolved player profile in case the player is currently online.
+            // Paper start - Use Online Players Skull
+            GameProfile newProfile = null;
             net.minecraft.server.level.ServerPlayer player = net.minecraft.server.MinecraftServer.getServer().getPlayerList().getPlayerByName(name);
-            this.setProfile(
-                player != null
-                    ? ResolvableProfile.createResolved(player.getGameProfile())
-                    : new ResolvableProfile.Dynamic(Either.left(name), PlayerSkin.Patch.EMPTY)
-            );
+            if (player != null) newProfile = player.getGameProfile();
+            if (newProfile == null) newProfile = new GameProfile(Util.NIL_UUID, name);
+            this.setProfile(new ResolvableProfile(newProfile));
+            // Paper end
         }
 
         return true;
@@ -200,9 +_,9 @@
         if (owner == null) {
             this.setProfile(null);
         } else if (owner instanceof CraftPlayer craftPlayer) {
-            this.setProfile(ResolvableProfile.createResolved(craftPlayer.getProfile()));
+            this.setProfile(new ResolvableProfile(craftPlayer.getProfile()));
         } else {
-            this.setProfile(new ResolvableProfile.Dynamic(Either.right(owner.getUniqueId()), PlayerSkin.Patch.EMPTY));
+            this.setProfile(new ResolvableProfile(new GameProfile(owner.getUniqueId(), (owner.getName() == null) ? "" : owner.getName())));
         }
 
         return true;
@@ -222,7 +_,7 @@
     @Deprecated
     public void setOwnerProfile(PlayerProfile profile) {
         if (profile instanceof final com.destroystokyo.paper.profile.SharedPlayerProfile sharedProfile) {
-            this.setProfile(sharedProfile.buildResolvableProfile());
+            this.setProfile(CraftPlayerProfile.validateSkullProfile(sharedProfile.buildResolvableProfile()));
         } else {
             this.setProfile(null);
         }
