--- a/src/main/java/io/papermc/paper/threadedregions/scheduler/FoliaGlobalRegionScheduler.java
+++ b/src/main/java/io/papermc/paper/threadedregions/scheduler/FoliaGlobalRegionScheduler.java
@@ -1,13 +_,14 @@
 package io.papermc.paper.threadedregions.scheduler;
 
 import ca.spottedleaf.concurrentutil.util.ConcurrentUtil;
+import ca.spottedleaf.concurrentutil.util.Validate;
 import it.unimi.dsi.fastutil.longs.Long2ObjectOpenHashMap;
 import org.bukkit.plugin.IllegalPluginAccessException;
 import org.bukkit.plugin.Plugin;
+
 import java.lang.invoke.VarHandle;
 import java.util.ArrayList;
 import java.util.List;
-import java.util.Objects;
 import java.util.function.Consumer;
 import java.util.logging.Level;
 
@@ -39,8 +_,8 @@
 
     @Override
     public void execute(final Plugin plugin, final Runnable run) {
-        Objects.requireNonNull(plugin, "Plugin may not be null");
-        Objects.requireNonNull(run, "Runnable may not be null");
+        Validate.notNull(plugin, "Plugin may not be null");
+        Validate.notNull(run, "Runnable may not be null");
 
         this.run(plugin, (final ScheduledTask task) -> {
             run.run();
@@ -54,8 +_,8 @@
 
     @Override
     public ScheduledTask runDelayed(final Plugin plugin, final Consumer<ScheduledTask> task, final long delayTicks) {
-        Objects.requireNonNull(plugin, "Plugin may not be null");
-        Objects.requireNonNull(task, "Task may not be null");
+        Validate.notNull(plugin, "Plugin may not be null");
+        Validate.notNull(task, "Task may not be null");
         if (delayTicks <= 0) {
             throw new IllegalArgumentException("Delay ticks may not be <= 0");
         }
@@ -78,8 +_,8 @@
 
     @Override
     public ScheduledTask runAtFixedRate(final Plugin plugin, final Consumer<ScheduledTask> task, final long initialDelayTicks, final long periodTicks) {
-        Objects.requireNonNull(plugin, "Plugin may not be null");
-        Objects.requireNonNull(task, "Task may not be null");
+        Validate.notNull(plugin, "Plugin may not be null");
+        Validate.notNull(task, "Task may not be null");
         if (initialDelayTicks <= 0) {
             throw new IllegalArgumentException("Initial delay ticks may not be <= 0");
         }
@@ -105,7 +_,7 @@
 
     @Override
     public void cancelTasks(final Plugin plugin) {
-        Objects.requireNonNull(plugin, "Plugin may not be null");
+        Validate.notNull(plugin, "Plugin may not be null");
 
         final List<GlobalScheduledTask> toCancel = new ArrayList<>();
         synchronized (this.stateLock) {
