--- a/src/main/java/io/papermc/paper/threadedregions/scheduler/FoliaEntityScheduler.java
+++ b/src/main/java/io/papermc/paper/threadedregions/scheduler/FoliaEntityScheduler.java
@@ -1,13 +_,14 @@
 package io.papermc.paper.threadedregions.scheduler;
 
 import ca.spottedleaf.concurrentutil.util.ConcurrentUtil;
+import ca.spottedleaf.concurrentutil.util.Validate;
 import net.minecraft.world.entity.Entity;
 import org.bukkit.craftbukkit.entity.CraftEntity;
 import org.bukkit.plugin.IllegalPluginAccessException;
 import org.bukkit.plugin.Plugin;
 import org.jetbrains.annotations.Nullable;
+
 import java.lang.invoke.VarHandle;
-import java.util.Objects;
 import java.util.function.Consumer;
 import java.util.logging.Level;
 
@@ -20,8 +_,8 @@
     }
 
     private static Consumer<? extends Entity> wrap(final Plugin plugin, final Runnable runnable) {
-        Objects.requireNonNull(plugin, "Plugin may not be null");
-        Objects.requireNonNull(runnable, "Runnable may not be null");
+        Validate.notNull(plugin, "Plugin may not be null");
+        Validate.notNull(runnable, "Runnable may not be null");
 
         return (final Entity nmsEntity) -> {
             if (!plugin.isEnabled()) {
@@ -53,8 +_,8 @@
     @Override
     public @Nullable ScheduledTask runDelayed(final Plugin plugin, final Consumer<ScheduledTask> task, final Runnable retired,
                                               final long delayTicks) {
-        Objects.requireNonNull(plugin, "Plugin may not be null");
-        Objects.requireNonNull(task, "Task may not be null");
+        Validate.notNull(plugin, "Plugin may not be null");
+        Validate.notNull(task, "Task may not be null");
         if (delayTicks <= 0) {
             throw new IllegalArgumentException("Delay ticks may not be <= 0");
         }
@@ -80,8 +_,8 @@
     @Override
     public @Nullable ScheduledTask runAtFixedRate(final Plugin plugin, final Consumer<ScheduledTask> task,
                                                   final Runnable retired, final long initialDelayTicks, final long periodTicks) {
-        Objects.requireNonNull(plugin, "Plugin may not be null");
-        Objects.requireNonNull(task, "Task may not be null");
+        Validate.notNull(plugin, "Plugin may not be null");
+        Validate.notNull(task, "Task may not be null");
         if (initialDelayTicks <= 0) {
             throw new IllegalArgumentException("Initial delay ticks may not be <= 0");
         }
