--- a/src/main/java/io/papermc/paper/adventure/PaperAdventure.java
+++ b/src/main/java/io/papermc/paper/adventure/PaperAdventure.java
@@ -350,7 +_,7 @@
             Filterable.passThrough(validateField(asPlain(book.title(), locale), WrittenBookContent.TITLE_MAX_LENGTH, "title")),
             asPlain(book.author(), locale),
             0,
-            book.pages().stream().map(c -> Filterable.passThrough(PaperAdventure.asVanilla(c))).toList(), // TODO should we validate length?
+            book.pages().stream().map(c -> Filterable.passThrough(PaperAdventure.asVanilla(c))).toList(), // TODO should we validate legnth?
             false
         ));
         return item;
@@ -503,5 +_,4 @@
         return AdventureCodecs.STYLE_MAP_CODEC.codec()
             .parse(ops, encoded).getOrThrow(IllegalStateException::new);
     }
-
 }
