--- a/src/main/java/org/purpurmc/purpur/controller/LookControllerWASD.java
+++ b/src/main/java/org/purpurmc/purpur/controller/LookControllerWASD.java
@@ -50,7 +_,7 @@
             (byte) Mth.floor(entity.getXRot() * 256.0F / 360.0F),
             entity.onGround
         );
-        ((ServerLevel) entity.level()).getChunkSource().sendToTrackingPlayers(entity, entityPacket);
+        ((ServerLevel) entity.level()).getChunkSource().broadcast(entity, entityPacket);
     }
 
     public void setOffsets(float yaw, float pitch) {
