--- a/src/generated/java/org/bukkit/craftbukkit/block/impl/CraftTrialSpawner.java
+++ b/src/generated/java/org/bukkit/craftbukkit/block/impl/CraftTrialSpawner.java
@@ -1,7 +_,7 @@
 package org.bukkit.craftbukkit.block.impl;
 
 import com.google.common.base.Preconditions;
-import io.papermc.paper.annotation.GeneratedClass;
+import io.papermc.paper.generated.GeneratedFrom;
 import net.minecraft.world.level.block.TrialSpawnerBlock;
 import net.minecraft.world.level.block.entity.trialspawner.TrialSpawnerState;
 import net.minecraft.world.level.block.state.BlockState;
@@ -9,10 +_,8 @@
 import net.minecraft.world.level.block.state.properties.EnumProperty;
 import org.bukkit.block.data.type.TrialSpawner;
 import org.bukkit.craftbukkit.block.data.CraftBlockData;
-import org.jspecify.annotations.NullMarked;
 
-@NullMarked
-@GeneratedClass
+@GeneratedFrom("1.21.8")
 public class CraftTrialSpawner extends CraftBlockData implements TrialSpawner {
     private static final BooleanProperty OMINOUS = TrialSpawnerBlock.OMINOUS;
 
