--- a/src/generated/java/org/bukkit/craftbukkit/block/impl/CraftSnowLayer.java
+++ b/src/generated/java/org/bukkit/craftbukkit/block/impl/CraftSnowLayer.java
@@ -1,15 +_,13 @@
 package org.bukkit.craftbukkit.block.impl;
 
-import io.papermc.paper.annotation.GeneratedClass;
+import io.papermc.paper.generated.GeneratedFrom;
 import net.minecraft.world.level.block.SnowLayerBlock;
 import net.minecraft.world.level.block.state.BlockState;
 import net.minecraft.world.level.block.state.properties.IntegerProperty;
 import org.bukkit.block.data.type.Snow;
 import org.bukkit.craftbukkit.block.data.CraftBlockData;
-import org.jspecify.annotations.NullMarked;
 
-@NullMarked
-@GeneratedClass
+@GeneratedFrom("1.21.8")
 public class CraftSnowLayer extends CraftBlockData implements Snow {
     private static final IntegerProperty LAYERS = SnowLayerBlock.LAYERS;
 
