--- a/src/generated/java/org/bukkit/craftbukkit/block/impl/CraftTripWireHook.java
+++ b/src/generated/java/org/bukkit/craftbukkit/block/impl/CraftTripWireHook.java
@@ -1,7 +_,7 @@
 package org.bukkit.craftbukkit.block.impl;
 
 import com.google.common.base.Preconditions;
-import io.papermc.paper.annotation.GeneratedClass;
+import io.papermc.paper.generated.GeneratedFrom;
 import java.util.Set;
 import net.minecraft.core.Direction;
 import net.minecraft.world.level.block.TripWireHookBlock;
@@ -11,10 +_,8 @@
 import org.bukkit.block.BlockFace;
 import org.bukkit.block.data.type.TripwireHook;
 import org.bukkit.craftbukkit.block.data.CraftBlockData;
-import org.jspecify.annotations.NullMarked;
 
-@NullMarked
-@GeneratedClass
+@GeneratedFrom("1.21.8")
 public class CraftTripWireHook extends CraftBlockData implements TripwireHook {
     private static final BooleanProperty ATTACHED = TripWireHookBlock.ATTACHED;
 
