--- a/src/generated/java/org/bukkit/craftbukkit/block/impl/CraftPointedDripstone.java
+++ b/src/generated/java/org/bukkit/craftbukkit/block/impl/CraftPointedDripstone.java
@@ -1,7 +_,7 @@
 package org.bukkit.craftbukkit.block.impl;
 
 import com.google.common.base.Preconditions;
-import io.papermc.paper.annotation.GeneratedClass;
+import io.papermc.paper.generated.GeneratedFrom;
 import java.util.Set;
 import net.minecraft.core.Direction;
 import net.minecraft.world.level.block.PointedDripstoneBlock;
@@ -12,10 +_,8 @@
 import org.bukkit.block.BlockFace;
 import org.bukkit.block.data.type.PointedDripstone;
 import org.bukkit.craftbukkit.block.data.CraftBlockData;
-import org.jspecify.annotations.NullMarked;
 
-@NullMarked
-@GeneratedClass
+@GeneratedFrom("1.21.8")
 public class CraftPointedDripstone extends CraftBlockData implements PointedDripstone {
     private static final EnumProperty<DripstoneThickness> THICKNESS = PointedDripstoneBlock.THICKNESS;
 
