--- a/src/generated/java/org/bukkit/craftbukkit/block/impl/CraftCommandBlock.java
+++ b/src/generated/java/org/bukkit/craftbukkit/block/impl/CraftCommandBlock.java
@@ -1,7 +_,7 @@
 package org.bukkit.craftbukkit.block.impl;
 
 import com.google.common.base.Preconditions;
-import io.papermc.paper.annotation.GeneratedClass;
+import io.papermc.paper.generated.GeneratedFrom;
 import java.util.Set;
 import net.minecraft.core.Direction;
 import net.minecraft.world.level.block.state.BlockState;
@@ -10,10 +_,8 @@
 import org.bukkit.block.BlockFace;
 import org.bukkit.block.data.type.CommandBlock;
 import org.bukkit.craftbukkit.block.data.CraftBlockData;
-import org.jspecify.annotations.NullMarked;
 
-@NullMarked
-@GeneratedClass
+@GeneratedFrom("1.21.8")
 public class CraftCommandBlock extends CraftBlockData implements CommandBlock {
     private static final BooleanProperty CONDITIONAL = net.minecraft.world.level.block.CommandBlock.CONDITIONAL;
 
