--- a/src/generated/java/org/bukkit/craftbukkit/block/impl/CraftTallFlower.java
+++ b/src/generated/java/org/bukkit/craftbukkit/block/impl/CraftTallFlower.java
@@ -1,17 +_,15 @@
 package org.bukkit.craftbukkit.block.impl;
 
 import com.google.common.base.Preconditions;
-import io.papermc.paper.annotation.GeneratedClass;
+import io.papermc.paper.generated.GeneratedFrom;
 import net.minecraft.world.level.block.TallFlowerBlock;
 import net.minecraft.world.level.block.state.BlockState;
 import net.minecraft.world.level.block.state.properties.DoubleBlockHalf;
 import net.minecraft.world.level.block.state.properties.EnumProperty;
 import org.bukkit.block.data.Bisected;
 import org.bukkit.craftbukkit.block.data.CraftBlockData;
-import org.jspecify.annotations.NullMarked;
 
-@NullMarked
-@GeneratedClass
+@GeneratedFrom("1.21.8")
 public class CraftTallFlower extends CraftBlockData implements Bisected {
     private static final EnumProperty<DoubleBlockHalf> HALF = TallFlowerBlock.HALF;
 
