--- a/src/generated/java/org/bukkit/craftbukkit/block/impl/CraftDetectorRail.java
+++ b/src/generated/java/org/bukkit/craftbukkit/block/impl/CraftDetectorRail.java
@@ -1,7 +_,7 @@
 package org.bukkit.craftbukkit.block.impl;
 
 import com.google.common.base.Preconditions;
-import io.papermc.paper.annotation.GeneratedClass;
+import io.papermc.paper.generated.GeneratedFrom;
 import java.util.Set;
 import net.minecraft.world.level.block.DetectorRailBlock;
 import net.minecraft.world.level.block.state.BlockState;
@@ -10,10 +_,8 @@
 import net.minecraft.world.level.block.state.properties.RailShape;
 import org.bukkit.block.data.type.RedstoneRail;
 import org.bukkit.craftbukkit.block.data.CraftBlockData;
-import org.jspecify.annotations.NullMarked;
 
-@NullMarked
-@GeneratedClass
+@GeneratedFrom("1.21.8")
 public class CraftDetectorRail extends CraftBlockData implements RedstoneRail {
     private static final BooleanProperty POWERED = DetectorRailBlock.POWERED;
 
