--- a/src/generated/java/org/bukkit/craftbukkit/block/impl/CraftHugeMushroom.java
+++ b/src/generated/java/org/bukkit/craftbukkit/block/impl/CraftHugeMushroom.java
@@ -2,7 +_,7 @@
 
 import com.google.common.base.Preconditions;
 import com.google.common.collect.ImmutableSet;
-import io.papermc.paper.annotation.GeneratedClass;
+import io.papermc.paper.generated.GeneratedFrom;
 import java.util.Collections;
 import java.util.Map;
 import java.util.Set;
@@ -13,10 +_,8 @@
 import org.bukkit.block.BlockFace;
 import org.bukkit.block.data.MultipleFacing;
 import org.bukkit.craftbukkit.block.data.CraftBlockData;
-import org.jspecify.annotations.NullMarked;
 
-@NullMarked
-@GeneratedClass
+@GeneratedFrom("1.21.8")
 public class CraftHugeMushroom extends CraftBlockData implements MultipleFacing {
     private static final Map<BlockFace, BooleanProperty> PROPERTY_BY_DIRECTION = Map.of(
         BlockFace.DOWN, HugeMushroomBlock.DOWN,
