--- a/src/generated/java/org/bukkit/craftbukkit/block/impl/CraftRedStoneWire.java
+++ b/src/generated/java/org/bukkit/craftbukkit/block/impl/CraftRedStoneWire.java
@@ -1,7 +_,7 @@
 package org.bukkit.craftbukkit.block.impl;
 
 import com.google.common.base.Preconditions;
-import io.papermc.paper.annotation.GeneratedClass;
+import io.papermc.paper.generated.GeneratedFrom;
 import java.util.Collections;
 import java.util.Map;
 import java.util.Set;
@@ -15,10 +_,8 @@
 import org.bukkit.block.data.type.RedstoneWire;
 import org.bukkit.craftbukkit.block.CraftBlock;
 import org.bukkit.craftbukkit.block.data.CraftBlockData;
-import org.jspecify.annotations.NullMarked;
 
-@NullMarked
-@GeneratedClass
+@GeneratedFrom("1.21.8")
 public class CraftRedStoneWire extends CraftBlockData implements RedstoneWire {
     private static final IntegerProperty POWER = RedStoneWireBlock.POWER;
 
