--- a/src/generated/java/org/bukkit/craftbukkit/block/impl/CraftBrewingStand.java
+++ b/src/generated/java/org/bukkit/craftbukkit/block/impl/CraftBrewingStand.java
@@ -1,17 +_,15 @@
 package org.bukkit.craftbukkit.block.impl;
 
 import com.google.common.collect.ImmutableSet;
-import io.papermc.paper.annotation.GeneratedClass;
+import io.papermc.paper.generated.GeneratedFrom;
 import java.util.Set;
 import net.minecraft.world.level.block.BrewingStandBlock;
 import net.minecraft.world.level.block.state.BlockState;
 import net.minecraft.world.level.block.state.properties.BooleanProperty;
 import org.bukkit.block.data.type.BrewingStand;
 import org.bukkit.craftbukkit.block.data.CraftBlockData;
-import org.jspecify.annotations.NullMarked;
 
-@NullMarked
-@GeneratedClass
+@GeneratedFrom("1.21.8")
 public class CraftBrewingStand extends CraftBlockData implements BrewingStand {
     private static final BooleanProperty[] HAS_BOTTLE = BrewingStandBlock.HAS_BOTTLE;
 
