--- a/src/main/java/com/destroystokyo/paper/network/PaperNetworkClient.java
+++ b/src/main/java/com/destroystokyo/paper/network/PaperNetworkClient.java
@@ -39,7 +_,7 @@
         }
 
         // When clients connect with a SRV record, their host contains a trailing '.'
-        if (len > 0 && host.charAt(len - 1) == '.') {
+        if (len > 0 && host.charAt(len -  1) == '.') {
             len--;
         }
 
