--- a/src/main/java/io/papermc/paper/configuration/serializer/ServerboundPacketClassSerializer.java
+++ b/src/main/java/io/papermc/paper/configuration/serializer/ServerboundPacketClassSerializer.java
@@ -73,7 +_,7 @@
             }
             ID_TO_INFO = idBuilder.buildOrThrow();
             CLASS_TO_INFO = classBuilder.buildOrThrow();
-            Preconditions.checkState(ID_TO_INFO.size() == 75, "Packet class map must have 75 entries");
+            Preconditions.checkState(ID_TO_INFO.size() == 74, "Packet class map must have 74 entries");
         } catch (final ReflectiveOperationException e) {
             throw new RuntimeException("Could not create packet class map", e);
         }
