--- a/src/main/java/io/papermc/paper/datacomponent/item/PaperPotionContents.java
+++ b/src/main/java/io/papermc/paper/datacomponent/item/PaperPotionContents.java
@@ -52,7 +_,7 @@
 
     @Override
     public @Unmodifiable List<PotionEffect> allEffects() {
-        //noinspection SimplifyStreamApiCallChains - explicitly want it unmodifiable, as toList() api doesnt guarantee this.
+        //noinspection SimplifyStreamApiCallChains - explicity want it unmodifiable, as toList() api doesnt guarantee this.
         return StreamSupport.stream(this.impl.getAllEffects().spliterator(), false)
             .map(CraftPotionUtil::toBukkit)
             .collect(Collectors.toUnmodifiableList());
