--- a/src/main/java/io/papermc/paper/registry/data/PaperFrogVariantRegistryEntry.java
+++ b/src/main/java/io/papermc/paper/registry/data/PaperFrogVariantRegistryEntry.java
@@ -3,7 +_,6 @@
 import io.papermc.paper.registry.PaperRegistryBuilder;
 import io.papermc.paper.registry.data.client.ClientTextureAsset;
 import io.papermc.paper.registry.data.util.Conversions;
-import net.minecraft.core.ClientAsset;
 import net.minecraft.world.entity.animal.frog.FrogVariant;
 import net.minecraft.world.entity.variant.SpawnPrioritySelectors;
 import org.bukkit.entity.Frog;
@@ -14,7 +_,7 @@
 
 public class PaperFrogVariantRegistryEntry implements FrogVariantRegistryEntry {
 
-    protected ClientAsset.@Nullable ResourceTexture clientTextureAsset;
+    protected net.minecraft.core.@Nullable ClientAsset clientTextureAsset;
     protected SpawnPrioritySelectors spawnConditions;
 
     protected final Conversions conversions;
