--- a/src/main/java/io/papermc/paper/registry/data/PaperPigVariantRegistryEntry.java
+++ b/src/main/java/io/papermc/paper/registry/data/PaperPigVariantRegistryEntry.java
@@ -3,7 +_,6 @@
 import io.papermc.paper.registry.PaperRegistryBuilder;
 import io.papermc.paper.registry.data.client.ClientTextureAsset;
 import io.papermc.paper.registry.data.util.Conversions;
-import net.minecraft.core.ClientAsset;
 import net.minecraft.world.entity.animal.PigVariant;
 import net.minecraft.world.entity.variant.ModelAndTexture;
 import net.minecraft.world.entity.variant.SpawnPrioritySelectors;
@@ -16,7 +_,7 @@
 public class PaperPigVariantRegistryEntry implements PigVariantRegistryEntry {
 
     protected PigVariant.@Nullable ModelType model;
-    protected ClientAsset.@Nullable ResourceTexture clientTextureAsset;
+    protected net.minecraft.core.@Nullable ClientAsset clientTextureAsset;
     protected SpawnPrioritySelectors spawnConditions;
 
     protected final Conversions conversions;
