--- a/src/main/java/io/papermc/paper/connection/HorriblePlayerLoginEventHack.java
+++ b/src/main/java/io/papermc/paper/connection/HorriblePlayerLoginEventHack.java
@@ -13,6 +_,7 @@
 import net.minecraft.server.level.ClientInformation;
 import net.minecraft.server.level.ServerPlayer;
 import net.minecraft.server.players.PlayerList;
+import net.minecraft.world.level.Level;
 import org.bukkit.craftbukkit.entity.CraftPlayer;
 import org.bukkit.event.player.PlayerLoginEvent;
 import org.bukkit.plugin.RegisteredListener;
@@ -28,10 +_,7 @@
     private static final Logger LOGGER = LogUtils.getClassLogger();
     private static boolean nagged = false;
 
-    public static boolean warnReenterConfiguration(final Connection connection) {
-        if (!connection.handledLegacyLoginEvent) {
-            return false;
-        }
+    public static void warnReenterConfiguration() {
         LOGGER.warn("""
                 
                 ============================================================
@@ -42,7 +_,6 @@
                 
                 Please look in your logs for the Plugins listening to this event.
                 ============================================================""", StackWalkerUtil.getFirstPluginCaller().getName());
-        return true;
     }
 
     public static @Nullable Component execute(final Connection connection, MinecraftServer server, GameProfile profile, PlayerList.LoginResult result) {
@@ -74,14 +_,13 @@
         // We need to account for the fact that during the config stage we now call this event to mimic the old behavior.
         // However, we want to make sure that we still properly hold the same player reference
         ServerPlayer player;
-        if (connection.savedPlayerForLegacyEvents != null) {
-            player = connection.savedPlayerForLegacyEvents;
+        if (connection.savedPlayerForLoginEventLegacy != null) {
+            player = connection.savedPlayerForLoginEventLegacy;
         } else {
-            ServerPlayer serverPlayer = new ServerPlayer(server, server.overworld(), profile, ClientInformation.createDefault());
-            connection.savedPlayerForLegacyEvents = serverPlayer;
+            ServerPlayer serverPlayer = new ServerPlayer(server, server.getLevel(Level.OVERWORLD), profile, ClientInformation.createDefault());
+            connection.savedPlayerForLoginEventLegacy = serverPlayer;
             player = serverPlayer;
         }
-        connection.handledLegacyLoginEvent = true;
 
         CraftPlayer horribleBukkitPlayer = player.getBukkitEntity();
         PlayerLoginEvent event = new PlayerLoginEvent(horribleBukkitPlayer, connection.hostname, ((java.net.InetSocketAddress) connection.getRemoteAddress()).getAddress(), ((java.net.InetSocketAddress) connection.channel.remoteAddress()).getAddress());
