--- a/src/main/java/io/papermc/paper/connection/PaperPlayerGameConnection.java
+++ b/src/main/java/io/papermc/paper/connection/PaperPlayerGameConnection.java
@@ -19,7 +_,8 @@
 
     @Override
     public void reenterConfiguration() {
-        if (HorriblePlayerLoginEventHack.warnReenterConfiguration(this.handle.connection)) {
+        if (this.handle.connection.savedPlayerForLoginEventLegacy != null) {
+            HorriblePlayerLoginEventHack.warnReenterConfiguration();
             return;
         }
         this.handle.switchToConfig();
