--- a/purpur-server/build.gradle.kts
+++ b/purpur-server/build.gradle.kts
@@ -32,8 +_,31 @@
             outputDir = rootDirectory.dir("paper-server")
         }
     }
-    activeFork = purpur
+    //activeFork = purpur // Pulvinar
     // Purpur end - Rebrand
+    // Pulvinar start - Rebrand
+    val pulvinar = forks.register("pulvinar") {
+        forks = purpur
+        upstream.patchRepo("paperServer") {
+            upstreamRepo = purpur.patchedRepo("paperServer")
+            patchesDir = rootDirectory.dir("pulvinar-server/paper-patches")
+            outputDir = rootDirectory.dir("paper-server")
+        }
+        upstream.patchDir("purpurServer") {
+            upstreamPath = "purpur-server"
+            excludes = setOf(
+                "src/minecraft",
+                "paper-patches",
+                "minecraft-patches",
+                "build.gradle.kts",
+                "build.gradle.kts.patch"
+            )
+            patchesDir = rootDirectory.dir("pulvinar-server/purpur-patches")
+            outputDir = rootDirectory.dir("purpur-server")
+        }
+    }
+    activeFork = pulvinar
+    // Pulvinar end - Rebrand
 
     spigot {
         enabled = false
@@ -122,10 +_,14 @@
     main {
         java { srcDir("../paper-server/src/main/java") }
         resources { srcDir("../paper-server/src/main/resources") }
+        java { srcDir("../purpur-server/src/main/java") } // Pulvinar
+        resources { srcDir("../purpur-server/src/main/resources") } // Pulvinar
     }
     test {
         java { srcDir("../paper-server/src/test/java") }
         resources { srcDir("../paper-server/src/test/resources") }
+        java { srcDir("../purpur-server/src/test/java") } // Pulvinar
+        resources { srcDir("../purpur-server/src/test/resources") } // Pulvinar
     }
 }
 val log4jPlugins = sourceSets.create("log4jPlugins") {
@@ -154,7 +_,9 @@
 }
 
 dependencies {
-    implementation(project(":purpur-api")) // Purpur
+    //implementation(project(":purpur-api")) // Purpur
+    implementation(project(":pulvinar-api")) // Pulvinar
+    implementation("net.kissenpvp:kissen-source:1.0.0-SNAPSHOT") // Pulvinar
     implementation("ca.spottedleaf:concurrentutil:0.0.7")
     implementation("org.jline:jline-terminal-ffm:3.27.1") // use ffm on java 22+
     implementation("org.jline:jline-terminal-jni:3.27.1") // fall back to jni on java 21
@@ -229,14 +_,14 @@
         val gitBranch = git.exec(providers, "rev-parse", "--abbrev-ref", "HEAD").get().trim()
         attributes(
             "Main-Class" to "org.bukkit.craftbukkit.Main",
-            "Implementation-Title" to "Purpur", // Purpur
+            "Implementation-Title" to "Pulvinar", // Purpur // Pulvinar
             "Implementation-Version" to implementationVersion,
             "Implementation-Vendor" to date,
-            "Specification-Title" to "Purpur", // Purpur
+            "Specification-Title" to "Pulvinar", // Purpur // Pulvinar
             "Specification-Version" to project.version,
-            "Specification-Vendor" to "Purpur Team", // Purpur
-            "Brand-Id" to "purpurmc:purpur", // Purpur
-            "Brand-Name" to "Purpur", // Purpur
+            "Specification-Vendor" to "KissenPvP Team", // Purpur // Pulvinar
+            "Brand-Id" to "kissenpvp:pulvinar", // Purpur // Pulvinar
+            "Brand-Name" to "Pulvinar", // Purpur // Pulvinar
             "Build-Number" to (build ?: ""),
             "Build-Time" to buildTime.toString(),
             "Git-Branch" to gitBranch,
