--- a/src/main/java/org/bukkit/craftbukkit/entity/CraftMinecartTNT.java
+++ b/src/main/java/org/bukkit/craftbukkit/entity/CraftMinecartTNT.java
@@ -69,13 +_,13 @@
 
     @Override
     public void explode() {
-        this.getHandle().explode(null, this.getHandle().getDeltaMovement().horizontalDistanceSqr());
+        this.getHandle().explode(this.getHandle().getDeltaMovement().horizontalDistanceSqr());
     }
 
     @Override
     public void explode(double power) {
         Preconditions.checkArgument(0 <= power && power <= Mth.square(5), "Power must be in range [0, 25] (got %s)", power);
 
-        this.getHandle().explode(null, power);
+        this.getHandle().explode(power);
     }
 }
