--- a/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java
+++ b/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java
@@ -7,7 +_,6 @@
 import com.google.common.io.BaseEncoding;
 import com.mojang.authlib.GameProfile;
 import com.mojang.datafixers.util.Pair;
-import com.mojang.logging.LogUtils;
 import io.papermc.paper.FeatureHooks;
 import io.papermc.paper.connection.PlayerGameConnection;
 import io.papermc.paper.connection.PluginMessageBridgeImpl;
@@ -54,10 +_,8 @@
 import net.minecraft.advancements.AdvancementProgress;
 import net.minecraft.commands.Commands;
 import net.minecraft.core.BlockPos;
-import net.minecraft.core.GlobalPos;
 import net.minecraft.core.Holder;
 import net.minecraft.core.SectionPos;
-import net.minecraft.nbt.CompoundTag;
 import net.minecraft.network.chat.Component;
 import net.minecraft.network.chat.PlayerChatMessage;
 import net.minecraft.network.protocol.common.ClientboundClearDialogPacket;
@@ -109,7 +_,6 @@
 import net.minecraft.sounds.SoundEvent;
 import net.minecraft.util.ProblemReporter;
 import net.minecraft.world.entity.Entity;
-import net.minecraft.world.entity.EntitySpawnReason;
 import net.minecraft.world.entity.ai.attributes.AttributeInstance;
 import net.minecraft.world.entity.ai.attributes.Attributes;
 import net.minecraft.world.entity.item.ItemEntity;
@@ -123,8 +_,6 @@
 import net.minecraft.world.level.saveddata.maps.MapDecoration;
 import net.minecraft.world.level.saveddata.maps.MapId;
 import net.minecraft.world.level.saveddata.maps.MapItemSavedData;
-import net.minecraft.world.level.storage.LevelData;
-import net.minecraft.world.level.storage.TagValueInput;
 import net.minecraft.world.level.storage.ValueInput;
 import net.minecraft.world.level.storage.ValueOutput;
 import org.bukkit.BanEntry;
@@ -215,7 +_,6 @@
 
 @DelegateDeserialization(CraftOfflinePlayer.class)
 public class CraftPlayer extends CraftHumanEntity implements Player, PluginMessageBridgeImpl {
-    private static final org.slf4j.Logger LOGGER = LogUtils.getClassLogger();
     private static final PointersSupplier<Player> POINTERS_SUPPLIER = PointersSupplier.<Player>builder()
         .parent(CraftEntity.POINTERS_SUPPLIER)
         .resolving(Identity.NAME, Player::getName)
@@ -281,7 +_,7 @@
 
     @Override
     public boolean isOp() {
-        return this.server.getHandle().isOp(this.getHandle().nameAndId());
+        return this.server.getHandle().isOp(this.getProfile());
     }
 
     @Override
@@ -289,9 +_,9 @@
         if (value == this.isOp()) return;
 
         if (value) {
-            this.server.getHandle().op(this.getHandle().nameAndId());
+            this.server.getHandle().op(this.getProfile());
         } else {
-            this.server.getHandle().deop(this.getHandle().nameAndId());
+            this.server.getHandle().deop(this.getProfile());
         }
 
         this.perm.recalculatePermissions();
@@ -582,7 +_,7 @@
         }
         this.getHandle().listName = name.equals(this.getName()) ? null : useMM ? io.papermc.paper.adventure.PaperAdventure.asVanilla(net.kyori.adventure.text.minimessage.MiniMessage.miniMessage().deserialize(name)) : CraftChatMessage.fromStringOrNull(name); // Purpur - AFK API
         if (this.getHandle().connection == null) return; // Paper - Updates are possible before the player has fully joined
-        for (ServerPlayer player : this.server.getHandle().players) {
+        for (ServerPlayer player : (List<ServerPlayer>) this.server.getHandle().players) {
             if (player.getBukkitEntity().canSee(this)) {
                 player.connection.send(new ClientboundPlayerInfoUpdatePacket(ClientboundPlayerInfoUpdatePacket.Action.UPDATE_DISPLAY_NAME, this.getHandle()));
             }
@@ -654,6 +_,13 @@
         this.getHandle().connection.disconnect(CraftChatMessage.fromStringOrEmpty(message, true), org.bukkit.event.player.PlayerKickEvent.Cause.PLUGIN); // Paper - kick event cause
     }
 
+    private static final net.kyori.adventure.text.Component DEFAULT_KICK_COMPONENT = net.kyori.adventure.text.Component.translatable("multiplayer.disconnect.kicked");
+
+    @Override
+    public void kick() {
+        this.kick(DEFAULT_KICK_COMPONENT);
+    }
+
     @Override
     public void kick(net.kyori.adventure.text.Component message, org.bukkit.event.player.PlayerKickEvent.Cause cause) {
         org.spigotmc.AsyncCatcher.catchOp("player kick");
@@ -691,7 +_,7 @@
     public void sendOpLevel(byte level) {
         Preconditions.checkArgument(level >= Commands.LEVEL_ALL && level <= Commands.LEVEL_OWNERS, "Level must be within [%s, %s]", Commands.LEVEL_ALL, Commands.LEVEL_OWNERS);
 
-        this.server.getServer().getPlayerList().sendPlayerPermissionLevel(this.getHandle(), level, false);
+        this.getHandle().getServer().getPlayerList().sendPlayerPermissionLevel(this.getHandle(), level, false);
     }
 
     @Override
@@ -717,16 +_,7 @@
         if (this.getHandle().connection == null) return;
 
         // Do not directly assign here, from the packethandler we'll assign it.
-        this.getHandle().connection.send(new ClientboundSetDefaultSpawnPositionPacket(
-            new LevelData.RespawnData(
-                GlobalPos.of(
-                    ((CraftWorld) loc.getWorld()).getHandle().dimension(),
-                    CraftLocation.toBlockPosition(loc)
-                ),
-                loc.getYaw(),
-                loc.getPitch()
-            )
-        ));
+        this.getHandle().connection.send(new ClientboundSetDefaultSpawnPositionPacket(CraftLocation.toBlockPosition(loc), loc.getYaw()));
     }
 
     @Override
@@ -763,6 +_,11 @@
     }
 
     @Override
+    public void playNote(Location loc, byte instrument, byte note) {
+        this.playNote(loc, Instrument.getByType(instrument), new Note(note));
+    }
+
+    @Override
     public void playNote(Location loc, Instrument instrument, Note note) {
         Preconditions.checkArgument(loc != null, "Location cannot be null");
         Preconditions.checkArgument(instrument != null, "Instrument cannot be null");
@@ -781,6 +_,16 @@
     }
 
     @Override
+    public void playSound(Location loc, Sound sound, float volume, float pitch) {
+        this.playSound(loc, sound, org.bukkit.SoundCategory.MASTER, volume, pitch);
+    }
+
+    @Override
+    public void playSound(Location loc, String sound, float volume, float pitch) {
+        this.playSound(loc, sound, org.bukkit.SoundCategory.MASTER, volume, pitch);
+    }
+
+    @Override
     public void playSound(Location loc, Sound sound, org.bukkit.SoundCategory category, float volume, float pitch) {
         this.playSound(loc, sound, category, volume, pitch, this.getHandle().random.nextLong());
     }
@@ -814,6 +_,16 @@
     }
 
     @Override
+    public void playSound(org.bukkit.entity.Entity entity, Sound sound, float volume, float pitch) {
+        this.playSound(entity, sound, org.bukkit.SoundCategory.MASTER, volume, pitch);
+    }
+
+    @Override
+    public void playSound(org.bukkit.entity.Entity entity, String sound, float volume, float pitch) {
+        this.playSound(entity, sound, org.bukkit.SoundCategory.MASTER, volume, pitch);
+    }
+
+    @Override
     public void playSound(org.bukkit.entity.Entity entity, Sound sound, org.bukkit.SoundCategory category, float volume, float pitch) {
         this.playSound(entity, sound, category, volume, pitch, this.getHandle().random.nextLong());
     }
@@ -850,6 +_,21 @@
     }
 
     @Override
+    public void stopSound(Sound sound) {
+        this.stopSound(sound, null);
+    }
+
+    @Override
+    public void stopSound(String sound) {
+        this.stopSound(sound, null);
+    }
+
+    @Override
+    public void stopSound(Sound sound, org.bukkit.SoundCategory category) {
+        this.stopSound(sound.getKey().getKey(), category);
+    }
+
+    @Override
     public void stopSound(String sound, org.bukkit.SoundCategory category) {
         if (this.getHandle().connection == null) return;
 
@@ -1026,8 +_,7 @@
     @Override
     public void sendBlockHighlight(Location location, int duration, String text, int argb) {
         if (this.getHandle().connection == null) return;
-        // Purpur - TODO: fix block highlight API
-        //this.getHandle().connection.send(new net.minecraft.network.protocol.common.ClientboundCustomPayloadPacket(new net.minecraft.network.protocol.common.custom.GameTestAddMarkerDebugPayload(org.bukkit.craftbukkit.util.CraftLocation.toBlockPosition(location), argb, text, duration)));
+        this.getHandle().connection.send(new net.minecraft.network.protocol.common.ClientboundCustomPayloadPacket(new net.minecraft.network.protocol.common.custom.GameTestAddMarkerDebugPayload(org.bukkit.craftbukkit.util.CraftLocation.toBlockPosition(location), argb, text, duration)));
     }
 
     @Override
@@ -1044,8 +_,7 @@
     @Override
     public void clearBlockHighlights() {
         if (this.getHandle().connection == null) return;
-        // Purpur - TODO: fix block highlight API
-        //this.getHandle().connection.send(new net.minecraft.network.protocol.common.ClientboundCustomPayloadPacket(new net.minecraft.network.protocol.common.custom.GameTestClearMarkersDebugPayload()));
+        this.getHandle().connection.send(new net.minecraft.network.protocol.common.ClientboundCustomPayloadPacket(new net.minecraft.network.protocol.common.custom.GameTestClearMarkersDebugPayload()));
     }
     // Purpur end - Debug Marker API
 
@@ -1058,6 +_,11 @@
     // Purpur end - Death screen API
 
     @Override
+    public void sendBlockDamage(Location loc, float progress) {
+        this.sendBlockDamage(loc, progress, this.getEntityId());
+    }
+
+    @Override
     public void sendBlockDamage(Location loc, float progress, org.bukkit.entity.Entity source) {
         Preconditions.checkArgument(source != null, "source must not be null");
         this.sendBlockDamage(loc, progress, source.getEntityId());
@@ -1240,35 +_,35 @@
     private BorderChangeListener createWorldBorderListener() {
         return new BorderChangeListener() {
             @Override
-            public void onSetSize(net.minecraft.world.level.border.WorldBorder border, double size) {
+            public void onBorderSizeSet(net.minecraft.world.level.border.WorldBorder border, double size) {
                 CraftPlayer.this.getHandle().connection.send(new ClientboundSetBorderSizePacket(border));
             }
 
             @Override
-            public void onLerpSize(net.minecraft.world.level.border.WorldBorder border, double fromSize, double toSize, long time) {
+            public void onBorderSizeLerping(net.minecraft.world.level.border.WorldBorder border, double fromSize, double toSize, long time) {
                 CraftPlayer.this.getHandle().connection.send(new ClientboundSetBorderLerpSizePacket(border));
             }
 
             @Override
-            public void onSetCenter(net.minecraft.world.level.border.WorldBorder border, double centerX, double centerZ) {
+            public void onBorderCenterSet(net.minecraft.world.level.border.WorldBorder border, double centerX, double centerZ) {
                 CraftPlayer.this.getHandle().connection.send(new ClientboundSetBorderCenterPacket(border));
             }
 
             @Override
-            public void onSetWarningTime(net.minecraft.world.level.border.WorldBorder border, int warningTime) {
+            public void onBorderSetWarningTime(net.minecraft.world.level.border.WorldBorder border, int warningTime) {
                 CraftPlayer.this.getHandle().connection.send(new ClientboundSetBorderWarningDelayPacket(border));
             }
 
             @Override
-            public void onSetWarningBlocks(net.minecraft.world.level.border.WorldBorder border, int warningBlockDistance) {
+            public void onBorderSetWarningBlocks(net.minecraft.world.level.border.WorldBorder border, int warningBlockDistance) {
                 CraftPlayer.this.getHandle().connection.send(new ClientboundSetBorderWarningDistancePacket(border));
             }
 
             @Override
-            public void onSetDamagePerBlock(net.minecraft.world.level.border.WorldBorder border, double damagePerBlock) {} // NO OP
+            public void onBorderSetDamagePerBlock(net.minecraft.world.level.border.WorldBorder border, double damagePerBlock) {} // NO OP
 
             @Override
-            public void onSetSafeZone(net.minecraft.world.level.border.WorldBorder border, double safeZoneRadius) {} // NO OP
+            public void onBorderSetDamageSafeZOne(net.minecraft.world.level.border.WorldBorder border, double safeZoneRadius) {} // NO OP
         };
     }
 
@@ -1367,7 +_,7 @@
     public void setRotation(float yaw, float pitch) {
         // Paper start - Teleport API
         if (this.getHandle().connection == null) return;
-        this.getHandle().forceSetRotation(yaw, false, pitch, false);
+        this.getHandle().forceSetRotation(yaw, pitch);
         // Paper end - Teleportation API
     }
 
@@ -1527,9 +_,7 @@
 
     @Override
     public void loadData() {
-        this.server.getHandle().playerIo.load(this.getHandle().nameAndId())
-            .map(tag -> TagValueInput.create(ProblemReporter.DISCARDING, this.server.getServer().registryAccess(), tag))
-            .ifPresent(this.getHandle()::load);
+        this.server.getHandle().playerIo.load(this.getHandle(), ProblemReporter.DISCARDING);
     }
 
     @Override
@@ -1558,35 +_,44 @@
     public Location getRespawnLocation(final boolean loadLocationAndValidate) {
         final ServerPlayer.RespawnConfig respawnConfig = this.getHandle().getRespawnConfig();
         if (respawnConfig == null) return null;
-        final LevelData.RespawnData respawnData = respawnConfig.respawnData();
 
-        final ServerLevel world = this.server.getServer().getLevel(respawnData.dimension());
+        final ServerLevel world = this.getHandle().getServer().getLevel(respawnConfig.dimension());
         if (world == null) return null;
 
         if (!loadLocationAndValidate) {
-            return CraftLocation.toBukkit(respawnData.pos(), world, respawnData.yaw(), respawnData.pitch());
+            return CraftLocation.toBukkit(respawnConfig.pos(), world.getWorld(), respawnConfig.angle(), 0);
         }
 
         return ServerPlayer.findRespawnAndUseSpawnBlock(world, respawnConfig, false)
-            .map(pos -> CraftLocation.toBukkit(pos.position(), world, pos.yaw(), pos.pitch()))
+            .map(pos -> CraftLocation.toBukkit(pos.position(), world.getWorld(), pos.yaw(), 0))
             .orElse(null);
     }
 
     @Override
+    public void setBedSpawnLocation(Location location) {
+        this.setBedSpawnLocation(location, false);
+    }
+
+    @Override
+    public void setRespawnLocation(Location location) {
+        this.setRespawnLocation(location, false);
+    }
+
+    @Override
+    public void setBedSpawnLocation(Location location, boolean override) {
+        this.setRespawnLocation(location, override);
+    }
+
+    @Override
     public void setRespawnLocation(Location location, boolean override) {
         if (location == null) {
             this.getHandle().setRespawnPosition(null, false, PlayerSetSpawnEvent.Cause.PLUGIN);
         } else {
             this.getHandle().setRespawnPosition(
                 new ServerPlayer.RespawnConfig(
-                    new LevelData.RespawnData(
-                        GlobalPos.of(
-                            ((CraftWorld) location.getWorld()).getHandle().dimension(),
-                            CraftLocation.toBlockPosition(location)
-                        ),
-                        location.getYaw(),
-                        location.getPitch()
-                    ),
+                    ((CraftWorld) location.getWorld()).getHandle().dimension(),
+                    CraftLocation.toBlockPosition(location),
+                    location.getYaw(),
                     override
                 ),
                 false,
@@ -1609,7 +_,7 @@
     public Location getBedLocation() {
         Preconditions.checkState(this.isSleeping(), "Not sleeping");
 
-        BlockPos bed = this.getHandle().getRespawnConfig().respawnData().pos();
+        BlockPos bed = this.getHandle().getRespawnConfig().pos();
         return CraftLocation.toBukkit(bed, this.getWorld());
     }
 
@@ -1828,15 +_,15 @@
 
     @Override
     public boolean isWhitelisted() {
-        return this.server.getHandle().getWhiteList().isWhiteListed(this.getHandle().nameAndId());
+        return this.server.getHandle().getWhiteList().isWhiteListed(this.getProfile());
     }
 
     @Override
     public void setWhitelisted(boolean value) {
         if (value) {
-            this.server.getHandle().getWhiteList().add(new UserWhiteListEntry(this.getHandle().nameAndId()));
+            this.server.getHandle().getWhiteList().add(new UserWhiteListEntry(this.getProfile()));
         } else {
-            this.server.getHandle().getWhiteList().remove(this.getHandle().nameAndId());
+            this.server.getHandle().getWhiteList().remove(this.getProfile());
         }
     }
 
@@ -2010,6 +_,11 @@
     }
 
     @Override
+    public void hidePlayer(Plugin plugin, Player player) {
+        this.hideEntity(plugin, player);
+    }
+
+    @Override
     public void hideEntity(Plugin plugin, org.bukkit.entity.Entity entity) {
         Preconditions.checkArgument(plugin != null, "Plugin cannot be null");
         Preconditions.checkArgument(plugin.isEnabled(), "Plugin (%s) cannot be disabled", plugin.getName());
@@ -2092,6 +_,11 @@
     }
 
     @Override
+    public void showPlayer(Plugin plugin, Player player) {
+        this.showEntity(plugin, player);
+    }
+
+    @Override
     public void showEntity(Plugin plugin, org.bukkit.entity.Entity entity) {
         Preconditions.checkArgument(plugin != null, "Plugin cannot be null");
         // Don't require that plugin be enabled. A plugin must be allowed to call
@@ -2131,20 +_,23 @@
     }
 
     private void trackAndShowEntity(org.bukkit.entity.Entity entity) {
+        // Paper start - uuid override
         this.trackAndShowEntity(entity, null);
     }
-
     private void trackAndShowEntity(org.bukkit.entity.Entity entity, final @Nullable UUID uuidOverride) {
-        ChunkMap tracker = this.getHandle().level().getChunkSource().chunkMap;
+        // Paper end
+        ChunkMap tracker = ((ServerLevel) this.getHandle().level()).getChunkSource().chunkMap;
         Entity other = ((CraftEntity) entity).getHandle();
 
         if (other instanceof ServerPlayer) {
             ServerPlayer otherPlayer = (ServerPlayer) other;
+            // Paper start - uuid override
             UUID original = null;
             if (uuidOverride != null) {
                 original = otherPlayer.getUUID();
                 otherPlayer.setUUID(uuidOverride);
             }
+            // Paper end
             this.getHandle().connection.send(ClientboundPlayerInfoUpdatePacket.createPlayerInitializing(List.of(otherPlayer), this.getHandle())); // Paper - Add Listing API for Player
             if (original != null) otherPlayer.setUUID(original); // Paper - uuid override
         }
@@ -2181,7 +_,7 @@
         for (ServerPlayer player : players) {
             CraftPlayer bukkitPlayer = player.getBukkitEntity();
             if (bukkitPlayer.canSee(this)) {
-                bukkitPlayer.trackAndShowEntity(self.getBukkitEntity(), gameProfile.id());
+                bukkitPlayer.trackAndShowEntity(self.getBukkitEntity(), gameProfile.getId());
             }
         }
 
@@ -2202,7 +_,7 @@
     }
     // Paper start
     public com.destroystokyo.paper.profile.PlayerProfile getPlayerProfile() {
-        return new com.destroystokyo.paper.profile.CraftPlayerProfile(this);
+        return new com.destroystokyo.paper.profile.CraftPlayerProfile(this).clone();
     }
 
     private void refreshPlayer() {
@@ -2215,7 +_,7 @@
         connection.send(new net.minecraft.network.protocol.game.ClientboundRespawnPacket(handle.createCommonSpawnInfo(level), net.minecraft.network.protocol.game.ClientboundRespawnPacket.KEEP_ALL_DATA));
         handle.onUpdateAbilities();
         connection.internalTeleport(net.minecraft.world.entity.PositionMoveRotation.of(this.getHandle()), java.util.Collections.emptySet());
-        net.minecraft.server.players.PlayerList playerList = this.server.getServer().getPlayerList();
+        net.minecraft.server.players.PlayerList playerList = handle.getServer().getPlayerList();
         playerList.sendPlayerPermissionLevel(handle, false);
         playerList.sendLevelInfo(handle, level);
         playerList.sendAllPlayerInfo(handle);
@@ -2285,7 +_,7 @@
 
     @Override
     public Map<String, Object> serialize() {
-        Map<String, Object> result = new LinkedHashMap<>();
+        Map<String, Object> result = new LinkedHashMap<String, Object>();
 
         result.put("name", this.getName());
 
@@ -2355,7 +_,7 @@
         data.putBoolean("keepLevel", handle.keepLevel);
         data.putLong("firstPlayed", this.getFirstPlayed());
         data.putLong("lastPlayed", System.currentTimeMillis());
-        data.putString("lastKnownName", handle.getGameProfile().name());
+        data.putString("lastKnownName", handle.getScoreboardName());
 
         // Paper start - persist for use in offline save data
         ValueOutput paper = output.child("Paper");
@@ -2406,6 +_,31 @@
     }
 
     @Override
+    public void setTexturePack(String url) {
+        this.setResourcePack(url);
+    }
+
+    @Override
+    public void setResourcePack(String url) {
+        this.setResourcePack(url, (byte[]) null);
+    }
+
+    @Override
+    public void setResourcePack(String url, byte @Nullable [] hash) {
+        this.setResourcePack(url, hash, false);
+    }
+
+    @Override
+    public void setResourcePack(String url, byte @Nullable [] hash, String prompt) {
+        this.setResourcePack(url, hash, prompt, false);
+    }
+
+    @Override
+    public void setResourcePack(String url, byte @Nullable [] hash, boolean force) {
+        this.setResourcePack(url, hash, (String) null, force);
+    }
+
+    @Override
     public void setResourcePack(String url, byte @Nullable [] hash, String prompt, boolean force) {
         Preconditions.checkArgument(url != null, "Resource pack URL cannot be null");
 
@@ -2596,7 +_,7 @@
         return true;
     }
 
-    public void disconnect() {
+    public void disconnect(String reason) {
         this.conversationTracker.abandonAllConversations();
         this.perm.clearPermissions();
     }
@@ -2664,7 +_,7 @@
 
     @Override
     public float getFlySpeed() {
-        return this.getHandle().getAbilities().flyingSpeed * 2f;
+        return (float) this.getHandle().getAbilities().flyingSpeed * 2f;
     }
 
     @Override
@@ -2862,14 +_,79 @@
 
     @Override
     public void resetTitle() {
-        ClientboundClearTitlesPacket packet = new ClientboundClearTitlesPacket(true);
-        this.getHandle().connection.send(packet);
+        ClientboundClearTitlesPacket packetReset = new ClientboundClearTitlesPacket(true);
+        this.getHandle().connection.send(packetReset);
+    }
+
+    @Override
+    public void spawnParticle(Particle particle, Location location, int count) {
+        this.spawnParticle(particle, location.getX(), location.getY(), location.getZ(), count);
+    }
+
+    @Override
+    public void spawnParticle(Particle particle, double x, double y, double z, int count) {
+        this.spawnParticle(particle, x, y, z, count, null);
+    }
+
+    @Override
+    public <T> void spawnParticle(Particle particle, Location location, int count, T data) {
+        this.spawnParticle(particle, location.getX(), location.getY(), location.getZ(), count, data);
+    }
+
+    @Override
+    public <T> void spawnParticle(Particle particle, double x, double y, double z, int count, T data) {
+        this.spawnParticle(particle, x, y, z, count, 0, 0, 0, data);
+    }
+
+    @Override
+    public void spawnParticle(Particle particle, Location location, int count, double offsetX, double offsetY, double offsetZ) {
+        this.spawnParticle(particle, location.getX(), location.getY(), location.getZ(), count, offsetX, offsetY, offsetZ);
+    }
+
+    @Override
+    public void spawnParticle(Particle particle, double x, double y, double z, int count, double offsetX, double offsetY, double offsetZ) {
+        this.spawnParticle(particle, x, y, z, count, offsetX, offsetY, offsetZ, null);
+    }
+
+    @Override
+    public <T> void spawnParticle(Particle particle, Location location, int count, double offsetX, double offsetY, double offsetZ, T data) {
+        this.spawnParticle(particle, location.getX(), location.getY(), location.getZ(), count, offsetX, offsetY, offsetZ, data);
+    }
+
+    @Override
+    public <T> void spawnParticle(Particle particle, double x, double y, double z, int count, double offsetX, double offsetY, double offsetZ, T data) {
+        this.spawnParticle(particle, x, y, z, count, offsetX, offsetY, offsetZ, 1, data);
+    }
+
+    @Override
+    public void spawnParticle(Particle particle, Location location, int count, double offsetX, double offsetY, double offsetZ, double extra) {
+        this.spawnParticle(particle, location.getX(), location.getY(), location.getZ(), count, offsetX, offsetY, offsetZ, extra);
+    }
+
+    @Override
+    public void spawnParticle(Particle particle, double x, double y, double z, int count, double offsetX, double offsetY, double offsetZ, double extra) {
+        this.spawnParticle(particle, x, y, z, count, offsetX, offsetY, offsetZ, extra, null);
+    }
+
+    @Override
+    public <T> void spawnParticle(Particle particle, Location location, int count, double offsetX, double offsetY, double offsetZ, double extra, T data) {
+        this.spawnParticle(particle, location.getX(), location.getY(), location.getZ(), count, offsetX, offsetY, offsetZ, extra, data);
+    }
+
+    @Override
+    public <T> void spawnParticle(Particle particle, double x, double y, double z, int count, double offsetX, double offsetY, double offsetZ, double extra, T data) {
+        this.spawnParticle(particle, x, y, z, count, offsetX, offsetY, offsetZ, extra, data, false);
+    }
+
+    @Override
+    public <T> void spawnParticle(Particle particle, Location location, int count, double offsetX, double offsetY, double offsetZ, double extra, T data, boolean force) {
+        this.spawnParticle(particle, location.getX(), location.getY(), location.getZ(), count, offsetX, offsetY, offsetZ, extra, data, force);
     }
 
     @Override
     public <T> void spawnParticle(Particle particle, double x, double y, double z, int count, double offsetX, double offsetY, double offsetZ, double extra, T data, boolean force) {
-        ClientboundLevelParticlesPacket packet = new ClientboundLevelParticlesPacket(CraftParticle.createParticleParam(particle, data), force, false, x, y, z, (float) offsetX, (float) offsetY, (float) offsetZ, (float) extra, count); // Paper - fix x/y/z precision loss
-        this.getHandle().connection.send(packet);
+        ClientboundLevelParticlesPacket packetplayoutworldparticles = new ClientboundLevelParticlesPacket(CraftParticle.createParticleParam(particle, data), force, false, x, y, z, (float) offsetX, (float) offsetY, (float) offsetZ, (float) extra, count); // Paper - fix x/y/z precision loss
+        this.getHandle().connection.send(packetplayoutworldparticles);
     }
 
     @Override
@@ -2924,7 +_,7 @@
     public void updateCommands() {
         if (this.getHandle().connection == null) return;
 
-        this.server.getServer().getCommands().sendCommands(this.getHandle());
+        this.getHandle().getServer().getCommands().sendCommands(this.getHandle());
     }
 
     @Override
@@ -2939,6 +_,11 @@
     }
 
     @Override
+    public void openSign(Sign sign) {
+        this.openSign(sign, Side.FRONT);
+    }
+
+    @Override
     public void openSign(@NonNull Sign sign, @NonNull Side side) {
         CraftSign.openSign(sign, this, side);
     }
@@ -3083,8 +_,8 @@
 
     @Override
     public <T> void sendTitlePart(final net.kyori.adventure.title.TitlePart<T> part, T value) {
-        Preconditions.checkArgument(part != null, "part cannot be null");
-        Preconditions.checkArgument(value != null, "value cannot be null");
+        java.util.Objects.requireNonNull(part, "part");
+        java.util.Objects.requireNonNull(value, "value");
         if (part == net.kyori.adventure.title.TitlePart.TITLE) {
             final ClientboundSetTitleTextPacket tp = new ClientboundSetTitleTextPacket(io.papermc.paper.adventure.PaperAdventure.asVanilla((net.kyori.adventure.text.Component) value));
             this.getHandle().connection.send(tp);
@@ -3473,81 +_,5 @@
     @Override
     public PlayerGameConnection getConnection() {
         return this.getHandle().connection.playerGameConnection;
-    }
-
-    @Override
-    public org.bukkit.entity.Entity releaseLeftShoulderEntity() {
-        if (!getHandle().getShoulderEntityLeft().isEmpty()) {
-            Entity entity = getHandle().releaseLeftShoulderEntity();
-            if (entity != null) {
-                return entity.getBukkitEntity();
-            }
-        }
-
-        return null;
-    }
-
-    @Override
-    public org.bukkit.entity.Entity releaseRightShoulderEntity() {
-        if (!getHandle().getShoulderEntityRight().isEmpty()) {
-            Entity entity = getHandle().releaseRightShoulderEntity();
-            if (entity != null) {
-                return entity.getBukkitEntity();
-            }
-        }
-
-        return null;
-    }
-
-    @Override
-    public org.bukkit.entity.Entity getShoulderEntityLeft() {
-        if (!this.getHandle().getShoulderEntityLeft().isEmpty()) {
-            try (ProblemReporter.ScopedCollector scopedCollector = new ProblemReporter.ScopedCollector(this.getHandle().problemPath(), LOGGER)) {
-                return net.minecraft.world.entity.EntityType.create(
-                    TagValueInput.create(scopedCollector.forChild(() -> ".shoulder"), this.getHandle().registryAccess(), this.getHandle().getShoulderEntityLeft()),
-                    this.getHandle().level(),
-                    EntitySpawnReason.LOAD
-                ).map(Entity::getBukkitEntity).orElse(null);
-            }
-        }
-
-        return null;
-    }
-
-    @Override
-    public void setShoulderEntityLeft(org.bukkit.entity.Entity entity) {
-        if (entity != null) {
-            Preconditions.checkArgument(((CraftEntity) entity).getHandle().getType().canSerialize(), "Cannot set entity of type %s as a shoulder entity", entity.getType().getKey());
-        }
-        this.getHandle().setShoulderEntityLeft(entity == null ? new CompoundTag() : ((CraftEntity) entity).save());
-        if (entity != null) {
-            entity.remove();
-        }
-    }
-
-    @Override
-    public org.bukkit.entity.Entity getShoulderEntityRight() {
-        if (!this.getHandle().getShoulderEntityRight().isEmpty()) {
-            try (ProblemReporter.ScopedCollector scopedCollector = new ProblemReporter.ScopedCollector(this.getHandle().problemPath(), LOGGER)) {
-                return net.minecraft.world.entity.EntityType.create(
-                    TagValueInput.create(scopedCollector.forChild(() -> ".shoulder"), this.getHandle().registryAccess(), this.getHandle().getShoulderEntityRight()),
-                    this.getHandle().level(),
-                    EntitySpawnReason.LOAD
-                ).map(Entity::getBukkitEntity).orElse(null);
-            }
-        }
-
-        return null;
-    }
-
-    @Override
-    public void setShoulderEntityRight(org.bukkit.entity.Entity entity) {
-        if (entity != null) {
-            Preconditions.checkArgument(((CraftEntity) entity).getHandle().getType().canSerialize(), "Cannot set entity of type %s as a shoulder entity", entity.getType().getKey());
-        }
-        this.getHandle().setShoulderEntityRight(entity == null ? new CompoundTag() : ((CraftEntity) entity).save());
-        if (entity != null) {
-            entity.remove();
-        }
     }
 }
