--- a/src/main/java/org/bukkit/craftbukkit/entity/CraftMob.java
+++ b/src/main/java/org/bukkit/craftbukkit/entity/CraftMob.java
@@ -22,19 +_,8 @@
     }
 
     @Override
-    public net.minecraft.world.entity.Mob getHandle() {
-        return (net.minecraft.world.entity.Mob) this.entity;
-    }
-
-    @Override
-    public void setHandle(net.minecraft.world.entity.Entity entity) {
-        super.setHandle(entity);
-        this.paperPathfinder.setHandle(this.getHandle());
-    }
-
-    @Override
     public boolean shouldDespawnInPeaceful() {
-        return this.getHandle().shouldDespawnInPeaceful();
+        return this.getHandle().shouldActuallyDespawnInPeaceful();
     }
 
     @Override
@@ -46,6 +_,17 @@
     @Override
     public TriState getDespawnInPeacefulOverride() {
         return this.getHandle().despawnInPeacefulOverride;
+    }
+
+    @Override
+    public net.minecraft.world.entity.Mob getHandle() {
+        return (net.minecraft.world.entity.Mob) this.entity;
+    }
+
+    @Override
+    public void setHandle(net.minecraft.world.entity.Entity entity) {
+        super.setHandle(entity);
+        this.paperPathfinder.setHandle(getHandle());
     }
 
     @Override
