From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Ivo Quiring <falk.quiring@gmail.com>
Date: Sun, 2 Nov 2025 15:56:26 +0100
Subject: [PATCH] Overhaul Operator Database


diff --git a/src/main/java/net/kissenpvp/base/PulvinarCore.java b/src/main/java/net/kissenpvp/base/PulvinarCore.java
index 5bb00abe99474c3ee0ad71c93e6c9ee8497437a3..0bbd671fa09db023f89cfb8a50098498900caaf3 100644
--- a/src/main/java/net/kissenpvp/base/PulvinarCore.java
+++ b/src/main/java/net/kissenpvp/base/PulvinarCore.java
@@ -4,6 +4,7 @@ import net.kissenpvp.api.base.Pulvinar;
 import net.kissenpvp.api.database.ConnectionProvider;
 import net.kissenpvp.api.network.actor.ConsoleClient;
 import net.kissenpvp.database.InternalConnectionProvider;
+import net.kissenpvp.network.actor.operator.PulvinarOperatorRepository;
 import org.bukkit.Bukkit;
 import org.bukkit.configuration.InvalidConfigurationException;
 import org.flywaydb.core.Flyway;
@@ -29,6 +30,8 @@ public class PulvinarCore extends KissenCore implements Pulvinar
     private final ConnectionProvider provider;
     private final ConnectionProvider privateProvider;
 
+    private PulvinarOperatorRepository operatorRepository;
+
     public PulvinarCore(@NonNull File optionsFile) throws IOException, InvalidConfigurationException, SQLException {
         config = new PulvinarConfig();
         config.init(optionsFile);
@@ -75,7 +78,7 @@ public class PulvinarCore extends KissenCore implements Pulvinar
 
     private void createRepositories(@NonNull DataSource globalDatasource, @NonNull DataSource privateDatasource)
     {
-
+        operatorRepository = new PulvinarOperatorRepository(privateDatasource);
     }
 
     public void shutdown()
@@ -100,4 +103,8 @@ public class PulvinarCore extends KissenCore implements Pulvinar
     public @NonNull ConnectionProvider privateConnectionProvider() {
         return privateProvider;
     }
+
+    public @NonNull PulvinarOperatorRepository operatorRepository() {
+        return operatorRepository;
+    }
 }
diff --git a/src/main/java/net/kissenpvp/network/actor/operator/PulvinarOperatorInfo.java b/src/main/java/net/kissenpvp/network/actor/operator/PulvinarOperatorInfo.java
new file mode 100644
index 0000000000000000000000000000000000000000..e1686de9702b07475edfa2b2c794ff216a298b63
--- /dev/null
+++ b/src/main/java/net/kissenpvp/network/actor/operator/PulvinarOperatorInfo.java
@@ -0,0 +1,18 @@
+package net.kissenpvp.network.actor.operator;
+
+import net.kissenpvp.api.database.PersistableEntity;
+import org.jspecify.annotations.NonNull;
+
+import java.util.UUID;
+
+public interface PulvinarOperatorInfo extends PersistableEntity<UUID> {
+
+    @NonNull UUID id();
+
+    @NonNull String name();
+
+    int getLevel();
+
+    boolean getBypassesPlayerLimit();
+
+}
diff --git a/src/main/java/net/kissenpvp/network/actor/operator/PulvinarOperatorList.java b/src/main/java/net/kissenpvp/network/actor/operator/PulvinarOperatorList.java
new file mode 100644
index 0000000000000000000000000000000000000000..89cd7dc5d06b8f9c609ab5c0b588f5705254acfb
--- /dev/null
+++ b/src/main/java/net/kissenpvp/network/actor/operator/PulvinarOperatorList.java
@@ -0,0 +1,53 @@
+package net.kissenpvp.network.actor.operator;
+
+import net.kissenpvp.api.database.Repository;
+import net.minecraft.server.notifications.NotificationService;
+import net.minecraft.server.players.ServerOpList;
+import net.minecraft.server.players.ServerOpListEntry;
+import org.jspecify.annotations.NonNull;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.File;
+import java.io.IOException;
+import java.util.Collection;
+import java.util.Objects;
+import java.util.UUID;
+
+public class PulvinarOperatorList extends ServerOpList
+{
+    private static final Logger log = LoggerFactory.getLogger(PulvinarOperatorList.class);
+    private final Repository<UUID, PulvinarOperatorInfo> repository;
+
+    public PulvinarOperatorList(@NonNull File file, @NonNull Repository<UUID, PulvinarOperatorInfo> repository, NotificationService notificationService)
+    {
+        super(file, notificationService);
+        this.repository = repository;
+    }
+
+    @Override
+    public void load() throws IOException {
+        Collection<PulvinarOperatorInfo> operators = repository.findAll().join();
+
+        for(PulvinarOperatorInfo operator : operators)
+        {
+            ServerOpListEntry entry = (ServerOpListEntry) operator;
+            if(Objects.isNull(entry.getUser()))
+            {
+                log.warn("Discarding operator entry {} due to missing user.", operator);
+                continue;
+            }
+
+            String key = getKeyForUser(entry.getUser());
+            getMap().put(key, entry);
+        }
+    }
+
+    @Override
+    public void save() throws IOException {
+        Collection<? extends PulvinarOperatorInfo> ops = getMap().values();
+
+        //noinspection unchecked
+        repository.saveAll((Iterable<PulvinarOperatorInfo>) ops); // look. I know this is stupid. But just accept that this is java
+    }
+}
diff --git a/src/main/java/net/kissenpvp/network/actor/operator/PulvinarOperatorRepository.java b/src/main/java/net/kissenpvp/network/actor/operator/PulvinarOperatorRepository.java
new file mode 100644
index 0000000000000000000000000000000000000000..9b2cca05b39b940395c8e8019fe62d0c3dd26109
--- /dev/null
+++ b/src/main/java/net/kissenpvp/network/actor/operator/PulvinarOperatorRepository.java
@@ -0,0 +1,134 @@
+package net.kissenpvp.network.actor.operator;
+
+import com.google.common.base.Preconditions;
+import net.kissenpvp.api.database.KissenRepository;
+import net.kissenpvp.base.PulvinarCore;
+import net.minecraft.server.permissions.LevelBasedPermissionSet;
+import net.minecraft.server.permissions.PermissionLevel;
+import net.minecraft.server.players.NameAndId;
+import net.minecraft.server.players.ServerOpListEntry;
+import org.bukkit.Bukkit;
+import org.jspecify.annotations.NonNull;
+
+import javax.sql.DataSource;
+import java.sql.PreparedStatement;
+import java.sql.ResultSet;
+import java.sql.SQLException;
+import java.util.*;
+import java.util.concurrent.CompletableFuture;
+
+public class PulvinarOperatorRepository extends KissenRepository<UUID, PulvinarOperatorInfo> {
+    public PulvinarOperatorRepository(@NonNull DataSource dataSource) throws NullPointerException {
+        super(dataSource);
+    }
+
+    private static @NonNull String globalDatabase() {
+        return ((PulvinarCore) Bukkit.getPulvinar()).config().databaseGlobal();
+    }
+
+    private static @NonNull PulvinarOperatorInfo toEntity(@NonNull UUID id, @NonNull ResultSet resultSet) throws SQLException {
+        Preconditions.checkNotNull(id, "Id must not be null.");
+        Preconditions.checkNotNull(resultSet, "Resultset must not be null.");
+
+        String username = resultSet.getString("username");
+
+        NameAndId identify = new NameAndId(id, username);
+        int level = resultSet.getInt("operator_level");
+        boolean canBypassPlayerLimit = resultSet.getBoolean("can_bypass_player_limit");
+
+        LevelBasedPermissionSet permissionSet = LevelBasedPermissionSet.forLevel(PermissionLevel.byId(level));
+        return new ServerOpListEntry(identify, permissionSet, canBypassPlayerLimit);
+    }
+
+
+    private static @NonNull Collection<PulvinarOperatorInfo> collectResults(@NonNull PreparedStatement statement) throws SQLException {
+        Preconditions.checkNotNull(statement, "statement must not be null.");
+
+        Collection<PulvinarOperatorInfo> collection = new HashSet<>();
+
+        try (ResultSet resultSet = statement.executeQuery()) {
+            while (resultSet.next()) {
+                UUID uuid = UUID.fromString(resultSet.getString("id"));
+                collection.add(toEntity(uuid, resultSet));
+            }
+        }
+
+        return Collections.unmodifiableCollection(collection);
+    }
+
+    @Override
+    public @NonNull CompletableFuture<@NonNull Optional<PulvinarOperatorInfo>> find(@NonNull UUID id) throws NullPointerException {
+        Preconditions.checkNotNull(id, "Id cannot be null.");
+
+        String sql = "SELECT o.id, p.username AS username, o.operator_level, o.can_bypass_player_limit FROM ksvp_operators o JOIN %s.ksvp_player p ON o.id = p.id WHERE o.id = ?;";
+        return CompletableFuture.supplyAsync(() -> assumeNotNull(this.query(String.format(sql, globalDatabase()), (statement) -> {
+            statement.setString(1, String.valueOf(id));
+            try (ResultSet resultSet = statement.executeQuery()) {
+                if (!resultSet.next()) {
+                    return Optional.empty();
+                }
+
+                UUID uuid = UUID.fromString(resultSet.getString("id"));
+                return Optional.of(toEntity(uuid, resultSet));
+            }
+        })));
+    }
+
+    @Override
+    public @NonNull CompletableFuture<@NonNull Collection<PulvinarOperatorInfo>> findAll(@NonNull Iterable<UUID> id) throws NullPointerException {
+        Preconditions.checkNotNull(id, "Iterable cannot be null.");
+
+        String placeHolders = String.join(", ", Collections.nCopies(this.computeIterableSize(id), "?"));
+        String sql = "SELECT o.id, p.username AS username, o.operator_level, o.can_bypass_player_limit FROM ksvp_operators o JOIN %s.ksvp_player p ON o.id = p.id WHERE o.id IN (" + placeHolders + ");";
+        return CompletableFuture.supplyAsync(() -> assumeNotNull(query(String.format(sql, globalDatabase()), (statement) -> {
+            int index = 1;
+
+            for (UUID current : id) {
+                statement.setString(index++, String.valueOf(current));
+            }
+
+            return collectResults(statement);
+        })));
+    }
+
+    @Override
+    public @NonNull CompletableFuture<@NonNull Collection<PulvinarOperatorInfo>> findAll() {
+        String sql = "SELECT o.id, p.username AS username, o.operator_level, o.can_bypass_player_limit FROM ksvp_operators o JOIN %s.ksvp_player p ON o.id = p.id;";
+        return CompletableFuture.supplyAsync(() -> assumeNotNull(query(String.format(sql, globalDatabase()), PulvinarOperatorRepository::collectResults)));
+    }
+
+    @Override
+    public @NonNull CompletableFuture<@NonNull Boolean> has(@NonNull UUID id) throws NullPointerException {
+        Preconditions.checkNotNull(id, "id must not be null.");
+
+        String sql = "SELECT id FROM ksvp_operators WHERE id = ?;";
+        return CompletableFuture.supplyAsync(() -> assumeNotNull(query(sql, (statement) -> {
+            statement.setString(1, String.valueOf(id));
+            try (ResultSet resultSet = statement.executeQuery()) {
+                return resultSet.next();
+            }
+        })));
+    }
+
+    @Override
+    public @NonNull CompletableFuture<Void> saveAll(@NonNull Iterable<PulvinarOperatorInfo> id) throws NullPointerException {
+        Preconditions.checkNotNull(id, "Iterable must not be null.");
+
+        String sql = "INSERT INTO ksvp_operators (id, operator_level, can_bypass_player_limit) VALUES (?, ?, ?) ON DUPLICATE KEY UPDATE operator_level = ?, can_bypass_player_limit = ?;";
+        return CompletableFuture.supplyAsync(() -> query(sql, (statement) -> {
+            for (PulvinarOperatorInfo pulvinarOperatorInfo : id) {
+                statement.setString(1, String.valueOf(pulvinarOperatorInfo.id()));
+
+                statement.setInt(2, pulvinarOperatorInfo.getLevel());
+                statement.setInt(4, pulvinarOperatorInfo.getLevel());
+
+                statement.setBoolean(3, pulvinarOperatorInfo.getBypassesPlayerLimit());
+                statement.setBoolean(5, pulvinarOperatorInfo.getBypassesPlayerLimit());
+                statement.addBatch();
+            }
+
+            statement.executeBatch();
+            return null;
+        }));
+    }
+}
diff --git a/src/main/resources/migrations/private/mariadb/V1_0__Intial_migration.sql b/src/main/resources/migrations/private/mariadb/V1_0__Intial_migration.sql
new file mode 100644
index 0000000000000000000000000000000000000000..2ccc7c0adb50c56458f4f5d529d040719f814534
--- /dev/null
+++ b/src/main/resources/migrations/private/mariadb/V1_0__Intial_migration.sql
@@ -0,0 +1,7 @@
+CREATE TABLE IF NOT EXISTS ksvp_operators (
+    id UUID NOT NULL,
+    operator_level INT NOT NULL,
+    can_bypass_player_limit BOOLEAN NOT NULL,
+    PRIMARY KEY (id),
+    FOREIGN KEY (id) REFERENCES ${global_database}.ksvp_player(id) ON DELETE CASCADE
+) COMMENT = "This table stores operators and their permissions.";
\ No newline at end of file
