From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Ivo Quiring <falk.quiring@gmail.com>
Date: Thu, 28 Aug 2025 19:05:59 +0200
Subject: [PATCH] Add pulvinar to bukkit


diff --git a/src/main/java/com/destroystokyo/paper/profile/PaperAuthenticationService.java b/src/main/java/com/destroystokyo/paper/profile/PaperAuthenticationService.java
index c36bbf2585507f12d7b7b91cae19d50c4213b4c7..f2d30ce7d9b94088ae9834e07cbdf5d6b691feea 100644
--- a/src/main/java/com/destroystokyo/paper/profile/PaperAuthenticationService.java
+++ b/src/main/java/com/destroystokyo/paper/profile/PaperAuthenticationService.java
@@ -29,7 +29,7 @@ public class PaperAuthenticationService extends YggdrasilAuthenticationService {
         return new PaperGameProfileRepository(
                 this.getProxy(),
                 this.environment,
-                org.bukkit.Bukkit.getPulvinar().playerModule().playerRepository()
+                ((net.kissenpvp.api.database.RepositoryHolder) org.bukkit.Bukkit.getPulvinar()).playerRepository()
         );
         // Pulvinar end - integrate player repository into paper
     }
diff --git a/src/main/java/com/destroystokyo/paper/profile/PaperGameProfileRepository.java b/src/main/java/com/destroystokyo/paper/profile/PaperGameProfileRepository.java
index 10fe41e926e8378c8e5e4b311aff3eef998325ba..6e5e3734e42cea762054c57175913ab844f52dc3 100644
--- a/src/main/java/com/destroystokyo/paper/profile/PaperGameProfileRepository.java
+++ b/src/main/java/com/destroystokyo/paper/profile/PaperGameProfileRepository.java
@@ -81,17 +81,12 @@ public class PaperGameProfileRepository extends YggdrasilGameProfileRepository {
     // Pulvinar start - implement player repository into paper
     @Override
     public java.util.Optional<GameProfile> findProfileByName(String name) {
-        net.kissenpvp.api.network.actor.PlayerClient playerClient = playerRepository.findByName(name).join();
+        java.util.Optional<net.kissenpvp.api.network.actor.PlayerClient> playerClient = playerRepository.findByName(name).join();
 
-        if(java.util.Objects.nonNull(playerClient))
-        {
-            return java.util.Optional.of(new com.mojang.authlib.GameProfile(
-                    playerClient.id(),
-                    playerClient.username()
-            ));
-        }
-
-        return super.findProfileByName(name);
+        return playerClient.map(client -> new GameProfile(
+                client.id(),
+                client.username()
+        )).or(() -> super.findProfileByName(name));
     }
 
     public boolean pulvinarHooked()
diff --git a/src/main/java/io/papermc/paper/plugin/manager/PaperPluginManagerImpl.java b/src/main/java/io/papermc/paper/plugin/manager/PaperPluginManagerImpl.java
index 097500a59336db1bbfffcd1aa4cff7a8586e46ec..3922e9d51fef0952a3344af6acb56e91b44c663c 100644
--- a/src/main/java/io/papermc/paper/plugin/manager/PaperPluginManagerImpl.java
+++ b/src/main/java/io/papermc/paper/plugin/manager/PaperPluginManagerImpl.java
@@ -243,4 +243,11 @@ public class PaperPluginManagerImpl implements PluginManager, DependencyContext
     public MutableGraph<String> getInstanceManagerGraph() {
         return instanceManager.getDependencyGraph();
     }
+
+    // Pulvinar start - implement player property repository
+    @Override
+    public @NotNull net.kissenpvp.api.network.actor. @org.jspecify.annotations.NonNull PlayerProperties playerProperties(@NotNull @org.jspecify.annotations.NonNull Plugin plugin, @NotNull org.bukkit.@org.jspecify.annotations.NonNull OfflinePlayer player) {
+        return ((net.kissenpvp.api.database.RepositoryHolder) org.bukkit.Bukkit.getPulvinar()).playerRepository().playerProperties(plugin, player);
+    }
+    // Pulvinar end - implement player property repository
 }
diff --git a/src/main/java/net/kissenpvp/base/PulvinarConfig.java b/src/main/java/net/kissenpvp/base/PulvinarConfig.java
index 29b557a6cf9d57602bc622fbec0f90271fbfc80a..80881cb8f73a6f8c40a2546b69b61ef6e463b413 100644
--- a/src/main/java/net/kissenpvp/base/PulvinarConfig.java
+++ b/src/main/java/net/kissenpvp/base/PulvinarConfig.java
@@ -2,7 +2,7 @@ package net.kissenpvp.base;
 
 import org.bukkit.configuration.InvalidConfigurationException;
 import org.bukkit.configuration.file.YamlConfiguration;
-import org.jetbrains.annotations.NotNull;
+import org.jspecify.annotations.NonNull;
 
 import java.io.File;
 import java.io.FileNotFoundException;
@@ -32,7 +32,7 @@ public class PulvinarConfig {
     private String databaseUser;
     private String databasePassword;
 
-    public void init(@NotNull File configFile) throws IOException, InvalidConfigurationException {
+    public void init(@NonNull File configFile) throws IOException, InvalidConfigurationException {
         file = configFile;
         config = new YamlConfiguration();
 
@@ -68,15 +68,15 @@ public class PulvinarConfig {
         return databaseConfigured;
     }
 
-    public @NotNull String databaseUrl() {
+    public @NonNull String databaseUrl() {
         return databaseUrl;
     }
 
-    public @NotNull String databaseUser() {
+    public @NonNull String databaseUser() {
         return databaseUser;
     }
 
-    public @NotNull String databasePassword() {
+    public @NonNull String databasePassword() {
         return databasePassword;
     }
 
@@ -85,19 +85,19 @@ public class PulvinarConfig {
         config.save(file);
     }
 
-    public void databaseUrl(@NotNull String url) throws IOException, NullPointerException {
+    public void databaseUrl(@NonNull String url) throws IOException, NullPointerException {
         Objects.requireNonNull(url, "url cannot be null");
         databaseUrl = url;
         config.save(file);
     }
 
-    public void databaseUser(@NotNull String user) throws IOException, NullPointerException {
+    public void databaseUser(@NonNull String user) throws IOException, NullPointerException {
         Objects.requireNonNull(user, "user cannot be null");
         databaseUser = user;
         config.save(file);
     }
 
-    public void databasePassword(@NotNull String password) throws IOException, NullPointerException {
+    public void databasePassword(@NonNull String password) throws IOException, NullPointerException {
         Objects.requireNonNull(password, "password cannot be null");
         databasePassword = password;
         config.save(file);
diff --git a/src/main/java/net/kissenpvp/base/PulvinarCore.java b/src/main/java/net/kissenpvp/base/PulvinarCore.java
index c42169f7ac76f2ebc2c0d48fba93183df0caf19a..dcbfc0b566ad95b0e44501356f1b5c638b315d38 100644
--- a/src/main/java/net/kissenpvp/base/PulvinarCore.java
+++ b/src/main/java/net/kissenpvp/base/PulvinarCore.java
@@ -1,27 +1,35 @@
 package net.kissenpvp.base;
 
+import net.kissenpvp.api.base.KissenPlugin;
+import net.kissenpvp.api.base.Pulvinar;
 import net.kissenpvp.api.database.ConnectionProvider;
-import net.kissenpvp.api.network.actor.PlayerModule;
-import net.kissenpvp.api.network.actor.rank.RankModule;
-import net.kissenpvp.api.punishment.PunishmentModule;
+import net.kissenpvp.api.network.actor.ConsoleClient;
+import net.kissenpvp.api.network.actor.PlayerClient;
+import net.kissenpvp.api.network.actor.PlayerProperties;
+import net.kissenpvp.api.network.actor.PlayerRepository;
 import net.kissenpvp.database.InternalConnectionProvider;
-import net.kissenpvp.network.actor.InternalPlayerModule;
+import net.kissenpvp.network.actor.InternalPlayerRepository;
 import net.kissenpvp.network.actor.PulvinarPlayerRepository;
-import net.kissenpvp.network.actor.rank.InternalRankModule;
+import net.kissenpvp.network.actor.operator.PulvinarOperatorRepository;
+import net.kissenpvp.network.actor.properties.PropertyRepository;
 import net.kissenpvp.network.actor.rank.repository.InternalRankRepository;
 import net.kissenpvp.network.actor.rank.repository.InternalRankSubscriptionRepository;
-import net.kissenpvp.punishment.InternalPunishmentModule;
-import net.kissenpvp.punishment.repository.PunishmentRepository;
-import net.kissenpvp.punishment.repository.PunishmentSubscriptionRepository;
+import net.kissenpvp.punishment.repository.InternalPunishmentRepository;
+import net.kissenpvp.punishment.repository.InternalPunishmentSubscriptionRepository;
+import org.bukkit.Bukkit;
 import org.bukkit.configuration.InvalidConfigurationException;
-import org.jetbrains.annotations.NotNull;
+import org.jspecify.annotations.NonNull;
+import org.jspecify.annotations.Nullable;
 import org.slf4j.Logger;
 import org.slf4j.LoggerFactory;
 
+import javax.sql.DataSource;
 import java.io.File;
 import java.io.IOException;
-import java.sql.Connection;
 import java.sql.SQLException;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Objects;
 
 /**
  * Allows access to the kissen api
@@ -29,9 +37,15 @@ import java.sql.SQLException;
 public class PulvinarCore extends KissenCore implements Pulvinar
 {
     private static final Logger log = LoggerFactory.getLogger(PulvinarCore.class);
-    private PulvinarConfig config;
+    private final PulvinarConfig config;
+
+    private ConnectionProvider provider;
+
+    private Map<KissenPlugin, PropertyRepository> playerRepositories;
+
+    public PulvinarCore(@NonNull File optionsFile) throws IOException, InvalidConfigurationException, SQLException {
+        this.playerRepositories = new HashMap<>();
 
-    public PulvinarCore(@NotNull File optionsFile) throws IOException, InvalidConfigurationException, SQLException {
         config = new PulvinarConfig();
         config.init(optionsFile);
 
@@ -46,36 +60,50 @@ public class PulvinarCore extends KissenCore implements Pulvinar
             throw new IllegalStateException("Could not connect to database, please check your configuration.");
         }
 
-        ConnectionProvider provider = new InternalConnectionProvider();
+        provider = new InternalConnectionProvider();
         provider.connect(config.databaseUrl(), config.databaseUser(), config.databasePassword());
 
-        if(provider.connection().isEmpty())
+        if(provider.dataSource().isEmpty())
         {
-            throw new SQLException("The providers connection is not present. Couldn't it connect?");
+            throw new SQLException("The providers dataSource is not present. Couldn't it connect?");
         }
 
-        Connection connection = provider.connection().get();
+        DataSource dataSource = provider.dataSource().get();
 
-        PunishmentRepository pRepository = new PunishmentRepository(connection);
-        PunishmentSubscriptionRepository psRepository = new PunishmentSubscriptionRepository(connection);
-        PunishmentModule punishmentModule = new InternalPunishmentModule(pRepository, psRepository);
-        log.debug("The punishment module has been initialized.");
+        InternalRepositoryHolder.RepositoryHolderBuilder builder = InternalRepositoryHolder.builder();
 
-        InternalRankRepository rRepository = new InternalRankRepository(connection);
-        InternalRankSubscriptionRepository rsRepository = new InternalRankSubscriptionRepository(connection);
-        RankModule rankModule = new InternalRankModule(rRepository, rsRepository);
-        log.debug("The rank module has been initialized.");
+        builder.punishmentRepository(new InternalPunishmentRepository(dataSource));
+        builder.punishmentSubscriptionRepository(new InternalPunishmentSubscriptionRepository(dataSource));
 
-        PulvinarPlayerRepository playerRepository = new PulvinarPlayerRepository(connection);
-        PlayerModule playerModule = new InternalPlayerModule(playerRepository);
-        log.debug("The player module has been initialized.");
+        builder.rankRepository(new InternalRankRepository(dataSource));
+        builder.rankSubscriptionRepository(new InternalRankSubscriptionRepository(dataSource));
+
+        builder.playerRepository(new PulvinarPlayerRepository(dataSource));
+        builder.operatorRepository(new PulvinarOperatorRepository(dataSource));
 
         log.debug("Proceeding to populate core instance...");
-        startCore(punishmentModule, rankModule, playerModule);
+        databaseModule(builder.build());
+        init();
+    }
 
+    public @NonNull PlayerProperties properties(@NonNull KissenPlugin plugin, @NonNull PlayerClient playerClient)
+    {
+        DataSource dataSource = provider.dataSource().orElseThrow(() -> new IllegalStateException("The provider's dataSource is not present."));
+        PropertyRepository repository = playerRepositories.computeIfAbsent(plugin, (ignored) -> new PropertyRepository(plugin, dataSource));
+        return repository.find(playerClient).join();
     }
 
-    public @NotNull PulvinarConfig config() {
+    public @NonNull PulvinarConfig config() {
         return config;
     }
+
+    @Override public ConsoleClient console()
+    {
+        return Bukkit.getConsoleSender();
+    }
+
+    @Override
+    public @NonNull ConnectionProvider connectionProvider() {
+        return provider;
+    }
 }
diff --git a/src/main/java/net/kissenpvp/network/actor/ActorData.java b/src/main/java/net/kissenpvp/network/actor/ActorData.java
new file mode 100644
index 0000000000000000000000000000000000000000..eb8a8fba85175befa220c55a487cef201eb09e54
--- /dev/null
+++ b/src/main/java/net/kissenpvp/network/actor/ActorData.java
@@ -0,0 +1,35 @@
+package net.kissenpvp.network.actor;
+
+import net.kissenpvp.api.network.actor.rank.Rank;
+import net.kissenpvp.api.temporal.timespan.DefinedTimeSpan;
+import net.kissenpvp.temporal.timespan.InternalDefinedTimeSpan;
+import org.jspecify.annotations.NonNull;
+
+import java.time.Duration;
+import java.time.Instant;
+import java.time.temporal.ChronoUnit;
+import java.util.Collections;
+import java.util.List;
+
+public class ActorData
+{
+    private final List<Rank> ranks;
+    private final DefinedTimeSpan definedTimeSpan;
+
+    public ActorData(@NonNull List<Rank> ranks, @NonNull DefinedTimeSpan definedTimeSpan)
+    {
+        this.ranks = ranks;
+        this.definedTimeSpan = definedTimeSpan;
+    }
+
+    public @NonNull List<Rank> ranks()
+    {
+        return Collections.unmodifiableList(ranks);
+    }
+
+    public @NonNull DefinedTimeSpan calculateTimePlayed(@NonNull Instant lastJoin)
+    {
+        Duration totalTimePlayed = Duration.between(lastJoin, Instant.now()).plus(Duration.from(definedTimeSpan));
+        return new InternalDefinedTimeSpan(totalTimePlayed.get(ChronoUnit.MILLIS));
+    }
+}
diff --git a/src/main/java/net/kissenpvp/network/actor/PlayerMock.java b/src/main/java/net/kissenpvp/network/actor/PlayerMock.java
new file mode 100644
index 0000000000000000000000000000000000000000..b12177fdb282965673c7273d020d4b89841abf71
--- /dev/null
+++ b/src/main/java/net/kissenpvp/network/actor/PlayerMock.java
@@ -0,0 +1,82 @@
+package net.kissenpvp.network.actor;
+
+import com.mojang.authlib.GameProfile;
+import net.kissenpvp.api.base.KissenPlugin;
+import net.kissenpvp.api.network.actor.PlayerProperties;
+import net.kissenpvp.api.punishment.Punishment;
+import net.kissenpvp.api.punishment.PunishmentSubscription;
+import net.kissenpvp.api.temporal.timespan.DefinedTimeSpan;
+import net.kissenpvp.temporal.timespan.InternalDefinedTimeSpan;
+import net.kyori.adventure.text.Component;
+import org.jspecify.annotations.NonNull;
+import org.jspecify.annotations.Nullable;
+
+import java.time.Instant;
+import java.util.Locale;
+import java.util.Objects;
+import java.util.UUID;
+
+public class PlayerMock extends InternalPlayerClient
+{
+    private final UUID id;
+    private final String username;
+
+    public static @NonNull PlayerMock of(@NonNull GameProfile gameProfile)
+    {
+        Objects.requireNonNull(gameProfile, "GameProfile cannot be null");
+        return new PlayerMock(gameProfile.getId(), gameProfile.getName());
+    }
+
+    private PlayerMock(@NonNull UUID id, @NonNull String username)
+    {
+        super(id);
+        this.id = id;
+        this.username = username;
+    }
+
+    @Override public @NonNull UUID getUniqueId()
+    {
+        return id;
+    }
+
+    @Override public long getLastLogin()
+    {
+        return Instant.now().getEpochSecond();
+    }
+
+    @Override public @NonNull DefinedTimeSpan timePlayed()
+    {
+        return new InternalDefinedTimeSpan(0L);
+    }
+
+    @Override public @NonNull PunishmentSubscription punish(@NonNull Punishment punishment) throws NullPointerException
+    {
+        return punish(punishment, null);
+    }
+
+    @Override
+    public @NonNull PunishmentSubscription punish(@NonNull Punishment punishment, @Nullable Component component) throws NullPointerException
+    {
+        throw new UnsupportedOperationException("Cannot punish a player mock.");
+    }
+
+    @Override
+    public @NonNull PlayerProperties persistentProperties(@NonNull KissenPlugin kissenPlugin) throws NullPointerException {
+        throw new UnsupportedOperationException("Cannot punish a player mock.");
+    }
+
+    @Override public @NonNull String username()
+    {
+        return username;
+    }
+
+    @Override public @NonNull Locale locale()
+    {
+        return Locale.US;
+    }
+
+    @Override public boolean isOp()
+    {
+        return false;
+    }
+}
diff --git a/src/main/java/net/kissenpvp/network/actor/PulvinarPlayerRepository.java b/src/main/java/net/kissenpvp/network/actor/PulvinarPlayerRepository.java
index 196774df5591be3691c89340bba0fe7a0fa91b10..8dae019c35775d1955dbb5f47f32bb3b48daf45b 100644
--- a/src/main/java/net/kissenpvp/network/actor/PulvinarPlayerRepository.java
+++ b/src/main/java/net/kissenpvp/network/actor/PulvinarPlayerRepository.java
@@ -3,11 +3,9 @@ package net.kissenpvp.network.actor;
 import com.mojang.authlib.GameProfile;
 import net.kissenpvp.api.network.actor.PlayerClient;
 import org.bukkit.Bukkit;
-import org.jetbrains.annotations.NotNull;
-import org.jetbrains.annotations.Nullable;
-import org.jetbrains.annotations.UnmodifiableView;
+import org.jspecify.annotations.NonNull;
 
-import java.sql.Connection;
+import javax.sql.DataSource;
 import java.sql.ResultSet;
 import java.sql.SQLException;
 import java.util.*;
@@ -15,19 +13,20 @@ import java.util.concurrent.CompletableFuture;
 
 public class PulvinarPlayerRepository extends InternalPlayerRepository {
 
-    public PulvinarPlayerRepository(@NotNull Connection connection) throws NullPointerException {
-        super(connection);
+    public PulvinarPlayerRepository(@NonNull DataSource dataSource) throws NullPointerException {
+        super(dataSource);
     }
 
-    public @NotNull CompletableFuture<@NotNull @UnmodifiableView Collection<GameProfile>> findAllProfiles()
+    public @NonNull CompletableFuture<@NonNull Collection<GameProfile>> findAllProfiles()
     {
-        return CompletableFuture.supplyAsync(() -> query("SELECT id, username FROM %s;", (statement) ->
+        String sql = "SELECT id, username FROM ksvp_player;";
+        return CompletableFuture.supplyAsync(() -> Objects.requireNonNull(query(sql, (statement) ->
         {
             Collection<GameProfile> clients = new ArrayList<>();
 
-            try(ResultSet resultSet = statement.executeQuery())
+            try (ResultSet resultSet = statement.executeQuery())
             {
-                while(resultSet.next())
+                while (resultSet.next())
                 {
                     clients.add(new GameProfile(
                             UUID.fromString(resultSet.getString("id")),
@@ -37,72 +36,18 @@ public class PulvinarPlayerRepository extends InternalPlayerRepository {
             }
 
             return Collections.unmodifiableCollection(clients);
-        }));
+        })));
     }
 
     @Override
-    public @NotNull CompletableFuture<@Nullable PlayerClient> findByName(@NotNull String name, boolean utilizeCache) throws NullPointerException {
-        String sql = "SELECT * FROM %s WHERE username = ?;";
-        return CompletableFuture.supplyAsync(() -> query(sql, (statement) ->
-        {
-            statement.setString(1, name);
-
-            try(ResultSet resultSet = statement.executeQuery())
-            {
-                if(!resultSet.next())
-                {
-                    return null;
-                }
-
-                return toEntity(resultSet);
-            }
-        }));
-    }
-
-    @Override
-    public @NotNull CompletableFuture<@UnmodifiableView Collection<PlayerClient>> findAllByName(@NotNull Iterable<String> iterable, boolean utilizeCache) throws NullPointerException {
-
-        int size = 0;
-        for (Object ignored : iterable) { size++; }
-
-        if(size == 0)
-        {
-            return CompletableFuture.completedFuture(Collections.emptyList());
-        }
-
-        String dynamicPlaceholders = String.join(", ", Collections.nCopies(size, "?"));
-        String sql = "SELECT * FROM %s WHERE username IN (" + dynamicPlaceholders + ");";
-
-        return CompletableFuture.supplyAsync(() -> query(sql, (statement) ->
-        {
-            int current = 0;
-            for (String xx : iterable) {
-                statement.setString(current + 1, xx);
-                current++;
-            }
-
-            List<PlayerClient> clients = new ArrayList<>();
-            try(ResultSet resultSet = statement.executeQuery())
-            {
-                while(resultSet.next())
-                {
-                    clients.add(toEntity(resultSet));
-                }
-            }
-
-            return Collections.unmodifiableList(clients);
-        }));
-    }
-
-    @Override
-    protected @NotNull PlayerClient toCachedEntity(@NotNull UUID uuid, @NotNull ResultSet resultSet) throws SQLException, NullPointerException {
+    protected @NonNull PlayerClient toCachedEntity(@NonNull UUID uuid, @NonNull ResultSet resultSet) throws SQLException, NullPointerException {
         PlayerClient playerClient = Bukkit.getOfflinePlayer(uuid);
         // TODO: Set data if necessary
         return playerClient; // will automatically adapt reachable player if online
     }
 
     @Override
-    protected @NotNull PlayerClient toCachedEntity(@NotNull ResultSet resultSet) throws SQLException, NullPointerException {
+    protected @NonNull PlayerClient toCachedEntity(@NonNull ResultSet resultSet) throws SQLException, NullPointerException {
         return toCachedEntity(UUID.fromString(resultSet.getString("id")), resultSet);
     }
 }
diff --git a/src/main/java/net/kissenpvp/network/actor/PulvinarProfileCache.java b/src/main/java/net/kissenpvp/network/actor/PulvinarProfileCache.java
index 9cd8f429b25647168d3141d78afe67f5f28068c4..9d7b254755bf3589e752eac238de2a2b6bfff846 100644
--- a/src/main/java/net/kissenpvp/network/actor/PulvinarProfileCache.java
+++ b/src/main/java/net/kissenpvp/network/actor/PulvinarProfileCache.java
@@ -1,31 +1,68 @@
 package net.kissenpvp.network.actor;
 
 import com.destroystokyo.paper.profile.PaperGameProfileRepository;
-import com.google.common.collect.Lists;
 import com.mojang.authlib.GameProfile;
 import com.mojang.authlib.GameProfileRepository;
-import net.kissenpvp.api.network.actor.PlayerClient;
-import net.minecraft.Util;
 import net.minecraft.server.players.GameProfileCache;
-import org.jetbrains.annotations.Contract;
-import org.jetbrains.annotations.NotNull;
-import org.jetbrains.annotations.Nullable;
+import org.jspecify.annotations.NonNull;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
 
-import java.time.Duration;
-import java.time.Instant;
-import java.time.temporal.Temporal;
+import java.io.File;
 import java.util.*;
-import java.util.concurrent.CompletableFuture;
-import java.util.function.BiConsumer;
 
 public class PulvinarProfileCache extends GameProfileCache
 {
-    public PulvinarProfileCache(@NotNull GameProfileRepository gameProfileRepository)
+    private static final Logger log = LoggerFactory.getLogger(PulvinarProfileCache.class);
+
+    /**
+     * Creates a new {@code PulvinarProfileCache} that manages player profile caching
+     * through the specified {@link GameProfileRepository} without using a cache file.
+     * <p>
+     * This constructor requires that the given repository is an instance of
+     * {@link PaperGameProfileRepository} backed by a {@link PulvinarPlayerRepository}.
+     * In such cases, profiles are stored directly in the database, and no file-based
+     * cache is used.
+     *
+     * @param gameProfileRepository the repository for managing game profiles; must not be {@code null}
+     * @throws IllegalArgumentException if the repository is not a {@code PaperGameProfileRepository}
+     *                                  backed by a {@code PulvinarPlayerRepository}
+     */
+    public PulvinarProfileCache(@NonNull GameProfileRepository gameProfileRepository) {
+        this(gameProfileRepository, null);
+    }
+
+    /**
+     * Creates a new {@code PulvinarProfileCache} that manages player profile caching
+     * through the specified {@link GameProfileRepository}, optionally using a cache file.
+     * <p>
+     * If a {@code file} is provided, profiles may be cached to disk as well as
+     * retrieved from the repository. If {@code file} is {@code null}, database-backed
+     * persistence is expected, in which case the repository must fulfill the same
+     * requirements enforced by the single-argument constructor.
+     *
+     * @param gameProfileRepository the repository for managing game profiles; must not be {@code null}
+     * @param file                  the cache file to use for storing profile data locally, or {@code null}
+     */
+    public PulvinarProfileCache(@NonNull GameProfileRepository gameProfileRepository, File file) {
+        super(gameProfileRepository, file);
+    }
+
+    @Override public void add(@NonNull GameProfile gameProfile)
     {
-        super(gameProfileRepository, null);
+        if(
+                getRepository() instanceof PaperGameProfileRepository paper &&
+                paper.playerRepository() instanceof PulvinarPlayerRepository playerRepository &&
+                get(gameProfile.getId()).isEmpty()
+        ) {
+            log.info("Profile {} has not been cached yet, saving to database...", gameProfile.getName());
+            playerRepository.save(PlayerMock.of(gameProfile)).join();
+        }
+
+        super.add(gameProfile);
     }
 
-    @SuppressWarnings("ClassEscapesDefinedScope") @Override public @NotNull List<GameProfileInfo> load()
+    @SuppressWarnings("ClassEscapesDefinedScope") @Override public @NonNull List<GameProfileInfo> load()
     {
         if(getRepository() instanceof PaperGameProfileRepository paper && paper.playerRepository() instanceof PulvinarPlayerRepository playerRepository)
         {
@@ -40,11 +77,18 @@ public class PulvinarProfileCache extends GameProfileCache
             }).toList();
         }
 
+
         return super.load();
     }
 
     @Override public void save(boolean asyncSave)
     {
-        // this can be skipped, profiles are automatically saved when they are updated
+        if(getRepository() instanceof PaperGameProfileRepository paper && paper.playerRepository() instanceof PulvinarPlayerRepository)
+        {
+            // this can be skipped, profiles are automatically saved when they are updated
+            return;
+        }
+
+        super.save(asyncSave);
     }
 }
diff --git a/src/main/java/net/kissenpvp/network/actor/operator/PulvinarOperatorList.java b/src/main/java/net/kissenpvp/network/actor/operator/PulvinarOperatorList.java
new file mode 100644
index 0000000000000000000000000000000000000000..44f61d0544dfd82b1ececa5919b89e6e352cf79b
--- /dev/null
+++ b/src/main/java/net/kissenpvp/network/actor/operator/PulvinarOperatorList.java
@@ -0,0 +1,59 @@
+package net.kissenpvp.network.actor.operator;
+
+import net.kissenpvp.api.database.Repository;
+import net.kissenpvp.api.network.actor.OperatorInfo;
+import net.minecraft.server.players.ServerOpList;
+import net.minecraft.server.players.ServerOpListEntry;
+import org.jspecify.annotations.NonNull;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.IOException;
+import java.util.Collection;
+import java.util.Objects;
+import java.util.UUID;
+
+public class PulvinarOperatorList extends ServerOpList
+{
+    private static final Logger log = LoggerFactory.getLogger(PulvinarOperatorList.class);
+    private final Repository<UUID, OperatorInfo> repository;
+
+    public PulvinarOperatorList(@NonNull Repository<UUID, OperatorInfo> repository)
+    {
+        super(null);
+        this.repository = repository;
+    }
+
+    @Override public void add(@NonNull ServerOpListEntry entry)
+    {
+        if(Objects.isNull(entry.getUser()))
+        {
+            log.warn("Discarding adding operator entry {} due to missing user.", entry);
+            return;
+        }
+
+        getMap().put(getKeyForUser(entry.getUser()), entry);
+        repository.save(entry).join();
+    }
+
+    @Override public void load() throws IOException
+    {
+        Collection<OperatorInfo> operators = repository.findAll().join();
+
+        for(OperatorInfo operator : operators)
+        {
+            ServerOpListEntry entry = (ServerOpListEntry) operator;
+            if(Objects.isNull(entry.getUser()))
+            {
+                log.warn("Discarding operator entry {} due to missing user.", operator);
+                continue;
+            }
+            getMap().put(getKeyForUser(entry.getUser()), entry);
+        }
+    }
+
+    @Override public void save() throws IOException
+    {
+        // saving is handled by the repository
+    }
+}
diff --git a/src/main/java/net/kissenpvp/network/actor/operator/PulvinarOperatorRepository.java b/src/main/java/net/kissenpvp/network/actor/operator/PulvinarOperatorRepository.java
new file mode 100644
index 0000000000000000000000000000000000000000..cb8d70fc070731276ca144df37586ffd3afdcc0d
--- /dev/null
+++ b/src/main/java/net/kissenpvp/network/actor/operator/PulvinarOperatorRepository.java
@@ -0,0 +1,30 @@
+package net.kissenpvp.network.actor.operator;
+
+import com.mojang.authlib.GameProfile;
+import net.kissenpvp.api.network.actor.OperatorInfo;
+import net.kissenpvp.network.actor.InternalOperatorRepository;
+import net.minecraft.server.players.ServerOpListEntry;
+import org.jspecify.annotations.NonNull;
+
+import javax.sql.DataSource;
+import java.sql.ResultSet;
+import java.sql.SQLException;
+import java.util.UUID;
+
+public class PulvinarOperatorRepository extends InternalOperatorRepository
+{
+    public PulvinarOperatorRepository(@NonNull DataSource dataSource) throws NullPointerException
+    {
+        super(dataSource);
+    }
+
+    @Override
+    public @NonNull OperatorInfo toEntity(@NonNull UUID uuid, @NonNull ResultSet resultSet) throws SQLException, NullPointerException
+    {
+        GameProfile profile = new GameProfile(uuid, resultSet.getString("username"));
+        int level = resultSet.getInt("operator_level");
+        boolean canBypass = resultSet.getBoolean("can_bypass_player_limit");
+
+        return new ServerOpListEntry(profile, level, canBypass);
+    }
+}
diff --git a/src/main/java/net/kissenpvp/punishment/PunishmentEntry.java b/src/main/java/net/kissenpvp/punishment/PunishmentEntry.java
new file mode 100644
index 0000000000000000000000000000000000000000..79ca61c8964825fee2476a27001d476330a866cf
--- /dev/null
+++ b/src/main/java/net/kissenpvp/punishment/PunishmentEntry.java
@@ -0,0 +1,81 @@
+package net.kissenpvp.punishment;
+
+import com.mojang.authlib.GameProfile;
+import net.kissenpvp.api.network.actor.PlayerClient;
+import net.kissenpvp.api.punishment.PunishmentSubscription;
+import net.kyori.adventure.text.serializer.plain.PlainTextComponentSerializer;
+import net.minecraft.server.players.UserBanListEntry;
+import org.jspecify.annotations.NonNull;
+import org.jspecify.annotations.Nullable;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.util.Date;
+import java.util.Objects;
+
+public class PunishmentEntry extends UserBanListEntry
+{
+    private static final Logger log = LoggerFactory.getLogger(PunishmentEntry.class);
+    private PunishmentSubscription punishment;
+
+    public static @NonNull PunishmentEntry from(@NonNull PunishmentSubscription subscription)
+    {
+        return new PunishmentEntry(subscription);
+    }
+
+    private PunishmentEntry(@NonNull PunishmentSubscription punishment)
+    {
+        super((GameProfile) null);
+        this.punishment = punishment;
+    }
+
+    @Override public @NonNull Date getCreated()
+    {
+        return Date.from(getPunishment().temporal().start());
+    }
+
+    @Override public @NonNull String getSource()
+    {
+        return getPunishment().operator().username();
+    }
+
+    @Override public @Nullable Date getExpires()
+    {
+        // convert expiry from instant to date if exists. If it doesn't exist, return null'
+        // null means a permanent ban.
+        return getPunishment().temporal().expiry().map(Date::from).orElse(null);
+    }
+
+    @Override public @NonNull String getReason()
+    {
+        PlainTextComponentSerializer serializer = PlainTextComponentSerializer.plainText();
+        return getPunishment().message().map(serializer::serialize).orElse("Banned by an operator.");
+    }
+
+    @Override public @Nullable GameProfile getUser() throws IllegalStateException
+    {
+        PlayerClient playerClient = getPunishment().target();
+        return new GameProfile(playerClient.id(), playerClient.username());
+    }
+
+    public @NonNull PunishmentSubscription getPunishment()
+    {
+        return punishment;
+    }
+
+    public void update(@NonNull PunishmentSubscription subscription)
+    {
+        if(!Objects.equals(subscription, punishment))
+        {
+            throw new IllegalArgumentException("Updated subscription must be the same as the original.");
+        }
+
+        if(Objects.equals(subscription.signature(), punishment.signature()))
+        {
+            log.warn("Updated subscription {} has the same signature as the original. Skipping update.", subscription);
+            return;
+        }
+
+        punishment = subscription;
+    }
+}
diff --git a/src/main/java/net/kissenpvp/punishment/PunishmentList.java b/src/main/java/net/kissenpvp/punishment/PunishmentList.java
new file mode 100644
index 0000000000000000000000000000000000000000..fa84c23d12a84082d11019c3e25c465aa256ecac
--- /dev/null
+++ b/src/main/java/net/kissenpvp/punishment/PunishmentList.java
@@ -0,0 +1,128 @@
+package net.kissenpvp.punishment;
+
+import com.mojang.authlib.GameProfile;
+import net.kissenpvp.api.database.RepositoryHolder;
+import net.kissenpvp.api.network.actor.PlayerClient;
+import net.kissenpvp.api.punishment.Punishment;
+import net.kissenpvp.api.punishment.PunishmentSubscription;
+import net.kissenpvp.api.punishment.PunishmentSubscriptionRepository;
+import net.kissenpvp.api.punishment.PunishmentType;
+import net.kissenpvp.api.temporal.WritableTemporalObject;
+import net.kissenpvp.temporal.InternalWritableTemporalObject;
+import net.kissenpvp.temporal.timespan.PermanentTimeSpan;
+import net.kyori.adventure.text.Component;
+import net.minecraft.server.players.UserBanList;
+import net.minecraft.server.players.UserBanListEntry;
+import org.bukkit.Bukkit;
+import org.jspecify.annotations.NonNull;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.IOException;
+import java.util.*;
+import java.util.stream.Collector;
+import java.util.stream.Collectors;
+
+public class PunishmentList extends UserBanList
+{
+    private static final Logger log = LoggerFactory.getLogger(PunishmentList.class);
+    private final PunishmentSubscriptionRepository repository;
+
+    private static @NonNull PunishmentSubscription convertLegacy(@NonNull UserBanListEntry entry)
+    {
+        if (Objects.isNull(entry.getUser()))
+        {
+            throw new IllegalArgumentException("Cannot create punishment entry from null user.");
+        }
+
+        UUID linkId = Bukkit.getOfflinePlayer(entry.getUser().getId()).linkId();
+        UUID operator = null; // TODO: determine source by the stupid vanilla string
+        WritableTemporalObject temporalObject = InternalWritableTemporalObject.toTemporal(new PermanentTimeSpan());
+
+        if (!((RepositoryHolder) Bukkit.getPulvinar()).punishmentRepository().has(-1).join())
+        {
+            //TODO prohibit creation of negative id's as they are reserved for internal use.
+            Punishment punishment = new InternalPunishment(-1, PunishmentType.BAN, new PermanentTimeSpan());
+            ((RepositoryHolder) Bukkit.getPulvinar()).punishmentRepository().save(punishment).join();
+        }
+
+        return new InternalPunishmentSubscription(-1, linkId, operator, temporalObject, Component.text(entry.getReason()));
+    }
+
+    public PunishmentList(@NonNull PunishmentSubscriptionRepository repository)
+    {
+        //noinspection DataFlowIssue we're not using files anymore. We're beyond that
+        super(null);
+        this.repository = repository;
+    }
+
+    @Override public void add(@NonNull UserBanListEntry entry)
+    {
+        GameProfile profile = entry.getUser(); assert profile != null;
+
+        if (entry instanceof PunishmentEntry punishmentEntry)
+        {
+            getMap().put(this.getKeyForUser(profile), entry);
+            getRepository().save(punishmentEntry.getPunishment()).join();
+            return;
+        }
+
+        log.warn("Received legacy ban entry {}. Converting to PunishmentEntry...", entry);
+        PunishmentEntry punishmentEntry = PunishmentEntry.from(convertLegacy(entry));
+
+        // will never end up in a recursion loop as we are passing a punishment entry.
+        add(punishmentEntry);
+    }
+
+    @Override public void save() throws IOException
+    {
+        // saving is handled by the repository
+    }
+
+    @Override public void remove(@NonNull GameProfile user)
+    {
+        UserBanListEntry entry = getMap().remove(getKeyForUser(user));
+        if(entry instanceof PunishmentEntry punishmentEntry)
+        {
+            PunishmentSubscription subscription = punishmentEntry.getPunishment();
+            subscription.temporal().expire();
+            getRepository().save(subscription).join();
+        }
+    }
+
+    @Override public void load() throws IOException
+    {
+        Collection<PunishmentSubscription> subscriptions = repository.findAll().join();
+        for (PunishmentSubscription subscription : subscriptions)
+        {
+            PlayerClient playerClient = subscription.target();
+            GameProfile gameProfile = new GameProfile(playerClient.id(), playerClient.username());
+            if(!subscription.temporal().expired())
+            {
+                getMap().put(getKeyForUser(gameProfile), PunishmentEntry.from(subscription));
+            }
+        }
+    }
+
+    /**
+     * Creates a collector that processes {@link PunishmentSubscription} objects by expiring their temporal context
+     * and maps them by their unique identifiers.
+     *
+     * @return a collector that collects {@link PunishmentSubscription} objects into a {@link Map}, where the keys
+     * are the identifiers of the subscriptions, and the values are the subscriptions themselves with their
+     * temporal context expired.
+     */
+    private static @NonNull Collector<PunishmentSubscription, ?, Map<String, PunishmentSubscription>> expire()
+    {
+        return Collectors.toMap(PunishmentSubscription::id, punishment ->
+        {
+            punishment.temporal().expire();
+            return punishment;
+        });
+    }
+
+    protected PunishmentSubscriptionRepository getRepository()
+    {
+        return repository;
+    }
+}
diff --git a/src/main/java/org/bukkit/craftbukkit/CraftOfflinePlayer.java b/src/main/java/org/bukkit/craftbukkit/CraftOfflinePlayer.java
index b54529d0efd4a989407423c6aa2b7e7dca2bb3a4..c1f7beeb65b8136dfb6885b02ef1aa4d9e18cc40 100644
--- a/src/main/java/org/bukkit/craftbukkit/CraftOfflinePlayer.java
+++ b/src/main/java/org/bukkit/craftbukkit/CraftOfflinePlayer.java
@@ -8,6 +8,10 @@ import java.util.Date;
 import java.util.LinkedHashMap;
 import java.util.Map;
 import java.util.UUID;
+
+import net.kissenpvp.api.punishment.Punishment;
+import net.kissenpvp.api.punishment.PunishmentSubscription;
+import net.kyori.adventure.text.Component;
 import net.minecraft.core.GlobalPos;
 import net.minecraft.nbt.CompoundTag;
 import net.minecraft.network.protocol.game.ClientboundTagQueryPacket;
@@ -35,6 +39,9 @@ import org.bukkit.configuration.serialization.SerializableAs;
 import org.bukkit.craftbukkit.util.CraftLocation;
 import org.bukkit.entity.EntityType;
 import org.bukkit.entity.Player;
+import org.jetbrains.annotations.NotNull;
+import org.jetbrains.annotations.Nullable;
+import org.jspecify.annotations.NonNull;
 
 @SerializableAs("Player")
 public class CraftOfflinePlayer implements OfflinePlayer, ConfigurationSerializable {
@@ -777,7 +784,7 @@ public class CraftOfflinePlayer implements OfflinePlayer, ConfigurationSerializa
 
     @Override public @org.jetbrains.annotations.NotNull net.kissenpvp.api.temporal.timespan.DefinedTimeSpan timePlayed()
     {
-        return null;
+        return new net.kissenpvp.temporal.timespan.InternalDefinedTimeSpan(0L);
     }
 
     @Override public @org.jetbrains.annotations.NotNull java.util.UUID id()
@@ -795,6 +802,17 @@ public class CraftOfflinePlayer implements OfflinePlayer, ConfigurationSerializa
         return false;
     }
 
+    @Override public @NotNull PunishmentSubscription punish(@NotNull Punishment punishment) throws NullPointerException
+    {
+        return punish(punishment, null);
+    }
+
+    @Override
+    public @NotNull PunishmentSubscription punish(@NotNull Punishment punishment, @Nullable Component component) throws NullPointerException
+    {
+        return null;
+    }
+
     @Override public @org.jetbrains.annotations.NotNull String username()
     {
         String userName = getName();
@@ -811,6 +829,11 @@ public class CraftOfflinePlayer implements OfflinePlayer, ConfigurationSerializa
         return java.util.Locale.ENGLISH;
     }
 
+    @Override
+    public net.kissenpvp.api.network.actor.@NonNull PlayerProperties persistentProperties(net.kissenpvp.api.base.@NonNull KissenPlugin plugin) {
+        return org.bukkit.Bukkit.getPluginManager().playerProperties((org.bukkit.plugin.Plugin) plugin, this);
+    }
+
     @Override public final boolean isClient()
     {
         return true;
diff --git a/src/main/java/org/bukkit/craftbukkit/command/CraftConsoleCommandSender.java b/src/main/java/org/bukkit/craftbukkit/command/CraftConsoleCommandSender.java
index 15ae2653c1809041b7ab7cb67015e82c226e9f77..527dcaf194501c144ca2ffc11377782e501ea9d4 100644
--- a/src/main/java/org/bukkit/craftbukkit/command/CraftConsoleCommandSender.java
+++ b/src/main/java/org/bukkit/craftbukkit/command/CraftConsoleCommandSender.java
@@ -105,4 +105,26 @@ public class CraftConsoleCommandSender extends ServerCommandSender implements Co
     public boolean hasPermission(org.bukkit.permissions.Permission perm) {
         return io.papermc.paper.configuration.GlobalConfiguration.get().console.hasAllPermissions || super.hasPermission(perm);
     }
+
+    // Pulvinar start - implement kissen command sender methods
+    @Override public @org.jetbrains.annotations.NotNull String username()
+    {
+        return getName();
+    }
+
+    @Override public @org.jetbrains.annotations.NotNull java.util.Locale locale()
+    {
+        return java.util.Locale.ROOT;
+    }
+
+    @Override public boolean isClient()
+    {
+        return false;
+    }
+
+    @Override public @org.jetbrains.annotations.NotNull net.kyori.adventure.audience.Audience audience()
+    {
+        return this;
+    }
+    // Pulvinar end - implement kissen command sender methods
 }
diff --git a/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java b/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java
index afb3df83b717578bdf672df8a1c0b80392ffe341..fec145efefa287b1b76fbf3811cc296056d21f3b 100644
--- a/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java
+++ b/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java
@@ -3625,7 +3625,7 @@ public class CraftPlayer extends CraftHumanEntity implements Player, PluginMessa
     }
 
     // Pulvinar start - add player actor methods
-    @Override public @org.jetbrains.annotations.NotNull net.kyori.adventure.audience.Audience audience()
+    @Override public net.kyori.adventure.audience.@NonNull Audience audience()
     {
         return this;
     }
@@ -3640,12 +3640,28 @@ public class CraftPlayer extends CraftHumanEntity implements Player, PluginMessa
         return java.time.Instant.ofEpochMilli(getLastLogin());
     }
 
-    @Override public @org.jetbrains.annotations.NotNull net.kissenpvp.api.temporal.timespan.DefinedTimeSpan timePlayed()
+    @Override public net.kissenpvp.api.temporal.timespan.@NonNull DefinedTimeSpan timePlayed()
     {
         // TODO - implement online time
         return new net.kissenpvp.temporal.timespan.InternalDefinedTimeSpan(0L);
     }
 
+    @Override public net.kissenpvp.api.punishment.@NonNull PunishmentSubscription punish(@org.jetbrains.annotations.NotNull net.kissenpvp.api.punishment.Punishment punishment) throws NullPointerException
+    {
+        return punish(punishment, null);
+    }
+
+    @Override
+    public net.kissenpvp.api.punishment.@NonNull PunishmentSubscription punish(@org.jetbrains.annotations.NotNull net.kissenpvp.api.punishment.Punishment punishment, @org.jetbrains.annotations.Nullable net.kyori.adventure.text.Component component) throws NullPointerException
+    {
+        return null;
+    }
+
+    @Override
+    public net.kissenpvp.api.network.actor.@NonNull PlayerProperties persistentProperties(net.kissenpvp.api.base.@NonNull KissenPlugin plugin) {
+        return org.bukkit.Bukkit.getPluginManager().playerProperties((org.bukkit.plugin.Plugin) plugin, this);
+    }
+
     @Override public @NonNull UUID id()
     {
         return getUniqueId();
