From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Ivo Quiring <falk.quiring@gmail.com>
Date: Mon, 17 Nov 2025 15:51:29 +0100
Subject: [PATCH] Overhaul Player Database


diff --git a/src/main/java/com/destroystokyo/paper/profile/PaperAuthenticationService.java b/src/main/java/com/destroystokyo/paper/profile/PaperAuthenticationService.java
index 48e774677edf17d4a99ae9ed23d1b371dab41abb..8b002f246d17c0e00f9b8ca1405e6f1254b9ba54 100644
--- a/src/main/java/com/destroystokyo/paper/profile/PaperAuthenticationService.java
+++ b/src/main/java/com/destroystokyo/paper/profile/PaperAuthenticationService.java
@@ -25,6 +25,13 @@ public class PaperAuthenticationService extends YggdrasilAuthenticationService {
 
     @Override
     public GameProfileRepository createProfileRepository() {
-        return new PaperGameProfileRepository(this.getProxy(), this.environment);
+        // Pulvinar start - overhaul player database
+        //return new PaperGameProfileRepository(this.getProxy(), this.environment);
+        return new net.kissenpvp.network.actor.PulvinarProfileRepository(
+                getProxy(),
+                environment,
+                (net.kissenpvp.network.actor.PulvinarPlayerRepository) (org.bukkit.Bukkit.getPulvinar()).playerRepository()
+        );
+        // Pulvinar end - overhaul player database
     }
 }
diff --git a/src/main/java/com/destroystokyo/paper/profile/PaperGameProfileRepository.java b/src/main/java/com/destroystokyo/paper/profile/PaperGameProfileRepository.java
index 58c1ff58e58e969a88b929663dfff785843dc1b2..f7a056fee14ed797eb284e68d1387089bbc1a425 100644
--- a/src/main/java/com/destroystokyo/paper/profile/PaperGameProfileRepository.java
+++ b/src/main/java/com/destroystokyo/paper/profile/PaperGameProfileRepository.java
@@ -31,6 +31,10 @@ public class PaperGameProfileRepository extends YggdrasilGameProfileRepository {
             }
         }
 
+        // Pulvinar start - overhaul player database
+        unfoundNames = databaseLookup(unfoundNames, callback);
+        // Pulvinar end - overhaul player database
+
         // Some things were not found.... Proceed to look up.
         if (!unfoundNames.isEmpty()) {
             String[] namesArr = unfoundNames.toArray(new String[unfoundNames.size()]);
@@ -62,4 +66,8 @@ public class PaperGameProfileRepository extends YggdrasilGameProfileRepository {
             this.callback.onProfileLookupFailed(profileName, exception);
         }
     }
+
+    // Pulvinar start - overhaul player database
+    protected Set<String> databaseLookup(Set<String> unfound, ProfileLookupCallback callback) { return unfound; }
+    // Pulvinar end - overhaul player database
 }
diff --git a/src/main/java/net/kissenpvp/base/PulvinarConfig.java b/src/main/java/net/kissenpvp/base/PulvinarConfig.java
index f334b84cb4b5576b51ef65ceb6e9d1b711d652fe..9b2191cdb5edc5d34231a84737ed8b897bb094dd 100644
--- a/src/main/java/net/kissenpvp/base/PulvinarConfig.java
+++ b/src/main/java/net/kissenpvp/base/PulvinarConfig.java
@@ -36,6 +36,8 @@ public class PulvinarConfig {
     private String databaseUser;
     private String databasePassword;
 
+    private boolean disableDatabaseStorage;
+
     public void init(@NonNull File configFile) throws IOException, InvalidConfigurationException {
         file = configFile;
         config = new YamlConfiguration();
@@ -59,12 +61,15 @@ public class PulvinarConfig {
         config.addDefault("database.user", "user");
         config.addDefault("database.password", "password");
 
+        config.addDefault("database.player.storage.disable-database", false);
+
         databaseConfigured = config.getBoolean("database.configured");
         databaseAddress = config.getString("database.url");
         databaseGlobal = config.getString("database.global.database");
         databasePrivate = config.getString("database.private.database");
         databaseUser = config.getString("database.user");
         databasePassword = config.getString("database.password");
+        disableDatabaseStorage = config.getBoolean("database.player.storage.disable-database");
 
         config.save(file);
     }
@@ -97,6 +102,8 @@ public class PulvinarConfig {
         return databasePassword;
     }
 
+    public boolean disableDatabaseStorage() { return disableDatabaseStorage; }
+
     public void databaseConfigured(boolean configured) throws IOException {
         databaseConfigured = configured;
         config.save(file);
@@ -132,4 +139,9 @@ public class PulvinarConfig {
         databasePassword = password;
         config.save(file);
     }
+
+    public void disableDatabaseStorage(boolean use) throws IOException, NullPointerException {
+        disableDatabaseStorage = use;
+        config.save(file);
+    }
 }
diff --git a/src/main/java/net/kissenpvp/base/PulvinarCore.java b/src/main/java/net/kissenpvp/base/PulvinarCore.java
index 3ebbb2b41362e4fccf577c144dba23722591a364..ccd45b754ba2768437a80f5cbab19d093b0404d3 100644
--- a/src/main/java/net/kissenpvp/base/PulvinarCore.java
+++ b/src/main/java/net/kissenpvp/base/PulvinarCore.java
@@ -3,7 +3,9 @@ package net.kissenpvp.base;
 import net.kissenpvp.api.base.Pulvinar;
 import net.kissenpvp.api.database.ConnectionProvider;
 import net.kissenpvp.api.network.actor.ConsoleClient;
+import net.kissenpvp.api.network.actor.PlayerRepository;
 import net.kissenpvp.database.mariadb.MariaDBConnectionProvider;
+import net.kissenpvp.network.actor.PulvinarPlayerRepository;
 import net.kissenpvp.network.actor.operator.PulvinarOperatorRepository;
 import net.kissenpvp.network.actor.whitelist.PulvinarWhiteListRepository;
 import org.bukkit.Bukkit;
@@ -33,6 +35,7 @@ public class PulvinarCore extends KissenCore implements Pulvinar
 
     private PulvinarOperatorRepository operatorRepository;
     private PulvinarWhiteListRepository whiteListRepository;
+    private PlayerRepository playerRepository;
 
     public PulvinarCore(@NonNull File optionsFile) throws IOException, InvalidConfigurationException, SQLException {
         config = new PulvinarConfig();
@@ -82,6 +85,7 @@ public class PulvinarCore extends KissenCore implements Pulvinar
     {
         operatorRepository = new PulvinarOperatorRepository(privateDatasource);
         whiteListRepository = new PulvinarWhiteListRepository(privateDatasource);
+        playerRepository = new PulvinarPlayerRepository(globalDatasource);
     }
 
     public void shutdown()
@@ -115,4 +119,9 @@ public class PulvinarCore extends KissenCore implements Pulvinar
     public @NonNull PulvinarWhiteListRepository whiteListRepository() {
         return whiteListRepository;
     }
+
+    @Override
+    public @NonNull PlayerRepository playerRepository() {
+        return playerRepository;
+    }
 }
diff --git a/src/main/java/net/kissenpvp/network/actor/PulvinarPlayerRepository.java b/src/main/java/net/kissenpvp/network/actor/PulvinarPlayerRepository.java
new file mode 100644
index 0000000000000000000000000000000000000000..9dda5813fe822b2374ad49cb6afaa963b010b284
--- /dev/null
+++ b/src/main/java/net/kissenpvp/network/actor/PulvinarPlayerRepository.java
@@ -0,0 +1,183 @@
+package net.kissenpvp.network.actor;
+
+import com.google.common.base.Preconditions;
+import net.kissenpvp.api.network.actor.PlayerClient;
+import net.kissenpvp.base.PulvinarCore;
+import net.kissenpvp.database.AsyncDatabaseQueue;
+import net.minecraft.server.players.NameAndId;
+import org.bukkit.Bukkit;
+import org.jspecify.annotations.NonNull;
+
+import javax.sql.DataSource;
+import javax.sql.rowset.serial.SerialBlob;
+import java.sql.Blob;
+import java.sql.ResultSet;
+import java.util.*;
+import java.util.concurrent.CompletableFuture;
+
+public class PulvinarPlayerRepository extends MariaDBPlayerRepository {
+
+    public PulvinarPlayerRepository(@NonNull DataSource dataSource) throws NullPointerException {
+        super(dataSource);
+    }
+
+    @Override
+    protected @NonNull PlayerClient toCachedEntity(@NonNull UUID uuid, @NonNull ResultSet resultSet) throws NullPointerException {
+        return Bukkit.getOfflinePlayer(uuid);
+    }
+
+    public @NonNull CompletableFuture<Collection<NameAndId>> findAllProfiles()
+    {
+        String sql = "SELECT id, username FROM ksvp_player;";
+        return CompletableFuture.supplyAsync(() -> assumeNotNull(query(sql, (statement) -> {
+            try (ResultSet resultSet = statement.executeQuery()) {
+                Collection<NameAndId> profile = new HashSet<>();
+                while(resultSet.next())
+                {
+                    UUID id = UUID.fromString(resultSet.getString("id"));
+                    String username = resultSet.getString("username");
+
+                    profile.add(new NameAndId(id, username));
+                }
+                return Collections.unmodifiableCollection(profile);
+            }
+        })));
+    }
+
+    public @NonNull Collection<NameAndId> findProfilesByNames(@NonNull String[] names) {
+        Set<NameAndId> discovered = new HashSet<>();
+        Set<String> undiscovered = new HashSet<>(Set.of(names));
+
+        for(PlayerClient cachedPlayerClient : cachedEntries().values())
+        {
+            if(!undiscovered.contains(cachedPlayerClient.username()))
+            {
+                continue;
+            }
+            undiscovered.remove(cachedPlayerClient.username());
+            discovered.add(new NameAndId(cachedPlayerClient.id(), cachedPlayerClient.username()));
+        }
+
+        if(undiscovered.isEmpty())
+        {
+            return discovered;
+        }
+
+        String placeholders = String.join(", ", Collections.nCopies(undiscovered.size(), "?"));
+        String sql = "SELECT id, username FROM ksvp_player WHERE username IN (" + placeholders + ");";
+        query(sql, (statement) -> {
+            int index = 1;
+            for(String argument : undiscovered)
+            {
+                statement.setString(index++, argument);
+            }
+
+            try(ResultSet resultSet = statement.executeQuery())
+            {
+                while(resultSet.next())
+                {
+                    UUID id = UUID.fromString(resultSet.getString("id"));
+                    String username = resultSet.getString("username");
+
+                    discovered.add(new NameAndId(id, username));
+                }
+            }
+
+            return null;
+        });
+
+        return discovered;
+    }
+
+    public @NonNull CompletableFuture<Void> generateProfileIfAbsent(@NonNull NameAndId[] nameAndId)
+    {
+        String globalDatabase = ((PulvinarCore) Bukkit.getPulvinar()).config().databaseGlobal();
+
+        String sql = "INSERT INTO %s.ksvp_identity (link_id) VALUES (?) ON DUPLICATE KEY UPDATE link_id = link_id;";
+        String sql2 = "INSERT INTO %s.ksvp_player (id, link_id, username) VALUES (?, ?, ?) ON DUPLICATE KEY UPDATE id = id, link_id = link_id, username = ?;";
+
+        return CompletableFuture.supplyAsync(() -> {
+            query(String.format(sql, globalDatabase), (statement) -> {
+                for(NameAndId player : nameAndId)
+                {
+                    if(cached(player.id()))
+                    {
+                        continue;
+                    }
+
+                    statement.setString(1, String.valueOf(player.id()));
+                    statement.addBatch();
+                }
+                statement.executeBatch();
+                return null;
+            });
+
+            query(String.format(sql2, globalDatabase), (statement) -> {
+                for(NameAndId player : nameAndId)
+                {
+                    if(cached(player.id()))
+                    {
+                        continue;
+                    }
+
+                    String id = String.valueOf(player.id());
+
+                    statement.setString(1, id);
+                    statement.setString(2, id);
+
+                    statement.setString(3, player.name());
+                    statement.setString(4, player.name());
+                    statement.addBatch();
+                }
+                statement.executeBatch();
+                return null;
+            });
+
+            return null;
+        });
+    }
+
+    /**
+     * Saves a byte array as a blob inside the database.
+     * This is done in the private part of the database.
+     *
+     * @param player the player the properties are attached to
+     * @param data the actual data
+     * @return a completable future when the task is done
+     */
+    public @NonNull CompletableFuture<Void> saveProperties(@NonNull UUID player, byte[] data)
+    {
+        Preconditions.checkNotNull(player);
+
+        String privateDatabase = ((PulvinarCore)PulvinarCore.getInstance()).config().databasePrivate();
+        String sql = "INSERT INTO " + privateDatabase + ".ksvp_player_properties (id, content) VALUES (?, ?) ON DUPLICATE KEY UPDATE content = ?;";
+        return AsyncDatabaseQueue.submitTask(() -> query(sql, statement ->
+        {
+            Blob blob = new SerialBlob(data);
+            statement.setString(1, String.valueOf(player));
+            statement.setBlob(2, blob);
+            statement.setBlob(3, blob);
+            statement.executeUpdate();
+            return null;
+        }));
+    }
+
+    public @NonNull CompletableFuture<Optional<Blob>> findProperties(@NonNull UUID uuid) {
+        Preconditions.checkNotNull(uuid, "The uuid cannot be null!");
+        String privateDatabase = ((PulvinarCore)PulvinarCore.getInstance()).config().databasePrivate();
+
+        String sql = "SELECT content FROM " + privateDatabase + ".ksvp_player_properties WHERE id = ?;";
+        return AsyncDatabaseQueue.submitTask(() -> this.query(sql, (statement) -> {
+            statement.setString(1, String.valueOf(uuid));
+
+            try (ResultSet resultSet = statement.executeQuery()) {
+                if (!resultSet.next())
+                {
+                    return Optional.empty();
+                }
+
+                return Optional.of(resultSet.getBlob("content"));
+            }
+        }));
+    }
+}
diff --git a/src/main/java/net/kissenpvp/network/actor/PulvinarProfileCache.java b/src/main/java/net/kissenpvp/network/actor/PulvinarProfileCache.java
new file mode 100644
index 0000000000000000000000000000000000000000..0953215e5e4188c1badb1e9259433326ad1d9c7d
--- /dev/null
+++ b/src/main/java/net/kissenpvp/network/actor/PulvinarProfileCache.java
@@ -0,0 +1,59 @@
+package net.kissenpvp.network.actor;
+
+import com.mojang.authlib.GameProfileRepository;
+import net.minecraft.server.players.CachedUserNameToIdResolver;
+import net.minecraft.server.players.NameAndId;
+import org.jetbrains.annotations.NotNull;
+import org.jspecify.annotations.NonNull;
+
+import java.io.File;
+import java.util.Calendar;
+import java.util.Date;
+import java.util.List;
+
+public class PulvinarProfileCache extends CachedUserNameToIdResolver {
+
+    public PulvinarProfileCache(GameProfileRepository profileRepository, File file) {
+        super(profileRepository, file);
+    }
+
+    @Override
+    protected @NonNull GameProfileInfo addInternal(@NonNull NameAndId nameAndId) {
+
+        if(getRepository() instanceof PulvinarProfileRepository profileRepository)
+        {
+            profileRepository.generateProfiles(nameAndId);
+        }
+
+        return super.addInternal(nameAndId);
+    }
+
+    @Override
+    protected @NotNull List<CachedUserNameToIdResolver.GameProfileInfo> load() {
+        if(getRepository() instanceof PulvinarProfileRepository profileRepository)
+        {
+            return profileRepository.findAllProfiles().stream().map(nameAndId ->
+            {
+                Calendar instance = Calendar.getInstance();
+                instance.setTime(new Date());
+                instance.add(Calendar.MONTH, 1);
+                Date time = instance.getTime(); //TODO maybe completely remove this
+
+                return new GameProfileInfo(nameAndId, time);
+            }).toList();
+        }
+
+        return super.load();
+    }
+
+    @Override public void save(boolean asyncSave)
+    {
+        if(getRepository() instanceof PulvinarProfileRepository)
+        {
+            // this can be skipped, profiles are automatically saved when they are updated
+            return;
+        }
+
+        super.save(asyncSave);
+    }
+}
diff --git a/src/main/java/net/kissenpvp/network/actor/PulvinarProfileRepository.java b/src/main/java/net/kissenpvp/network/actor/PulvinarProfileRepository.java
new file mode 100644
index 0000000000000000000000000000000000000000..b4301a1e85a9cbd9def40a872adcf2af8da4765c
--- /dev/null
+++ b/src/main/java/net/kissenpvp/network/actor/PulvinarProfileRepository.java
@@ -0,0 +1,45 @@
+package net.kissenpvp.network.actor;
+
+import com.destroystokyo.paper.profile.PaperGameProfileRepository;
+import com.mojang.authlib.Environment;
+import com.mojang.authlib.ProfileLookupCallback;
+import net.minecraft.server.players.NameAndId;
+import org.jspecify.annotations.NonNull;
+
+import java.net.Proxy;
+import java.util.Collection;
+import java.util.HashSet;
+import java.util.Set;
+
+public class PulvinarProfileRepository extends PaperGameProfileRepository {
+
+    private final PulvinarPlayerRepository repository;
+
+    public PulvinarProfileRepository(Proxy proxy, Environment environment, PulvinarPlayerRepository repository) {
+        super(proxy, environment);
+        this.repository = repository;
+    }
+
+    @Override
+    protected Set<String> databaseLookup(Set<String> unfound, ProfileLookupCallback callback) {
+        Set<String> stillUndiscovered = new HashSet<>(unfound);
+
+        for(NameAndId current : repository.findProfilesByNames(unfound.toArray(new String[0])))
+        {
+            stillUndiscovered.remove(current.name());
+            callback.onProfileLookupSucceeded(current.name(), current.id());
+        }
+
+        return stillUndiscovered;
+    }
+
+    public void generateProfiles(NameAndId... nameAndIds)
+    {
+        repository.generateProfileIfAbsent(nameAndIds).join();
+    }
+
+    public @NonNull Collection<NameAndId> findAllProfiles()
+    {
+        return repository.findAllProfiles().join();
+    }
+}
diff --git a/src/main/java/org/bukkit/craftbukkit/CraftOfflinePlayer.java b/src/main/java/org/bukkit/craftbukkit/CraftOfflinePlayer.java
index 7283eb950a0d1c92c451a0379d7a409d147bb84c..c644d284a93d229f2e25e0b7b49cb8102146e69c 100644
--- a/src/main/java/org/bukkit/craftbukkit/CraftOfflinePlayer.java
+++ b/src/main/java/org/bukkit/craftbukkit/CraftOfflinePlayer.java
@@ -764,4 +764,47 @@ public class CraftOfflinePlayer implements OfflinePlayer, ConfigurationSerializa
         }
     }
     // Purpur end - OfflinePlayer API
+
+    // Pulvinar start - overhaul player database
+    @Override
+    public @org.jspecify.annotations.NonNull UUID id() {
+        return getUniqueId();
+    }
+
+    @Override
+    public @org.jspecify.annotations.NonNull UUID linkId() {
+        return getUniqueId(); // TODO
+    }
+
+    @Override
+    public net.kissenpvp.api.punishment.@org.jspecify.annotations.NonNull PunishmentSubscription punish(net.kissenpvp.api.punishment.@org.jspecify.annotations.NonNull Punishment punishment) {
+        return null;
+    }
+
+    @Override
+    public net.kissenpvp.api.punishment.@org.jspecify.annotations.NonNull PunishmentSubscription punish(net.kissenpvp.api.punishment.@org.jspecify.annotations.NonNull Punishment punishment, net.kyori.adventure.text.@org.jspecify.annotations.Nullable Component component) {
+        return null;
+    }
+
+    @Override
+    public java.util.@org.jspecify.annotations.NonNull List<net.kissenpvp.api.network.actor.rank.RankSubscription> rankHistory() {
+        return java.util.List.of();
+    }
+
+    @Override
+    public net.kissenpvp.api.network.actor.rank.@org.jspecify.annotations.NonNull RankSubscription rank() {
+        return null;
+    }
+
+    @Override
+    public @org.jspecify.annotations.NonNull String username() {
+        java.util.Objects.requireNonNull(getName(), "Username cannot be null.");
+        return getName();
+    }
+
+    @Override
+    public final boolean isClient() {
+        return false;
+    }
+    // Pulvinar end - overhaul player database
 }
diff --git a/src/main/java/org/bukkit/craftbukkit/command/CraftConsoleCommandSender.java b/src/main/java/org/bukkit/craftbukkit/command/CraftConsoleCommandSender.java
index 15ae2653c1809041b7ab7cb67015e82c226e9f77..1f26e86d5d2ce8f46bd09ce70b4d3de1078af104 100644
--- a/src/main/java/org/bukkit/craftbukkit/command/CraftConsoleCommandSender.java
+++ b/src/main/java/org/bukkit/craftbukkit/command/CraftConsoleCommandSender.java
@@ -105,4 +105,21 @@ public class CraftConsoleCommandSender extends ServerCommandSender implements Co
     public boolean hasPermission(org.bukkit.permissions.Permission perm) {
         return io.papermc.paper.configuration.GlobalConfiguration.get().console.hasAllPermissions || super.hasPermission(perm);
     }
+
+    // Pulvinar start - overhaul player database
+    @Override
+    public @org.jetbrains.annotations.NotNull String username() {
+        return getName();
+    }
+
+    @Override
+    public boolean isClient() {
+        return false;
+    }
+
+    @Override
+    public net.kyori.adventure.audience.@org.jetbrains.annotations.NotNull Audience audience() {
+        return this;
+    }
+    // Pulvinar end - overhaul player database
 }
diff --git a/src/main/java/org/bukkit/craftbukkit/entity/CraftHumanEntity.java b/src/main/java/org/bukkit/craftbukkit/entity/CraftHumanEntity.java
index b9b4114bfa8db1cac63a43b979a0a58f872ab018..33556af01e28df2ceb6e7a16408adda02bcfe4ac 100644
--- a/src/main/java/org/bukkit/craftbukkit/entity/CraftHumanEntity.java
+++ b/src/main/java/org/bukkit/craftbukkit/entity/CraftHumanEntity.java
@@ -920,4 +920,16 @@ public class CraftHumanEntity extends CraftLivingEntity implements HumanEntity {
     public org.bukkit.entity.Entity copy(Location location) {
         throw new UnsupportedOperationException("Cannot copy human entities");
     }
+
+    // Pulvinar start - overhaul player database
+    @Override
+    public @org.jspecify.annotations.NonNull String username() {
+        return getName();
+    }
+
+    @Override
+    public final boolean isClient() {
+        return true;
+    }
+    // Pulvinar end - overhaul player database
 }
diff --git a/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java b/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java
index cc273d30cbec5cf9ae861d18be4d14073a052d06..800c0fa06fa9edec7322f5fe8fef0f69bb37267f 100644
--- a/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java
+++ b/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java
@@ -3468,4 +3468,41 @@ public class CraftPlayer extends CraftHumanEntity implements Player, PluginMessa
         super.knockback(strength, directionX, directionZ);
         this.entity.hurtMarked = true;
     }
+
+    // Pulvinar start - overhaul player database
+    @Override
+    public @NonNull UUID id() {
+        return getUniqueId();
+    }
+
+    @Override
+    public net.kyori.adventure.audience.@NonNull Audience audience() {
+        return this;
+    }
+
+    @Override
+    public @NonNull UUID linkId() {
+        return getUniqueId(); // TODO
+    }
+
+    @Override
+    public net.kissenpvp.api.punishment.@NonNull PunishmentSubscription punish(net.kissenpvp.api.punishment.@NonNull Punishment punishment) {
+        return null;
+    }
+
+    @Override
+    public net.kissenpvp.api.punishment.@NonNull PunishmentSubscription punish(net.kissenpvp.api.punishment.@NonNull Punishment punishment, net.kyori.adventure.text.@Nullable Component component) {
+        return null;
+    }
+
+    @Override
+    public @NonNull List<net.kissenpvp.api.network.actor.rank.RankSubscription> rankHistory() {
+        return List.of();
+    }
+
+    @Override
+    public net.kissenpvp.api.network.actor.rank.@NonNull RankSubscription rank() {
+        return null;
+    }
+    // Pulvinar end - overhaul player database
 }
diff --git a/src/main/resources/migrations/private/mariadb/V1_0__Intial_migration.sql b/src/main/resources/migrations/private/mariadb/V1_0__Intial_migration.sql
index 85931a5ef70790bb8e81df43014947a605344c48..ef2d58765f1ac5ee62426273d3f77ce492fbca72 100644
--- a/src/main/resources/migrations/private/mariadb/V1_0__Intial_migration.sql
+++ b/src/main/resources/migrations/private/mariadb/V1_0__Intial_migration.sql
@@ -10,4 +10,11 @@ CREATE TABLE IF NOT EXISTS ksvp_whitelist (
     id UUID NOT NULL,
     PRIMARY KEY (id),
     FOREIGN KEY (id) REFERENCES ${global_database}.ksvp_player(id) ON DELETE CASCADE
-) COMMENT = "This table stores who is on the whitelist.";
\ No newline at end of file
+) COMMENT = "This table stores who is on the whitelist.";
+
+CREATE TABLE IF NOT EXISTS ksvp_player_properties (
+    id UUID NOT NULL,
+    content MEDIUMBLOB NOT NULL,
+    PRIMARY KEY (id),
+    FOREIGN KEY (id) REFERENCES ${global_database}.ksvp_player(id) ON DELETE CASCADE
+);
\ No newline at end of file
