From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Ivo Quiring <falk.quiring@gmail.com>
Date: Mon, 17 Nov 2025 15:51:29 +0100
Subject: [PATCH] Overhaul Player Database


diff --git a/src/main/java/com/destroystokyo/paper/profile/PaperAuthenticationService.java b/src/main/java/com/destroystokyo/paper/profile/PaperAuthenticationService.java
index 48e774677edf17d4a99ae9ed23d1b371dab41abb..8b002f246d17c0e00f9b8ca1405e6f1254b9ba54 100644
--- a/src/main/java/com/destroystokyo/paper/profile/PaperAuthenticationService.java
+++ b/src/main/java/com/destroystokyo/paper/profile/PaperAuthenticationService.java
@@ -25,6 +25,13 @@ public class PaperAuthenticationService extends YggdrasilAuthenticationService {
 
     @Override
     public GameProfileRepository createProfileRepository() {
-        return new PaperGameProfileRepository(this.getProxy(), this.environment);
+        // Pulvinar start - overhaul player database
+        //return new PaperGameProfileRepository(this.getProxy(), this.environment);
+        return new net.kissenpvp.network.actor.PulvinarProfileRepository(
+                getProxy(),
+                environment,
+                (net.kissenpvp.network.actor.PulvinarPlayerRepository) (org.bukkit.Bukkit.getPulvinar()).playerRepository()
+        );
+        // Pulvinar end - overhaul player database
     }
 }
diff --git a/src/main/java/com/destroystokyo/paper/profile/PaperGameProfileRepository.java b/src/main/java/com/destroystokyo/paper/profile/PaperGameProfileRepository.java
index 58c1ff58e58e969a88b929663dfff785843dc1b2..f7a056fee14ed797eb284e68d1387089bbc1a425 100644
--- a/src/main/java/com/destroystokyo/paper/profile/PaperGameProfileRepository.java
+++ b/src/main/java/com/destroystokyo/paper/profile/PaperGameProfileRepository.java
@@ -31,6 +31,10 @@ public class PaperGameProfileRepository extends YggdrasilGameProfileRepository {
             }
         }
 
+        // Pulvinar start - overhaul player database
+        unfoundNames = databaseLookup(unfoundNames, callback);
+        // Pulvinar end - overhaul player database
+
         // Some things were not found.... Proceed to look up.
         if (!unfoundNames.isEmpty()) {
             String[] namesArr = unfoundNames.toArray(new String[unfoundNames.size()]);
@@ -62,4 +66,8 @@ public class PaperGameProfileRepository extends YggdrasilGameProfileRepository {
             this.callback.onProfileLookupFailed(profileName, exception);
         }
     }
+
+    // Pulvinar start - overhaul player database
+    protected Set<String> databaseLookup(Set<String> unfound, ProfileLookupCallback callback) { return unfound; }
+    // Pulvinar end - overhaul player database
 }
diff --git a/src/main/java/net/kissenpvp/base/PulvinarCore.java b/src/main/java/net/kissenpvp/base/PulvinarCore.java
index b77c1f144473808c1bcf77a0f0e4252b4c2a0959..8c5f401fc9dce754ee32e2243a603f2a1bcc2a7f 100644
--- a/src/main/java/net/kissenpvp/base/PulvinarCore.java
+++ b/src/main/java/net/kissenpvp/base/PulvinarCore.java
@@ -3,7 +3,9 @@ package net.kissenpvp.base;
 import net.kissenpvp.api.base.Pulvinar;
 import net.kissenpvp.api.database.ConnectionProvider;
 import net.kissenpvp.api.network.actor.ConsoleClient;
+import net.kissenpvp.api.network.actor.PlayerRepository;
 import net.kissenpvp.database.InternalConnectionProvider;
+import net.kissenpvp.network.actor.PulvinarPlayerRepository;
 import net.kissenpvp.network.actor.operator.PulvinarOperatorRepository;
 import net.kissenpvp.network.actor.whitelist.PulvinarWhiteListRepository;
 import org.bukkit.Bukkit;
@@ -33,6 +35,7 @@ public class PulvinarCore extends KissenCore implements Pulvinar
 
     private PulvinarOperatorRepository operatorRepository;
     private PulvinarWhiteListRepository whiteListRepository;
+    private PlayerRepository playerRepository;
 
     public PulvinarCore(@NonNull File optionsFile) throws IOException, InvalidConfigurationException, SQLException {
         config = new PulvinarConfig();
@@ -82,6 +85,7 @@ public class PulvinarCore extends KissenCore implements Pulvinar
     {
         operatorRepository = new PulvinarOperatorRepository(privateDatasource);
         whiteListRepository = new PulvinarWhiteListRepository(privateDatasource);
+        playerRepository = new PulvinarPlayerRepository(globalDatasource);
     }
 
     public void shutdown()
@@ -114,4 +118,9 @@ public class PulvinarCore extends KissenCore implements Pulvinar
     public @NonNull PulvinarWhiteListRepository whiteListRepository() {
         return whiteListRepository;
     }
+
+    @Override
+    public PlayerRepository playerRepository() {
+        return playerRepository;
+    }
 }
diff --git a/src/main/java/net/kissenpvp/network/actor/PulvinarPlayerRepository.java b/src/main/java/net/kissenpvp/network/actor/PulvinarPlayerRepository.java
new file mode 100644
index 0000000000000000000000000000000000000000..2cd70f7de4b255ca59a2e454d6f1e89d70188060
--- /dev/null
+++ b/src/main/java/net/kissenpvp/network/actor/PulvinarPlayerRepository.java
@@ -0,0 +1,130 @@
+package net.kissenpvp.network.actor;
+
+import net.kissenpvp.api.network.actor.PlayerClient;
+import net.kissenpvp.base.PulvinarCore;
+import net.minecraft.server.players.NameAndId;
+import org.bukkit.Bukkit;
+import org.jspecify.annotations.NonNull;
+
+import javax.sql.DataSource;
+import java.sql.ResultSet;
+import java.util.*;
+import java.util.concurrent.CompletableFuture;
+
+public class PulvinarPlayerRepository extends InternalPlayerRepository {
+
+    public PulvinarPlayerRepository(@NonNull DataSource dataSource) throws NullPointerException {
+        super(dataSource);
+    }
+
+    @Override
+    protected @NonNull PlayerClient toCachedEntity(@NonNull UUID uuid, @NonNull ResultSet resultSet) throws NullPointerException {
+        return Bukkit.getOfflinePlayer(uuid);
+    }
+
+    public @NonNull CompletableFuture<Collection<NameAndId>> findAllProfiles()
+    {
+        String sql = "SELECT id, username FROM ksvp_player;";
+        return CompletableFuture.supplyAsync(() -> assumeNotNull(query(sql, (statement) -> {
+            try (ResultSet resultSet = statement.executeQuery()) {
+                Collection<NameAndId> profile = new HashSet<>();
+                while(resultSet.next())
+                {
+                    UUID id = UUID.fromString(resultSet.getString("id"));
+                    String username = resultSet.getString("username");
+
+                    profile.add(new NameAndId(id, username));
+                }
+                return Collections.unmodifiableCollection(profile);
+            }
+        })));
+    }
+
+    public @NonNull Collection<NameAndId> findProfilesByNames(@NonNull String[] names) {
+        Set<NameAndId> discovered = new HashSet<>();
+        Set<String> undiscovered = new HashSet<>(Set.of(names));
+
+        for(PlayerClient cachedPlayerClient : cachedEntries().values())
+        {
+            if(!undiscovered.contains(cachedPlayerClient.username()))
+            {
+                continue;
+            }
+            undiscovered.remove(cachedPlayerClient.username());
+            discovered.add(new NameAndId(cachedPlayerClient.id(), cachedPlayerClient.username()));
+        }
+
+        String placeholders = String.join(", ", Collections.nCopies(undiscovered.size(), "?"));
+        String sql = "SELECT id, username FROM ksvp_player WHERE username IN (" + placeholders + ");";
+        query(sql, (statement) -> {
+            int index = 1;
+            for(String argument : undiscovered)
+            {
+                statement.setString(index++, argument);
+            }
+
+            try(ResultSet resultSet = statement.executeQuery())
+            {
+                while(resultSet.next())
+                {
+                    UUID id = UUID.fromString(resultSet.getString("id"));
+                    String username = resultSet.getString("username");
+
+                    discovered.add(new NameAndId(id, username));
+                }
+            }
+
+            return null;
+        });
+
+        return discovered;
+    }
+
+    public @NonNull CompletableFuture<Void> generateProfileIfAbsent(@NonNull NameAndId[] nameAndId)
+    {
+        String globalDatabase = ((PulvinarCore) Bukkit.getPulvinar()).config().databaseGlobal();
+
+        String sql = "INSERT INTO %s.ksvp_identity (link_id) VALUES (?) ON DUPLICATE KEY UPDATE link_id = link_id;";
+        String sql2 = "INSERT INTO %s.ksvp_player (id, link_id, username) VALUES (?, ?, ?) ON DUPLICATE KEY UPDATE id = id, link_id = link_id, username = ?;";
+
+        return CompletableFuture.supplyAsync(() -> {
+            query(String.format(sql, globalDatabase), (statement) -> {
+                for(NameAndId player : nameAndId)
+                {
+                    if(cached(player.id()))
+                    {
+                        continue;
+                    }
+
+                    statement.setString(1, String.valueOf(player.id()));
+                    statement.addBatch();
+                }
+                statement.executeBatch();
+                return null;
+            });
+
+            query(String.format(sql2, globalDatabase), (statement) -> {
+                for(NameAndId player : nameAndId)
+                {
+                    if(cached(player.id()))
+                    {
+                        continue;
+                    }
+
+                    String id = String.valueOf(player.id());
+
+                    statement.setString(1, id);
+                    statement.setString(2, id);
+
+                    statement.setString(3, player.name());
+                    statement.setString(4, player.name());
+                    statement.addBatch();
+                }
+                statement.executeBatch();
+                return null;
+            });
+
+            return null;
+        });
+    }
+}
diff --git a/src/main/java/net/kissenpvp/network/actor/PulvinarProfileCache.java b/src/main/java/net/kissenpvp/network/actor/PulvinarProfileCache.java
new file mode 100644
index 0000000000000000000000000000000000000000..0953215e5e4188c1badb1e9259433326ad1d9c7d
--- /dev/null
+++ b/src/main/java/net/kissenpvp/network/actor/PulvinarProfileCache.java
@@ -0,0 +1,59 @@
+package net.kissenpvp.network.actor;
+
+import com.mojang.authlib.GameProfileRepository;
+import net.minecraft.server.players.CachedUserNameToIdResolver;
+import net.minecraft.server.players.NameAndId;
+import org.jetbrains.annotations.NotNull;
+import org.jspecify.annotations.NonNull;
+
+import java.io.File;
+import java.util.Calendar;
+import java.util.Date;
+import java.util.List;
+
+public class PulvinarProfileCache extends CachedUserNameToIdResolver {
+
+    public PulvinarProfileCache(GameProfileRepository profileRepository, File file) {
+        super(profileRepository, file);
+    }
+
+    @Override
+    protected @NonNull GameProfileInfo addInternal(@NonNull NameAndId nameAndId) {
+
+        if(getRepository() instanceof PulvinarProfileRepository profileRepository)
+        {
+            profileRepository.generateProfiles(nameAndId);
+        }
+
+        return super.addInternal(nameAndId);
+    }
+
+    @Override
+    protected @NotNull List<CachedUserNameToIdResolver.GameProfileInfo> load() {
+        if(getRepository() instanceof PulvinarProfileRepository profileRepository)
+        {
+            return profileRepository.findAllProfiles().stream().map(nameAndId ->
+            {
+                Calendar instance = Calendar.getInstance();
+                instance.setTime(new Date());
+                instance.add(Calendar.MONTH, 1);
+                Date time = instance.getTime(); //TODO maybe completely remove this
+
+                return new GameProfileInfo(nameAndId, time);
+            }).toList();
+        }
+
+        return super.load();
+    }
+
+    @Override public void save(boolean asyncSave)
+    {
+        if(getRepository() instanceof PulvinarProfileRepository)
+        {
+            // this can be skipped, profiles are automatically saved when they are updated
+            return;
+        }
+
+        super.save(asyncSave);
+    }
+}
diff --git a/src/main/java/net/kissenpvp/network/actor/PulvinarProfileRepository.java b/src/main/java/net/kissenpvp/network/actor/PulvinarProfileRepository.java
new file mode 100644
index 0000000000000000000000000000000000000000..b4301a1e85a9cbd9def40a872adcf2af8da4765c
--- /dev/null
+++ b/src/main/java/net/kissenpvp/network/actor/PulvinarProfileRepository.java
@@ -0,0 +1,45 @@
+package net.kissenpvp.network.actor;
+
+import com.destroystokyo.paper.profile.PaperGameProfileRepository;
+import com.mojang.authlib.Environment;
+import com.mojang.authlib.ProfileLookupCallback;
+import net.minecraft.server.players.NameAndId;
+import org.jspecify.annotations.NonNull;
+
+import java.net.Proxy;
+import java.util.Collection;
+import java.util.HashSet;
+import java.util.Set;
+
+public class PulvinarProfileRepository extends PaperGameProfileRepository {
+
+    private final PulvinarPlayerRepository repository;
+
+    public PulvinarProfileRepository(Proxy proxy, Environment environment, PulvinarPlayerRepository repository) {
+        super(proxy, environment);
+        this.repository = repository;
+    }
+
+    @Override
+    protected Set<String> databaseLookup(Set<String> unfound, ProfileLookupCallback callback) {
+        Set<String> stillUndiscovered = new HashSet<>(unfound);
+
+        for(NameAndId current : repository.findProfilesByNames(unfound.toArray(new String[0])))
+        {
+            stillUndiscovered.remove(current.name());
+            callback.onProfileLookupSucceeded(current.name(), current.id());
+        }
+
+        return stillUndiscovered;
+    }
+
+    public void generateProfiles(NameAndId... nameAndIds)
+    {
+        repository.generateProfileIfAbsent(nameAndIds).join();
+    }
+
+    public @NonNull Collection<NameAndId> findAllProfiles()
+    {
+        return repository.findAllProfiles().join();
+    }
+}
diff --git a/src/main/java/org/bukkit/craftbukkit/CraftOfflinePlayer.java b/src/main/java/org/bukkit/craftbukkit/CraftOfflinePlayer.java
index 7283eb950a0d1c92c451a0379d7a409d147bb84c..c644d284a93d229f2e25e0b7b49cb8102146e69c 100644
--- a/src/main/java/org/bukkit/craftbukkit/CraftOfflinePlayer.java
+++ b/src/main/java/org/bukkit/craftbukkit/CraftOfflinePlayer.java
@@ -764,4 +764,47 @@ public class CraftOfflinePlayer implements OfflinePlayer, ConfigurationSerializa
         }
     }
     // Purpur end - OfflinePlayer API
+
+    // Pulvinar start - overhaul player database
+    @Override
+    public @org.jspecify.annotations.NonNull UUID id() {
+        return getUniqueId();
+    }
+
+    @Override
+    public @org.jspecify.annotations.NonNull UUID linkId() {
+        return getUniqueId(); // TODO
+    }
+
+    @Override
+    public net.kissenpvp.api.punishment.@org.jspecify.annotations.NonNull PunishmentSubscription punish(net.kissenpvp.api.punishment.@org.jspecify.annotations.NonNull Punishment punishment) {
+        return null;
+    }
+
+    @Override
+    public net.kissenpvp.api.punishment.@org.jspecify.annotations.NonNull PunishmentSubscription punish(net.kissenpvp.api.punishment.@org.jspecify.annotations.NonNull Punishment punishment, net.kyori.adventure.text.@org.jspecify.annotations.Nullable Component component) {
+        return null;
+    }
+
+    @Override
+    public java.util.@org.jspecify.annotations.NonNull List<net.kissenpvp.api.network.actor.rank.RankSubscription> rankHistory() {
+        return java.util.List.of();
+    }
+
+    @Override
+    public net.kissenpvp.api.network.actor.rank.@org.jspecify.annotations.NonNull RankSubscription rank() {
+        return null;
+    }
+
+    @Override
+    public @org.jspecify.annotations.NonNull String username() {
+        java.util.Objects.requireNonNull(getName(), "Username cannot be null.");
+        return getName();
+    }
+
+    @Override
+    public final boolean isClient() {
+        return false;
+    }
+    // Pulvinar end - overhaul player database
 }
diff --git a/src/main/java/org/bukkit/craftbukkit/command/CraftConsoleCommandSender.java b/src/main/java/org/bukkit/craftbukkit/command/CraftConsoleCommandSender.java
index 15ae2653c1809041b7ab7cb67015e82c226e9f77..1f26e86d5d2ce8f46bd09ce70b4d3de1078af104 100644
--- a/src/main/java/org/bukkit/craftbukkit/command/CraftConsoleCommandSender.java
+++ b/src/main/java/org/bukkit/craftbukkit/command/CraftConsoleCommandSender.java
@@ -105,4 +105,21 @@ public class CraftConsoleCommandSender extends ServerCommandSender implements Co
     public boolean hasPermission(org.bukkit.permissions.Permission perm) {
         return io.papermc.paper.configuration.GlobalConfiguration.get().console.hasAllPermissions || super.hasPermission(perm);
     }
+
+    // Pulvinar start - overhaul player database
+    @Override
+    public @org.jetbrains.annotations.NotNull String username() {
+        return getName();
+    }
+
+    @Override
+    public boolean isClient() {
+        return false;
+    }
+
+    @Override
+    public net.kyori.adventure.audience.@org.jetbrains.annotations.NotNull Audience audience() {
+        return this;
+    }
+    // Pulvinar end - overhaul player database
 }
diff --git a/src/main/java/org/bukkit/craftbukkit/entity/CraftHumanEntity.java b/src/main/java/org/bukkit/craftbukkit/entity/CraftHumanEntity.java
index e24eb67ad0742ac6116e7a5b84998db61dc0fc93..bbffe713655b4c5c941fefa1a78afa9ab581467b 100644
--- a/src/main/java/org/bukkit/craftbukkit/entity/CraftHumanEntity.java
+++ b/src/main/java/org/bukkit/craftbukkit/entity/CraftHumanEntity.java
@@ -919,4 +919,16 @@ public class CraftHumanEntity extends CraftLivingEntity implements HumanEntity {
     public org.bukkit.entity.Entity copy(Location location) {
         throw new UnsupportedOperationException("Cannot copy human entities");
     }
+
+    // Pulvinar start - overhaul player database
+    @Override
+    public @org.jspecify.annotations.NonNull String username() {
+        return getName();
+    }
+
+    @Override
+    public final boolean isClient() {
+        return true;
+    }
+    // Pulvinar end - overhaul player database
 }
diff --git a/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java b/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java
index dd0dcc6ae5cfb72df0483417b22eddd90febc9e6..cfa61480fd1a8208846f9c6fd28a8faad81a2152 100644
--- a/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java
+++ b/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java
@@ -3467,4 +3467,41 @@ public class CraftPlayer extends CraftHumanEntity implements Player, PluginMessa
             entity.remove();
         }
     }
+
+    // Pulvinar start - overhaul player database
+    @Override
+    public @NonNull UUID id() {
+        return getUniqueId();
+    }
+
+    @Override
+    public net.kyori.adventure.audience.@NonNull Audience audience() {
+        return this;
+    }
+
+    @Override
+    public @NonNull UUID linkId() {
+        return getUniqueId(); // TODO
+    }
+
+    @Override
+    public net.kissenpvp.api.punishment.@NonNull PunishmentSubscription punish(net.kissenpvp.api.punishment.@NonNull Punishment punishment) {
+        return null;
+    }
+
+    @Override
+    public net.kissenpvp.api.punishment.@NonNull PunishmentSubscription punish(net.kissenpvp.api.punishment.@NonNull Punishment punishment, net.kyori.adventure.text.@Nullable Component component) {
+        return null;
+    }
+
+    @Override
+    public @NonNull List<net.kissenpvp.api.network.actor.rank.RankSubscription> rankHistory() {
+        return List.of();
+    }
+
+    @Override
+    public net.kissenpvp.api.network.actor.rank.@NonNull RankSubscription rank() {
+        return null;
+    }
+    // Pulvinar end - overhaul player database
 }
