From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Ivo Quiring <falk.quiring@gmail.com>
Date: Sat, 29 Nov 2025 09:49:20 +0100
Subject: [PATCH] Implement Rank System


diff --git a/src/main/java/net/kissenpvp/base/PulvinarCore.java b/src/main/java/net/kissenpvp/base/PulvinarCore.java
index 3900d93fb3827dcb66df61b6bc64a881a509951c..681697b4dff7302be81572d104d466dd90e53980 100644
--- a/src/main/java/net/kissenpvp/base/PulvinarCore.java
+++ b/src/main/java/net/kissenpvp/base/PulvinarCore.java
@@ -2,11 +2,14 @@ package net.kissenpvp.base;
 
 import net.kissenpvp.api.base.Pulvinar;
 import net.kissenpvp.api.database.ConnectionProvider;
+import net.kissenpvp.api.database.Repository;
 import net.kissenpvp.api.network.actor.ConsoleClient;
 import net.kissenpvp.api.network.actor.PlayerRepository;
+import net.kissenpvp.api.network.actor.rank.Rank;
 import net.kissenpvp.database.mariadb.MariaDBConnectionProvider;
 import net.kissenpvp.network.actor.PulvinarPlayerRepository;
 import net.kissenpvp.network.actor.operator.PulvinarOperatorRepository;
+import net.kissenpvp.network.actor.rank.repository.MariaDBRankRepository;
 import net.kissenpvp.network.actor.whitelist.PulvinarWhiteListRepository;
 import org.bukkit.Bukkit;
 import org.bukkit.configuration.InvalidConfigurationException;
@@ -36,6 +39,7 @@ public class PulvinarCore extends KissenCore implements Pulvinar
     private PulvinarOperatorRepository operatorRepository;
     private PulvinarWhiteListRepository whiteListRepository;
     private PlayerRepository playerRepository;
+    private Repository<String, Rank> rankRepository;
 
     public PulvinarCore(@NonNull File optionsFile) throws IOException, InvalidConfigurationException, SQLException {
         config = new PulvinarConfig();
@@ -88,6 +92,7 @@ public class PulvinarCore extends KissenCore implements Pulvinar
         operatorRepository = new PulvinarOperatorRepository(privateDatasource);
         whiteListRepository = new PulvinarWhiteListRepository(privateDatasource);
         playerRepository = new PulvinarPlayerRepository(globalDatasource);
+        rankRepository = new MariaDBRankRepository(globalDatasource);
     }
 
     public void shutdown()
@@ -126,4 +131,9 @@ public class PulvinarCore extends KissenCore implements Pulvinar
     public @NonNull PlayerRepository playerRepository() {
         return playerRepository;
     }
+
+    @Override
+    public @NonNull Repository<String, Rank> rankRepository() {
+        return rankRepository;
+    }
 }
