From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Ivo Quiring <falk.quiring@gmail.com>
Date: Sat, 29 Nov 2025 09:50:40 +0100
Subject: [PATCH] Overhaul Ban Database


diff --git a/src/main/java/net/kissenpvp/base/PulvinarCore.java b/src/main/java/net/kissenpvp/base/PulvinarCore.java
index 375a00e8024873cea7d4fbb32b6a80a921e75f1f..3effbe8015c391798530bcfd60ec27313fae7161 100644
--- a/src/main/java/net/kissenpvp/base/PulvinarCore.java
+++ b/src/main/java/net/kissenpvp/base/PulvinarCore.java
@@ -7,11 +7,15 @@ import net.kissenpvp.api.network.actor.ConsoleClient;
 import net.kissenpvp.api.network.actor.PlayerRepository;
 import net.kissenpvp.api.network.actor.rank.Rank;
 import net.kissenpvp.api.network.actor.rank.RankSubscriptionRepository;
+import net.kissenpvp.api.punishment.Punishment;
+import net.kissenpvp.api.punishment.PunishmentSubscriptionRepository;
 import net.kissenpvp.database.InternalConnectionProvider;
 import net.kissenpvp.network.actor.PulvinarPlayerRepository;
 import net.kissenpvp.network.actor.operator.PulvinarOperatorRepository;
 import net.kissenpvp.network.actor.rank.repository.InternalRankRepository;
 import net.kissenpvp.network.actor.whitelist.PulvinarWhiteListRepository;
+import net.kissenpvp.punishment.repository.InternalPunishmentRepository;
+import net.kissenpvp.punishment.repository.InternalPunishmentSubscriptionRepository;
 import org.bukkit.Bukkit;
 import org.bukkit.configuration.InvalidConfigurationException;
 import org.flywaydb.core.Flyway;
@@ -39,6 +43,8 @@ public class PulvinarCore extends KissenCore implements Pulvinar
     private PulvinarWhiteListRepository whiteListRepository;
     private PlayerRepository playerRepository;
     private Repository<String, Rank> rankRepository;
+    private Repository<Integer, Punishment> punishmentRepository;
+    private PunishmentSubscriptionRepository punishmentSubscriptionRepository;
 
     public PulvinarCore(@NonNull File optionsFile) throws IOException, InvalidConfigurationException, SQLException {
         config = new PulvinarConfig();
@@ -86,6 +92,8 @@ public class PulvinarCore extends KissenCore implements Pulvinar
         whiteListRepository = new PulvinarWhiteListRepository(privateDatasource);
         playerRepository = new PulvinarPlayerRepository(globalDatasource);
         rankRepository = new InternalRankRepository(globalDatasource);
+        punishmentRepository = new InternalPunishmentRepository(globalDatasource);
+        punishmentSubscriptionRepository = new InternalPunishmentSubscriptionRepository(globalDatasource);
     }
 
     public void shutdown()
@@ -128,4 +136,14 @@ public class PulvinarCore extends KissenCore implements Pulvinar
     public Repository<String, Rank> rankRepository() {
         return rankRepository;
     }
+
+    @Override
+    public Repository<Integer, Punishment> punishmentRepository() {
+        return punishmentRepository;
+    }
+
+    @Override
+    public PunishmentSubscriptionRepository punishmentSubscriptionRepository() {
+        return punishmentSubscriptionRepository;
+    }
 }
