From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Ivo Quiring <falk.quiring@gmail.com>
Date: Mon, 11 May 2026 17:08:50 +0200
Subject: [PATCH] Add Kissen Base


diff --git a/net/minecraft/server/Main.java b/net/minecraft/server/Main.java
index 1862df44a3319e459b547cebffb16972674856d6..a890268c784ae50992dcbff9b0773dfacbbd0aa3 100644
--- a/net/minecraft/server/Main.java
+++ b/net/minecraft/server/Main.java
@@ -178,6 +178,13 @@ public class Main {
 
             org.spigotmc.SpigotConfig.disabledAdvancements = spigotConfiguration.getStringList("advancements.disabled"); // Paper - fix SPIGOT-5885, must be set early in init
 
+            // Pulvinar start - add kissen base
+            LOGGER.info("Pulvinar is now going to hook in pulvinar.");
+            java.io.File pulvinarSettings = (java.io.File) options.valueOf("pulvinar-settings");
+            net.kissenpvp.api.base.Pulvinar pulvinar = new net.kissenpvp.base.PulvinarCore(pulvinarSettings);
+            org.bukkit.Bukkit.setPulvinar(pulvinar);
+            // Pulvinar end - add kissen base
+
             // Paper start - fix SPIGOT-5824
             File universePath;
             File userCacheFile = new File(Services.USERID_CACHE_FILE);
