From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Ivo Quiring <falk.quiring@gmail.com>
Date: Mon, 11 May 2026 17:09:47 +0200
Subject: [PATCH] Allow Access To Stored User List Map


diff --git a/net/minecraft/server/players/StoredUserList.java b/net/minecraft/server/players/StoredUserList.java
index d97d79e65c672e31e408f7627189ab65f5aedc2c..1ba6feae4309df3ce19b5bb53ba716e1e073b7d1 100644
--- a/net/minecraft/server/players/StoredUserList.java
+++ b/net/minecraft/server/players/StoredUserList.java
@@ -160,4 +160,11 @@ public abstract class StoredUserList<K, V extends StoredUserEntry<K>> {
             // Spigot end
         }
     }
+
+    // Pulvinar start - allow protected access to map
+    protected Map<String, V> getMap()
+    {
+        return map;
+    }
+    // Pulvinar end - allow protected access to map
 }
