From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Ivo Quiring <falk.quiring@gmail.com>
Date: Mon, 11 May 2026 17:19:20 +0200
Subject: [PATCH] Overhaul Player Database


diff --git a/net/minecraft/server/Services.java b/net/minecraft/server/Services.java
index c3c23a152c098e4dddfe0d0d10676fc20a35f0e0..7d782af71352e094d9e9b7a002045327433dd99f 100644
--- a/net/minecraft/server/Services.java
+++ b/net/minecraft/server/Services.java
@@ -41,7 +41,7 @@ public record Services(
     public static Services create(final YggdrasilAuthenticationService serviceAccess, final File nameCacheDir, final File userCacheFile, final joptsimple.OptionSet options) throws Exception { // Paper - add options to load paper config files; add userCacheFile parameter
         MinecraftSessionService sessionService = serviceAccess.createMinecraftSessionService();
         GameProfileRepository profileRepository = serviceAccess.createProfileRepository();
-        UserNameToIdResolver profileCache = new CachedUserNameToIdResolver(profileRepository, userCacheFile); // Paper
+        UserNameToIdResolver profileCache = new net.kissenpvp.network.actor.PulvinarProfileCache(profileRepository, userCacheFile); // new CachedUserNameToIdResolver(gameProfileRepository, userCacheFile); // Paper // Pulvinar - overhaul player database
         // Paper start - load paper config files from cli options
         final java.nio.file.Path legacyConfigPath = ((File) options.valueOf("paper-settings")).toPath();
         final java.nio.file.Path configDirPath = ((File) options.valueOf("paper-settings-directory")).toPath();
diff --git a/net/minecraft/server/players/CachedUserNameToIdResolver.java b/net/minecraft/server/players/CachedUserNameToIdResolver.java
index ce26768b951a5cd3e3337ef5347ce1537d593486..6f205b1e7f44ea725ae10dda6d2b713fc722d06c 100644
--- a/net/minecraft/server/players/CachedUserNameToIdResolver.java
+++ b/net/minecraft/server/players/CachedUserNameToIdResolver.java
@@ -91,7 +91,7 @@ public class CachedUserNameToIdResolver implements UserNameToIdResolver {
         this.addInternal(nameAndId);
     }
 
-    private CachedUserNameToIdResolver.GameProfileInfo addInternal(final NameAndId profile) {
+    protected CachedUserNameToIdResolver.GameProfileInfo addInternal(final NameAndId profile) { // Pulvinar - Overhaul Player Database  - allow to override
         Calendar c = Calendar.getInstance(TimeZone.getDefault(), Locale.ROOT);
         c.setTime(new Date());
         c.add(2, 1);
@@ -177,7 +177,7 @@ public class CachedUserNameToIdResolver implements UserNameToIdResolver {
         return new SimpleDateFormat("yyyy-MM-dd HH:mm:ss Z", Locale.ROOT);
     }
 
-    private List<CachedUserNameToIdResolver.GameProfileInfo> load() {
+    protected List<CachedUserNameToIdResolver.GameProfileInfo> load() { // Pulvinar - Overhaul Player Database - allow to override
         List<CachedUserNameToIdResolver.GameProfileInfo> result = Lists.newArrayList();
 
         try (Reader reader = Files.newReader(this.file, StandardCharsets.UTF_8)) {
@@ -278,12 +278,12 @@ public class CachedUserNameToIdResolver implements UserNameToIdResolver {
         return Optional.empty();
     }
 
-    private static class GameProfileInfo {
+    protected static class GameProfileInfo { // Pulvinar - overhaul player database - allow children to access
         private final NameAndId nameAndId;
         private final Date expirationDate;
         private volatile long lastAccess;
 
-        private GameProfileInfo(final NameAndId nameAndId, final Date expirationDate) {
+        public GameProfileInfo(final NameAndId nameAndId, final Date expirationDate) { // Pulvinar - overhaul player database - allow children to access
             this.nameAndId = nameAndId;
             this.expirationDate = expirationDate;
         }
@@ -304,4 +304,11 @@ public class CachedUserNameToIdResolver implements UserNameToIdResolver {
             return this.lastAccess;
         }
     }
+
+    // Pulvinar start - overhaul player data
+    protected GameProfileRepository getRepository()
+    {
+        return profileRepository;
+    }
+    // Pulvinar end - overhaul player data
 }
diff --git a/net/minecraft/world/level/storage/PlayerDataStorage.java b/net/minecraft/world/level/storage/PlayerDataStorage.java
index eddb6ebe3fa50a818e54079b73ba117be0abd195..2e9db95e4bbdb09a5dae445df8c700bc4a03ff92 100644
--- a/net/minecraft/world/level/storage/PlayerDataStorage.java
+++ b/net/minecraft/world/level/storage/PlayerDataStorage.java
@@ -35,9 +35,25 @@ public class PlayerDataStorage {
         try (ProblemReporter.ScopedCollector reporter = new ProblemReporter.ScopedCollector(player.problemPath(), LOGGER)) {
             TagValueOutput output = TagValueOutput.createWithContext(reporter, player.registryAccess());
             player.saveWithoutId(output);
+
+            // Pulvinar start - overhaul player database
+            boolean disableDatabaseStorage = ((net.kissenpvp.base.PulvinarCore) org.bukkit.Bukkit.getPulvinar()).config().disableDatabaseStorage();
+            CompoundTag dataToStore = output.buildResult();
+            if (!disableDatabaseStorage) {
+                net.kissenpvp.network.actor.PulvinarPlayerRepository
+                playerRepository = ((net.kissenpvp.network.actor.PulvinarPlayerRepository) org.bukkit.Bukkit.getPulvinar().playerRepository());
+
+                try (java.io.ByteArrayOutputStream byteArrayOutputStream = new java.io.ByteArrayOutputStream()) {
+                    net.minecraft.nbt.NbtIo.writeCompressed(dataToStore, byteArrayOutputStream);
+                    byte[] data = byteArrayOutputStream.toByteArray();
+                    playerRepository.saveProperties(player.getUUID(), data);
+                }
+            }
+            // Pulvinar end - overhaul player database
+
             Path playerDirPath = this.playerDir.toPath();
             Path tmpFile = Files.createTempFile(playerDirPath, player.getStringUUID() + "-", ".dat");
-            CompoundTag dataToStore = output.buildResult();
+            //CompoundTag dataToStore = output.buildResult(); // Pulvinar - overhaul player database
             NbtIo.writeCompressed(dataToStore, tmpFile);
             Path realFile = playerDirPath.resolve(player.getStringUUID() + ".dat");
             Path oldFile = playerDirPath.resolve(player.getStringUUID() + ".dat_old");
@@ -62,6 +78,25 @@ public class PlayerDataStorage {
     }
 
     private Optional<CompoundTag> load(final NameAndId nameAndId, final String suffix) {
+
+        // Pulvinar start - overhaul player database
+        boolean disableDatabaseStorage = ((net.kissenpvp.base.PulvinarCore) org.bukkit.Bukkit.getPulvinar()).config().disableDatabaseStorage();
+        if (!disableDatabaseStorage) {
+            net.kissenpvp.network.actor.PulvinarPlayerRepository playerRepository = ((net.kissenpvp.network.actor.PulvinarPlayerRepository) org.bukkit.Bukkit.getPulvinar().playerRepository());
+            Optional<java.sql.Blob> data = playerRepository.findProperties(nameAndId.id()).join();
+
+            if (data.isPresent()) {
+                java.sql.Blob blob = data.get();
+                try {
+                    return Optional.of(NbtIo.readCompressed(blob.getBinaryStream(), NbtAccounter.unlimitedHeap()));
+                } catch (java.sql.SQLException | java.io.IOException exception) {
+                    LOGGER.warn("Failed to load player data from database for player {}.", nameAndId.name());
+                }
+            }
+            LOGGER.info("Fall back to file system for player {}.", nameAndId.name());
+        }
+        // Pulvinar end - overhaul player database
+
         File realFile = new File(this.playerDir, nameAndId.id() + suffix);
         // Spigot start
         boolean usingWrongFile = false;
