From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Ivo Quiring <falk.quiring@gmail.com>
Date: Mon, 17 Nov 2025 15:50:46 +0100
Subject: [PATCH] Overhaul Player Database


diff --git a/net/minecraft/server/Services.java b/net/minecraft/server/Services.java
index db4ad03b30e1e7dc8856ce759b610402acb076e0..9e842751f5f5f74ddd96c895d3667feea966dd3d 100644
--- a/net/minecraft/server/Services.java
+++ b/net/minecraft/server/Services.java
@@ -41,7 +41,7 @@ public record Services(
     public static Services create(YggdrasilAuthenticationService authenticationService, File profileRepository, File userCacheFile, joptsimple.OptionSet optionSet) throws Exception { // Paper - add optionset to load paper config files; add userCacheFile parameter
         MinecraftSessionService minecraftSessionService = authenticationService.createMinecraftSessionService();
         GameProfileRepository gameProfileRepository = authenticationService.createProfileRepository();
-        UserNameToIdResolver userNameToIdResolver = new CachedUserNameToIdResolver(gameProfileRepository, userCacheFile); // Paper
+        UserNameToIdResolver userNameToIdResolver = new net.kissenpvp.network.actor.PulvinarProfileCache(gameProfileRepository, userCacheFile); // new CachedUserNameToIdResolver(gameProfileRepository, userCacheFile); // Paper // Pulvinar - overhaul player database
         // Paper start - load paper config files from cli options
         final java.nio.file.Path legacyConfigPath = ((File) optionSet.valueOf("paper-settings")).toPath();
         final java.nio.file.Path configDirPath = ((File) optionSet.valueOf("paper-settings-directory")).toPath();
diff --git a/net/minecraft/server/players/CachedUserNameToIdResolver.java b/net/minecraft/server/players/CachedUserNameToIdResolver.java
index 6d3f496881492ac67465746984d6ef8302750305..56adf03920eb4c56305793f62044dc442b71761f 100644
--- a/net/minecraft/server/players/CachedUserNameToIdResolver.java
+++ b/net/minecraft/server/players/CachedUserNameToIdResolver.java
@@ -177,7 +177,7 @@ public class CachedUserNameToIdResolver implements UserNameToIdResolver {
         return new SimpleDateFormat("yyyy-MM-dd HH:mm:ss Z", Locale.ROOT);
     }
 
-    private List<CachedUserNameToIdResolver.GameProfileInfo> load() {
+    protected List<CachedUserNameToIdResolver.GameProfileInfo> load() { // Pulvinar - overhaul player database - allow to override
         List<CachedUserNameToIdResolver.GameProfileInfo> list = Lists.newArrayList();
 
         try {
@@ -284,12 +284,12 @@ public class CachedUserNameToIdResolver implements UserNameToIdResolver {
         return Optional.empty();
     }
 
-    static class GameProfileInfo {
+    protected static class GameProfileInfo { // Pulvinar - overhaul player database - allow children to access
         private final NameAndId nameAndId;
         final Date expirationDate;
         private volatile long lastAccess;
 
-        GameProfileInfo(NameAndId nameAndId, Date expirationDate) {
+        public GameProfileInfo(NameAndId nameAndId, Date expirationDate) { // Pulvinar - overhaul player database - allow children to access
             this.nameAndId = nameAndId;
             this.expirationDate = expirationDate;
         }
@@ -310,4 +310,11 @@ public class CachedUserNameToIdResolver implements UserNameToIdResolver {
             return this.lastAccess;
         }
     }
+
+    // Pulvinar start - overhaul player data
+    protected GameProfileRepository getRepository()
+    {
+        return profileRepository;
+    }
+    // Pulvinar end - overhaul player data
 }
