From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Ivo Quiring <falk.quiring@gmail.com>
Date: Mon, 11 May 2026 17:21:00 +0200
Subject: [PATCH] Implement System-Message Event


diff --git a/net/minecraft/server/level/ServerPlayer.java b/net/minecraft/server/level/ServerPlayer.java
index a41e8e61df1f3d7a861cd1b4060bc3ec424d45bd..1112f88dd14dc38c11c77affc7a789dd08cda9b7 100644
--- a/net/minecraft/server/level/ServerPlayer.java
+++ b/net/minecraft/server/level/ServerPlayer.java
@@ -2501,14 +2501,26 @@ public class ServerPlayer extends Player implements ca.spottedleaf.moonrise.patc
 
     public void sendSystemMessage(final Component message, final boolean overlay) {
         if (this.acceptsSystemMessages(overlay)) {
+
+            // Pulvinar start - implement system message event
+            net.kyori.adventure.text.Component adventure = io.papermc.paper.adventure.PaperAdventure.asAdventure(message);
+            net.kissenpvp.api.event.SystemMessageEvent messageEvent = new net.kissenpvp.api.event.SystemMessageEvent(adventure);
+            org.bukkit.Bukkit.getPluginManager().callEvent(messageEvent);
+
+            if (messageEvent.isCancelled()) {
+                return;
+            }
+            Component outputMessage = io.papermc.paper.adventure.PaperAdventure.asVanilla(messageEvent.getMessage());
+            // Pulvinar end - implement system message event
+
             this.connection
                 .send(
-                    new ClientboundSystemChatPacket(message, overlay),
+                        new ClientboundSystemChatPacket(/*message*/ outputMessage, overlay), // Pulvinar - implement system message event,
                     PacketSendListener.exceptionallySend(
                         () -> {
                             if (this.acceptsSystemMessages(false)) {
                                 int truncatedMessageSize = 256;
-                                String contents = message.getString(256);
+                                String contents = /*message*/outputMessage.getString(256); // Pulvinar - implement system message event
                                 Component contentsMsg = Component.literal(contents).withStyle(ChatFormatting.YELLOW);
                                 return new ClientboundSystemChatPacket(
                                     Component.translatable("multiplayer.message_not_delivered", contentsMsg).withStyle(ChatFormatting.RED), false
