--- a/src/main/java/io/papermc/paper/configuration/GlobalConfiguration.java
+++ b/src/main/java/io/papermc/paper/configuration/GlobalConfiguration.java
@@ -29,7 +_,7 @@
 @SuppressWarnings({"CanBeFinal", "FieldCanBeLocal", "FieldMayBeFinal", "NotNullFieldNotInitialized", "InnerClassMayBeStatic"})
 public class GlobalConfiguration extends ConfigurationPart {
     private static final Logger LOGGER = LogUtils.getLogger();
-    static final int CURRENT_VERSION = 31; // (when you change the version, change the comment, so it conflicts on rebases): allow-nether property to config
+    static final int CURRENT_VERSION = 30; // (when you change the version, change the comment, so it conflicts on rebases): upgrade packet to use ids
     private static GlobalConfiguration instance;
     public static boolean isFirstStart = false;
     public static GlobalConfiguration get() {
@@ -224,10 +_,27 @@
 
         public int ioThreads = -1;
         public int workerThreads = -1;
+        public String genParallelism = "default";
 
         @PostProcess
         private void postProcess() {
             ca.spottedleaf.moonrise.common.util.MoonriseCommon.adjustWorkerThreads(this.workerThreads, this.ioThreads);
+            String newChunkSystemGenParallelism = this.genParallelism;
+            if (newChunkSystemGenParallelism.equalsIgnoreCase("default")) {
+                newChunkSystemGenParallelism = "true";
+            }
+
+            final boolean useParallelGen;
+            if (newChunkSystemGenParallelism.equalsIgnoreCase("on") || newChunkSystemGenParallelism.equalsIgnoreCase("enabled")
+                || newChunkSystemGenParallelism.equalsIgnoreCase("true")) {
+                useParallelGen = true;
+            } else if (newChunkSystemGenParallelism.equalsIgnoreCase("off") || newChunkSystemGenParallelism.equalsIgnoreCase("disabled")
+                || newChunkSystemGenParallelism.equalsIgnoreCase("false")) {
+                useParallelGen = false;
+            } else {
+                throw new IllegalStateException("Invalid option for gen-parallelism: must be one of [on, off, enabled, disabled, true, false, default]");
+            }
+            FeatureHooks.initChunkTaskScheduler(useParallelGen);
         }
     }
 
@@ -343,8 +_,6 @@
         public IntOr.Default xpOrbGroupsPerArea = IntOr.Default.USE_DEFAULT;
         @Comment("See Fix MC-163962; prevent villager demand from going negative.")
         public boolean preventNegativeVillagerDemand = false;
-        @Comment("Whether the nether dimension is enabled and will be loaded.")
-        public boolean enableNether = true;
     }
 
     public BlockUpdates blockUpdates;
