--- a/src/test/java/io/papermc/paper/adventure/AdventureCodecsTest.java
+++ b/src/test/java/io/papermc/paper/adventure/AdventureCodecsTest.java
@@ -59,7 +_,6 @@
 import static net.kyori.adventure.text.Component.blockNBT;
 import static net.kyori.adventure.text.Component.entityNBT;
 import static net.kyori.adventure.text.Component.keybind;
-import static net.kyori.adventure.text.Component.object;
 import static net.kyori.adventure.text.Component.score;
 import static net.kyori.adventure.text.Component.selector;
 import static net.kyori.adventure.text.Component.storageNBT;
@@ -72,9 +_,6 @@
 import static net.kyori.adventure.text.format.Style.style;
 import static net.kyori.adventure.text.format.TextColor.color;
 import static net.kyori.adventure.text.minimessage.MiniMessage.miniMessage;
-import static net.kyori.adventure.text.object.ObjectContents.playerHead;
-import static net.kyori.adventure.text.object.ObjectContents.sprite;
-import static net.kyori.adventure.text.object.PlayerHeadObjectContents.property;
 import static org.junit.jupiter.api.Assertions.assertEquals;
 import static org.junit.jupiter.api.Assertions.assertNotNull;
 import static org.junit.jupiter.api.Assertions.assertThrows;
@@ -326,8 +_,7 @@
     @Target({ElementType.PARAMETER, ElementType.ANNOTATION_TYPE})
     @MethodParameterSource({
         "testTexts", "testTranslatables", "testKeybinds", "testScores",
-        "testSelectors", "testBlockNbts", "testEntityNbts", "testStorageNbts",
-        "testObjects"
+        "testSelectors", "testBlockNbts", "testEntityNbts", "testStorageNbts"
     })
     @interface TestComponents {
     }
@@ -431,20 +_,6 @@
             storageNBT().nbtPath("abc").storage(key("doom:apple")).build(),
             storageNBT().nbtPath("abc").storage(key("doom:apple")).separator(text(", ")).build(),
             storageNBT().nbtPath("abc").storage(key("doom:apple")).interpret(true).build()
-        );
-    }
-
-    static List<Component> testObjects() {
-        return List.of(
-            object(sprite(key("block/fire_0"))),
-            object(sprite(key("ae2", "blocks"), key("block/controller"))),
-            object(sprite(key("blocks"), key("block/fire_0"))),
-            object(playerHead("pig")),
-            object(playerHead().build()),
-            object(playerHead().name("sheep").id(UUID.randomUUID()).build()),
-            object(playerHead(UUID.randomUUID())),
-            object(playerHead().profileProperties(List.of(property("textures", "abc"))).build()),
-            object(playerHead().texture(key("apples")).build())
         );
     }
 }
