--- a/src/test/java/org/bukkit/support/RegistryHelper.java
+++ b/src/test/java/org/bukkit/support/RegistryHelper.java
@@ -23,7 +_,6 @@
 import net.minecraft.tags.TagLoader;
 import net.minecraft.world.flag.FeatureFlagSet;
 import net.minecraft.world.level.biome.Biome;
-import net.minecraft.world.level.chunk.PalettedContainerFactory;
 import org.bukkit.Keyed;
 import org.bukkit.NamespacedKey;
 
@@ -32,7 +_,6 @@
     private static ReloadableServerResources dataPack;
     private static RegistryAccess.Frozen registry;
     private static Registry<Biome> biomes;
-    private static PalettedContainerFactory palettedContainerFactory;
 
     private RegistryHelper() {
     }
@@ -58,13 +_,6 @@
         return RegistryHelper.biomes;
     }
 
-    public static PalettedContainerFactory palettedContainerFactory() {
-        if (RegistryHelper.palettedContainerFactory == null) {
-            RegistryHelper.throwError("palettedContainerFactory");
-        }
-        return RegistryHelper.palettedContainerFactory;
-    }
-
     public static RegistryAccess.Frozen createRegistry(FeatureFlagSet featureFlagSet) {
         MultiPackResourceManager ireloadableresourcemanager = RegistryHelper.createResourceManager(featureFlagSet);
         // add tags and loot tables for unit tests
@@ -107,8 +_,6 @@
         RegistryHelper.dataPack.updateStaticRegistryTags();
         // Biome shortcut
         RegistryHelper.biomes = RegistryHelper.registry.lookupOrThrow(Registries.BIOME);
-        // PalettedContainerFactory shortcut
-        RegistryHelper.palettedContainerFactory = PalettedContainerFactory.create(RegistryHelper.registry);
     }
 
     public static <T extends Keyed> Class<T> updateClass(Class<T> aClass, NamespacedKey key) {
