--- a/src/main/java/org/bukkit/craftbukkit/util/Commodore.java
+++ b/src/main/java/org/bukkit/craftbukkit/util/Commodore.java
@@ -375,9 +_,6 @@
                                     case "SCUTE":
                                         name = "TURTLE_SCUTE";
                                         break;
-                                    case "CHAIN":
-                                        name = "IRON_CHAIN";
-                                        break;
                                 }
                             }
 
